<!DOCTYPE html>
<html lang="zh-CN">

  <head>
  <meta charset="utf-8">
  <meta name="author" content="zchengsite, 1451426471@qq.com" />
  <meta name="viewport" content="width=device-width, initial-scale=1" />
  
  
  <title>scRNA的细胞质控 | Hexo</title>

  
    <link rel="apple-touch-icon" href="/images/favicon.png">
    <link rel="icon" href="/images/favicon.png">
  

  <!-- Raleway-Font -->
  <link href="https://fonts.googleapis.com/css?family=Raleway&display=swap" rel="stylesheet">

  <!-- hexo site css -->
  
<link rel="stylesheet" href="/css/base.css">
<link rel="stylesheet" href="/iconfont/iconfont.css">
<link rel="stylesheet" href="/css/github-markdown.css">
<link rel="stylesheet" href="/css/highlight.css">


  <!-- jquery3.3.1 -->
  <script src="https://cdn.bootcss.com/jquery/3.3.1/jquery.min.js"></script>

  <!-- fancybox -->
  <link href="https://cdn.bootcss.com/fancybox/3.5.2/jquery.fancybox.min.css" rel="stylesheet">
  <script async src="https://cdn.bootcss.com/fancybox/3.5.2/jquery.fancybox.min.js"></script>
  
<script src="/js/fancybox.js"></script>


  

<meta name="generator" content="Hexo 5.2.0"><link rel="alternate" href="/atom.xml" title="Hexo" type="application/atom+xml">
<style>mjx-container[jax="SVG"] {
  direction: ltr;
}

mjx-container[jax="SVG"] > svg {
  overflow: visible;
}

mjx-container[jax="SVG"] > svg a {
  fill: blue;
  stroke: blue;
}

mjx-container[jax="SVG"][display="true"] {
  display: block;
  text-align: center;
  margin: 1em 0;
}

mjx-container[jax="SVG"][justify="left"] {
  text-align: left;
}

mjx-container[jax="SVG"][justify="right"] {
  text-align: right;
}

g[data-mml-node="merror"] > g {
  fill: red;
  stroke: red;
}

g[data-mml-node="merror"] > rect[data-background] {
  fill: yellow;
  stroke: none;
}

g[data-mml-node="mtable"] > line[data-line] {
  stroke-width: 70px;
  fill: none;
}

g[data-mml-node="mtable"] > rect[data-frame] {
  stroke-width: 70px;
  fill: none;
}

g[data-mml-node="mtable"] > .mjx-dashed {
  stroke-dasharray: 140;
}

g[data-mml-node="mtable"] > .mjx-dotted {
  stroke-linecap: round;
  stroke-dasharray: 0,140;
}

g[data-mml-node="mtable"] > svg {
  overflow: visible;
}

[jax="SVG"] mjx-tool {
  display: inline-block;
  position: relative;
  width: 0;
  height: 0;
}

[jax="SVG"] mjx-tool > mjx-tip {
  position: absolute;
  top: 0;
  left: 0;
}

mjx-tool > mjx-tip {
  display: inline-block;
  padding: .2em;
  border: 1px solid #888;
  font-size: 70%;
  background-color: #F8F8F8;
  color: black;
  box-shadow: 2px 2px 5px #AAAAAA;
}

g[data-mml-node="maction"][data-toggle] {
  cursor: pointer;
}

mjx-status {
  display: block;
  position: fixed;
  left: 1em;
  bottom: 1em;
  min-width: 25%;
  padding: .2em .4em;
  border: 1px solid #888;
  font-size: 90%;
  background-color: #F8F8F8;
  color: black;
}

foreignObject[data-mjx-xml] {
  font-family: initial;
  line-height: normal;
  overflow: visible;
}

.MathJax path {
  stroke-width: 3;
}

mjx-container[display="true"] {
  overflow: auto hidden;
}

mjx-container[display="true"] + br {
  display: none;
}
</style></head>


  <body>
    <div id="app">
      <div class="header">
  <div class="avatar">
    <a href="/">
      <!-- 头像取消懒加载，添加no-lazy -->
      
        <img src="/images/mooner.jpg" alt="">
      
    </a>
    <div class="nickname"><a href="/">mooner</a></div>
  </div>
  <div class="navbar">
    <ul>
      
        <li class="nav-item" data-path="/">
          <a href="/">主页</a>
        </li>
      
        <li class="nav-item" data-path="/archives/">
          <a href="/archives/">归档</a>
        </li>
      
        <li class="nav-item" data-path="/tags/">
          <a href="/tags/">标签</a>
        </li>
      
        <li class="nav-item" data-path="/about/">
          <a href="/about/">关于</a>
        </li>
      
    </ul>
  </div>
</div>


<script src="/js/activeNav.js"></script>



      <div class="flex-container">
        <!-- 文章详情页，展示文章具体内容，url形式：https://yoursite/文章标题/ -->
<!-- 同时为「标签tag」，「朋友friend」，「分类categories」，「关于about」页面的承载页面，具体展示取决于page.type -->



  

  

  

  
  <!-- 文章内容页 url形式：https://yoursite/文章标题/ -->
  <div class="container post-details" id="post-details">
    <div class="post-content">
      <div class="post-title">scRNA的细胞质控</div>
      <div class="post-attach">
        <span class="post-pubtime">
          <i class="iconfont icon-updatetime" title="更新时间"></i>
          2020-08-12
        </span>
        
              <span class="post-tags">
                <i class="iconfont icon-tags" title="标签"></i>
                
                <span class="span--tag">
                  <a href="/tags/SingleCell/" title="SingleCell">
                    <b>#</b> SingleCell
                  </a>
                </span>
                
              </span>
          
      </div>
      <div class="markdown-body">
        <h2 id="动机">动机</h2>
<p>scRNA-seq数据的低质量文库可能来自于：细胞分选环节的破坏、文库制备失误（没有足够的反转录或PCR次数）… 表现在：细胞总表达量低、基本没有表达的基因、高线粒体或spike-in占比。</p>
<p>这些低质量的库是有问题的，因为它们可能在下游分析中导致误导的结果 。</p>
<ul>
<li><strong>聚类问题：</strong> 低质量的细胞会聚集成一群，对结果的解释造成干扰，因为从这群细胞中得不到什么有用的信息，但是它的确也是一群。这种现象产生的原因有可能是：细胞破坏以后，线粒体或核RNAs占比升高。 最差的情况就是：不同类型的低质量细胞，也能聚在一起，因为相比于固有的生物差异，更相似的低质量让它们相依相偎。除此以外，本来非常小的细胞文库也能聚成一群，因为log转换后它们的平均表达量会发生很大的变化(A. Lun <a target="_blank" rel="noopener" href="https://osca.bioconductor.org/quality-control.html#ref-lun2018overcoming">2018</a>).</li>
<li><strong>方差估计或主成分分析</strong>:首先在PCA分析时，低质量和高质量之间的差异相比于生物学差异会体现更明显，占据主要的成分，减少降维结果的可靠性。另外，某个基因可能在两个细胞之间没什么表达差异，但是如果所处环境差异很大（一个细胞质量很低，另一个细胞质量正常），那么在差异估算过程中，就会把这个差异也会被当成差异表达基因。例如：一个低质量细胞文库的总表达量非常低（接近0），但恰巧还存在一个基因有表达量，那么这个基因的表达量在后续的文库归一化过程中就会尤为突出<br />
</li>
<li><strong>奇怪的转录本上调</strong>：实验难免会混入外源的污染转录本，但这个量很少并且在所有细胞中都应该是差不多水平的。但如果某个细胞质量低，文库小，那么在校正文库差异过程中，其中的污染转录本表达量就会“突飞猛进”，看起来是一些明显上调的“奇怪基因”。实际上，这些奇怪的基因依然在其他细胞中存在，并且真实的表达量差不多，并且是不应该占据主体地位的。</li>
</ul>
<p>为了最大程度避免上面一种或多种情况的发生，应该在归一化之前去掉这些低质量的细胞，这个过程就是<strong>细胞的质控 </strong></p>
<p>使用A. T. L. Lun et al. (2017)的小型scRNA数据（未进行QC）进行测试</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#--- loading ---#</span></span><br><span class="line"><span class="keyword">library</span>(scRNAseq)</span><br><span class="line">sce.416b &lt;- LunSpikeInData(which=<span class="string">&quot;416b&quot;</span>) </span><br><span class="line">table(sce.416b$block)</span><br><span class="line"><span class="number">20160113</span> <span class="number">20160325</span> </span><br><span class="line">      <span class="number">96</span>       <span class="number">96</span> </span><br><span class="line">sce.416b$block &lt;- factor(sce.416b$block)</span><br><span class="line">sce.416b</span><br></pre></td></tr></table></figure>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## class: SingleCellExperiment </span></span><br><span class="line"><span class="comment">## dim: 46604 192 </span></span><br><span class="line"><span class="comment">## metadata(0):</span></span><br><span class="line"><span class="comment">## assays(1): counts</span></span><br><span class="line"><span class="comment">## rownames(46604): ENSMUSG00000102693 ENSMUSG00000064842 ...</span></span><br><span class="line"><span class="comment">##   ENSMUSG00000095742 CBFB-MYH11-mcherry</span></span><br><span class="line"><span class="comment">## rowData names(1): Length</span></span><br><span class="line"><span class="comment">## colnames(192): SLX-9555.N701_S502.C89V9ANXX.s_1.r_1</span></span><br><span class="line"><span class="comment">##   SLX-9555.N701_S503.C89V9ANXX.s_1.r_1 ...</span></span><br><span class="line"><span class="comment">##   SLX-11312.N712_S508.H5H5YBBXX.s_8.r_1</span></span><br><span class="line"><span class="comment">##   SLX-11312.N712_S517.H5H5YBBXX.s_8.r_1</span></span><br><span class="line"><span class="comment">## colData names(9): Source Name cell line ... spike-in addition block</span></span><br><span class="line"><span class="comment">## reducedDimNames(0):</span></span><br><span class="line"><span class="comment">## altExpNames(2): ERCC SIRV</span></span><br></pre></td></tr></table></figure>
<h2 id="qc指标的选择">QC指标的选择</h2>
<p>鉴定细胞是否是低质量的，需要用到几个指标。虽然下面这些指标是使用SMART-seq2数据进行展示的，但依然适用于UMI数据（比如MARS-seq、droplet-based技术）</p>
<ul>
<li><strong>文库大小（library size）</strong> ：指的是每个细胞中所有基因的表达量之和。细胞的文库如果很小，说明在文库制备过程中存在RNA的损失，要么是由于细胞裂解，要么是cDNA捕获不足导致后续的扩增量少<br />
</li>
<li><strong>每个细胞中表达基因的数目（number of expressed features in each cell）</strong>：指的是细胞中非0表达量的基因数量。如果细胞中基本没有基因表达，很可能是转录本捕获失败<br />
</li>
<li><strong>比对到spike-in转录本的reads比例（proportion of reads mapped to spike-in transcripts）</strong>：计算方法是：<code>spike-in counts / all features (including spike-ins) for each cell</code>。每个细胞都应该加入等量的外源转录本（spike-in），如果哪个细胞的spike-in比例提高了，说明它的内源RNA含量减少（比如在细胞分选阶段出现的细胞裂解或者RNA降解）<br />
</li>
<li><strong>比对到线粒体基因组的reads比例（proportion of reads mapped to genes in the mitochondrial genome）</strong> ：如果没有spike-in，那么使用线粒体指标也是能说明问题的(Islam et al. 2014; Ilicic et al. 2016)。比对到线粒体的reads增多，说明细胞质中的RNA减少，可能存在细胞穿孔的情况，而这个孔的大小，可能只是将细胞质中存在的mRNA流出去，但线粒体的体积超过了孔的大小，所以还留在了细胞中，造成一定程度的富集，导致线粒体RNA占比升高。</li>
</ul>
<p>对于每个细胞，可以用scater包的<code>perCellQCMetrics()</code>函数进行计算，其中<code>sum</code>这一列表示每个细胞的文库大小；<code>detected</code>这一列表示检测到的基因数量；<code>subsets_Mito_percent</code>这一列表示比对到线粒体基因组的reads占比；<code>altexps_ERCC_percent</code>表示比对到ERCC spike-in的reads占比</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Retrieving the mitochondrial transcripts using genomic locations included in</span></span><br><span class="line"><span class="comment"># the row-level annotation for the SingleCellExperiment.</span></span><br><span class="line">location &lt;- rowRanges(sce.416b)</span><br><span class="line">is.mito &lt;- any(seqnames(location)==<span class="string">&quot;MT&quot;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># ALTERNATIVELY: using resources in AnnotationHub to retrieve chromosomal</span></span><br><span class="line"><span class="comment"># locations given the Ensembl IDs; this should yield the same result.</span></span><br><span class="line"><span class="keyword">library</span>(AnnotationHub)</span><br><span class="line">ens.mm.v97 &lt;- AnnotationHub()[[<span class="string">&quot;AH73905&quot;</span>]]</span><br><span class="line">chr.loc &lt;- mapIds(ens.mm.v97, keys=rownames(sce.416b),</span><br><span class="line">    keytype=<span class="string">&quot;GENEID&quot;</span>, column=<span class="string">&quot;SEQNAME&quot;</span>)</span><br><span class="line">is.mito.alt &lt;- which(chr.loc==<span class="string">&quot;MT&quot;</span>)</span><br><span class="line"></span><br><span class="line"><span class="keyword">library</span>(scater)</span><br><span class="line">df &lt;- perCellQCMetrics(sce.416b, subsets=list(Mito=is.mito))</span><br><span class="line">df</span><br></pre></td></tr></table></figure>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## DataFrame with 192 rows and 16 columns</span></span><br><span class="line"><span class="comment">##           sum  detected percent_top_50 percent_top_100 percent_top_200</span></span><br><span class="line"><span class="comment">##     &lt;integer&gt; &lt;integer&gt;      &lt;numeric&gt;       &lt;numeric&gt;       &lt;numeric&gt;</span></span><br><span class="line"><span class="comment">## 1      865936      7618        26.7218         32.2773         39.7208</span></span><br><span class="line"><span class="comment">## 2     1076277      7521        29.4043         35.0354         42.2581</span></span><br><span class="line"><span class="comment">## 3     1180138      8306        27.3454         32.4770         39.3296</span></span><br><span class="line"><span class="comment">## 4     1342593      8143        35.8092         40.2666         46.2460</span></span><br><span class="line"><span class="comment">## 5     1668311      7154        34.1198         39.0901         45.6660</span></span><br><span class="line"><span class="comment">## ...       ...       ...            ...             ...             ...</span></span><br><span class="line"><span class="comment">## 188    776622      8174        45.9362         49.7010         54.6101</span></span><br><span class="line"><span class="comment">## 189   1299950      8956        38.0829         42.8930         49.0622</span></span><br><span class="line"><span class="comment">## 190   1800696      9530        30.6675         35.5839         41.8550</span></span><br><span class="line"><span class="comment">## 191     46731      6649        32.2998         37.9149         44.5999</span></span><br><span class="line"><span class="comment">## 192   1866692     10964        26.6632         31.2584         37.5608</span></span><br><span class="line"><span class="comment">##     percent_top_500 subsets_Mito_sum subsets_Mito_detected subsets_Mito_percent</span></span><br><span class="line"><span class="comment">##           &lt;numeric&gt;        &lt;integer&gt;             &lt;integer&gt;            &lt;numeric&gt;</span></span><br><span class="line"><span class="comment">## 1           52.9038            78790                    20              9.09882</span></span><br><span class="line"><span class="comment">## 2           55.7454            98613                    20              9.16242</span></span><br><span class="line"><span class="comment">## 3           51.9337           100341                    19              8.50248</span></span><br><span class="line"><span class="comment">## 4           57.1210           104882                    20              7.81190</span></span><br><span class="line"><span class="comment">## 5           58.2004           129559                    22              7.76588</span></span><br><span class="line"><span class="comment">## ...             ...              ...                   ...                  ...</span></span><br><span class="line"><span class="comment">## 188         64.4249            48126                    20              6.19684</span></span><br><span class="line"><span class="comment">## 189         60.6675           112225                    25              8.63302</span></span><br><span class="line"><span class="comment">## 190         53.6781           135693                    23              7.53559</span></span><br><span class="line"><span class="comment">## 191         56.5235             3505                    16              7.50037</span></span><br><span class="line"><span class="comment">## 192         48.9489           150375                    29              8.05569</span></span><br><span class="line"><span class="comment">##     altexps_ERCC_sum altexps_ERCC_detected altexps_ERCC_percent</span></span><br><span class="line"><span class="comment">##            &lt;integer&gt;             &lt;integer&gt;            &lt;numeric&gt;</span></span><br><span class="line"><span class="comment">## 1              65278                    39              6.80658</span></span><br><span class="line"><span class="comment">## 2              74748                    40              6.28030</span></span><br><span class="line"><span class="comment">## 3              60878                    42              4.78949</span></span><br><span class="line"><span class="comment">## 4              60073                    42              4.18567</span></span><br><span class="line"><span class="comment">## 5             136810                    44              7.28887</span></span><br><span class="line"><span class="comment">## ...              ...                   ...                  ...</span></span><br><span class="line"><span class="comment">## 188            61575                    39              7.17620</span></span><br><span class="line"><span class="comment">## 189            94982                    41              6.65764</span></span><br><span class="line"><span class="comment">## 190           113707                    40              5.81467</span></span><br><span class="line"><span class="comment">## 191             7580                    44             13.48898</span></span><br><span class="line"><span class="comment">## 192            48664                    39              2.51930</span></span><br><span class="line"><span class="comment">##     altexps_SIRV_sum altexps_SIRV_detected altexps_SIRV_percent     total</span></span><br><span class="line"><span class="comment">##            &lt;integer&gt;             &lt;integer&gt;            &lt;numeric&gt; &lt;integer&gt;</span></span><br><span class="line"><span class="comment">## 1              27828                     7              2.90165    959042</span></span><br><span class="line"><span class="comment">## 2              39173                     7              3.29130   1190198</span></span><br><span class="line"><span class="comment">## 3              30058                     7              2.36477   1271074</span></span><br><span class="line"><span class="comment">## 4              32542                     7              2.26741   1435208</span></span><br><span class="line"><span class="comment">## 5              71850                     7              3.82798   1876971</span></span><br><span class="line"><span class="comment">## ...              ...                   ...                  ...       ...</span></span><br><span class="line"><span class="comment">## 188            19848                     7             2.313165    858045</span></span><br><span class="line"><span class="comment">## 189            31729                     7             2.224004   1426661</span></span><br><span class="line"><span class="comment">## 190            41116                     7             2.102562   1955519</span></span><br><span class="line"><span class="comment">## 191             1883                     7             3.350892     56194</span></span><br><span class="line"><span class="comment">## 192            16289                     7             0.843271   1931645</span></span><br></pre></td></tr></table></figure>
<p>另外，还可以使用<code>addPerCellQC()</code>，它会把每个细胞的QC指标加到<code>SingleCellExperiment</code>对象的<code>colData</code>中，方便后面调取</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">sce.416b &lt;- addPerCellQC(sce.416b, subsets=list(Mito=is.mito))</span><br><span class="line">colnames(colData(sce.416b))</span><br></pre></td></tr></table></figure>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">##  [1] &quot;Source Name&quot;              &quot;cell line&quot;               </span></span><br><span class="line"><span class="comment">##  [3] &quot;cell type&quot;                &quot;single cell well quality&quot;</span></span><br><span class="line"><span class="comment">##  [5] &quot;genotype&quot;                 &quot;phenotype&quot;               </span></span><br><span class="line"><span class="comment">##  [7] &quot;strain&quot;                   &quot;spike-in addition&quot;       </span></span><br><span class="line"><span class="comment">##  [9] &quot;block&quot;                    &quot;sum&quot;                     </span></span><br><span class="line"><span class="comment">## [11] &quot;detected&quot;                 &quot;percent_top_50&quot;          </span></span><br><span class="line"><span class="comment">## [13] &quot;percent_top_100&quot;          &quot;percent_top_200&quot;         </span></span><br><span class="line"><span class="comment">## [15] &quot;percent_top_500&quot;          &quot;subsets_Mito_sum&quot;        </span></span><br><span class="line"><span class="comment">## [17] &quot;subsets_Mito_detected&quot;    &quot;subsets_Mito_percent&quot;    </span></span><br><span class="line"><span class="comment">## [19] &quot;altexps_ERCC_sum&quot;         &quot;altexps_ERCC_detected&quot;   </span></span><br><span class="line"><span class="comment">## [21] &quot;altexps_ERCC_percent&quot;     &quot;altexps_SIRV_sum&quot;        </span></span><br><span class="line"><span class="comment">## [23] &quot;altexps_SIRV_detected&quot;    &quot;altexps_SIRV_percent&quot;    </span></span><br><span class="line"><span class="comment">## [25] &quot;total&quot;</span></span><br></pre></td></tr></table></figure>
<p>当然，这里做QC统计的前提假设是：每个细胞的qc指标都是独立于生物学状态的。也就是说，qc指标不会那么智能地识别细胞类型然后进行判断。它会认为（如文库太小、高线粒体占比）都是由技术误差引起的，而非生物因素。但是有一个问题，<strong>如果某些细胞类型本身的RNA含量就很低，或者它们本来就含有很多的线粒体转录本呢？再根据这个指标进行过滤，会不会损失一些细胞类型呢？</strong></p>
<h2 id="识别低质量细胞">识别低质量细胞</h2>
<h3 id="使用固定的阈值">1. 使用固定的阈值</h3>
<p>识别低质量细胞最简单方法是在QC度量上应用阈值。例如设定文库低于10万reads的细胞是低质量的，或者表达基因数量少于5000个，spike-in或线粒体占比高于10%。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">qc.lib &lt;- df$sum &lt; <span class="number">1e5</span></span><br><span class="line">qc.nexprs &lt;- df$detected &lt; <span class="number">5e3</span></span><br><span class="line">qc.spike &lt;- df$altexps_ERCC_percent &gt; <span class="number">10</span></span><br><span class="line">qc.mito &lt;- df$subsets_Mito_percent &gt; <span class="number">10</span></span><br><span class="line">discard &lt;- qc.lib | qc.nexprs | qc.spike | qc.mito</span><br><span class="line"></span><br><span class="line"><span class="comment"># Summarize the number of cells removed for each reason.</span></span><br><span class="line">DataFrame(LibSize=sum(qc.lib), NExprs=sum(qc.nexprs),</span><br><span class="line">    SpikeProp=sum(qc.spike), MitoProp=sum(qc.mito), Total=sum(discard))</span><br></pre></td></tr></table></figure>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## DataFrame with 1 row and 5 columns</span></span><br><span class="line"><span class="comment">##     LibSize    NExprs SpikeProp  MitoProp     Total</span></span><br><span class="line"><span class="comment">##   &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></span><br><span class="line"><span class="comment">## 1         3         0        19        14        33</span></span><br></pre></td></tr></table></figure>
<p>虽然看起来简单，但使用这种方法需要丰富的经验，了解实验设计和细胞状态；另外即使使用同一种文库制备方法，但由于细胞捕获效率和测序深度的不同，这个阈值依然需要适时调整。因此对于研究人员要求很高。</p>
<h3 id="使用相对阈值">2. 使用相对阈值</h3>
<h4 id="鉴定离群点">鉴定离群点</h4>
<p>为了获得相对阈值，先假设大部分细胞都是高质量的，然后去找离群点作为低质量。那么按照什么方法找离群点呢？常用的一个函数<code>isOutlier</code>使用的是MAD指标（绝对中位差来估计方差,先计算出数据与它们的中位数之间的偏差，然后这些偏差的绝对值的中位数就是MAD，median absolute deviation）。如果<strong>超过中位数3倍MAD的值就是离群值</strong>。</p>
<p>使用<code>isOutlier</code>时，如果要相减（例如：<code>df$sum - 3* MAD</code>），就用<code>type="lower"</code>，此时一般还要做个log转换<code>log=TRUE</code> ，保证得到的结果不是负数</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">qc.lib2 &lt;- isOutlier(df$sum, log=<span class="literal">TRUE</span>, type=<span class="string">&quot;lower&quot;</span>)</span><br><span class="line">qc.nexprs2 &lt;- isOutlier(df$detected, log=<span class="literal">TRUE</span>, type=<span class="string">&quot;lower&quot;</span>)</span><br><span class="line">qc.spike2 &lt;- isOutlier(df$altexps_ERCC_percent, type=<span class="string">&quot;higher&quot;</span>)</span><br><span class="line">qc.mito2 &lt;- isOutlier(df$subsets_Mito_percent, type=<span class="string">&quot;higher&quot;</span>)</span><br></pre></td></tr></table></figure>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">attr(qc.lib2, <span class="string">&quot;thresholds&quot;</span>)</span><br><span class="line">   lower   higher </span><br><span class="line"><span class="number">434082.9</span>      <span class="literal">Inf</span> </span><br><span class="line">attr(qc.nexprs2, <span class="string">&quot;thresholds&quot;</span>)</span><br><span class="line">   lower   higher </span><br><span class="line"><span class="number">5231.468</span>      <span class="literal">Inf</span> </span><br><span class="line">attr(qc.spike2, <span class="string">&quot;thresholds&quot;</span>)</span><br><span class="line">   lower   higher </span><br><span class="line">    -<span class="literal">Inf</span> <span class="number">14.15371</span> </span><br><span class="line">attr(qc.mito2, <span class="string">&quot;thresholds&quot;</span>)</span><br><span class="line">   lower   higher </span><br><span class="line">    -<span class="literal">Inf</span> <span class="number">11.91734</span> </span><br></pre></td></tr></table></figure>
<p>用相对阈值过滤的细胞数量统计：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">discard2 &lt;- qc.lib2 | qc.nexprs2 | qc.spike2 | qc.mito2</span><br><span class="line"></span><br><span class="line"><span class="comment"># Summarize the number of cells removed for each reason.</span></span><br><span class="line">DataFrame(LibSize=sum(qc.lib2), NExprs=sum(qc.nexprs2),</span><br><span class="line">    SpikeProp=sum(qc.spike2), MitoProp=sum(qc.mito2), Total=sum(discard2))</span><br></pre></td></tr></table></figure>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## DataFrame with 1 row and 5 columns</span></span><br><span class="line"><span class="comment">##     LibSize    NExprs SpikeProp  MitoProp     Total</span></span><br><span class="line"><span class="comment">##   &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;</span></span><br><span class="line"><span class="comment">## 1         4         0         1         2         6</span></span><br></pre></td></tr></table></figure>
<p><strong>除此以外，还有一种更快的计算方法，一步整合了上面的操作：</strong></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">reasons &lt;- quickPerCellQC(df, percent_subsets=c(<span class="string">&quot;subsets_Mito_percent&quot;</span>,</span><br><span class="line">    <span class="string">&quot;altexps_ERCC_percent&quot;</span>))</span><br><span class="line">colSums(as.matrix(reasons))</span><br></pre></td></tr></table></figure>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">##              low_lib_size            low_n_features high_subsets_Mito_percent </span></span><br><span class="line"><span class="comment">##                         4                         0                         2 </span></span><br><span class="line"><span class="comment">## high_altexps_ERCC_percent                   discard </span></span><br><span class="line"><span class="comment">##                         1                         6</span></span><br></pre></td></tr></table></figure>
<p>使用”相对“的阈值一个好处就是可以根据测序深度、cDNA捕获效率、线粒体含量等等进行阈值的调整，在经验不是足够丰富的时候，可以辅助判断。但仍需要注意的是，<strong>使用相对阈值是有前提的，那就是：认为大部分细胞都是高质量的</strong></p>
<h3 id="离群点检测的假设">离群点检测的假设</h3>
<ul>
<li>离群点的检测的假设是<strong>大部分细胞的质量都不错</strong>。这一点假设可以通过实验去验证（比如肉眼检查细胞完整性）。<strong>但当大部分细胞质量都很低，使用相对阈值结果就对大量的低质量细胞无计可施</strong>。因为它使用了MAD值，和中位数有关系，那么可以试想：如果一堆数都不合格，那么它们的中位数也不合格，因此原来正确的值，其实在这群不合格的数值中，就是“离群”的。<br />
</li>
<li>另一个假设就是：<strong>每个细胞的qc指标都是独立于生物学状态的</strong>。也就是说，qc指标不会那么智能地识别细胞类型然后进行判断。在异质性很高的组织中， 就是存在内源RNA含量低，而线粒体基因占比高的细胞。如果使用”一刀切“的固定阈值，它们就很可能会被当成离群点被过滤。而是用MAD计算方法检测的结果可能就是：虽然一堆细胞的某个qc指标差异很大，但中位数也在变，随之变化的还有MAD值，这样最后的结果不至于和真实生物学情况差太多</li>
</ul>
<h3 id="考虑实验的因素">考虑实验的因素</h3>
<p>很多大型的实验都包含多个细胞系，而且可能采用的实验方法不同（比如选用不同的测序深度），这就产生了实验的不同批次。<strong>这种情况下， 应该对每个批次分别进行检测。</strong></p>
<p>如果每个批次都是一个<code>SingleCellExperiment</code>对象，那么<code>isOutlier()</code>函数可以按上面的方法直接使用；但是如果不同批次的细胞已经混合成一整个<code>SingleCellExperiment</code>对象，那么<code>batch=</code>这个参数就派上用场了。</p>
<p>同样以这个416B数据集为例，他包含了两种不同的实验类型。然后我们就可以使用<code>batch=</code>参数去进行质控。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">batch &lt;- paste0(sce.416b$phenotype, <span class="string">&quot;-&quot;</span>, sce.416b$Plate)</span><br><span class="line">batch.reasons &lt;- quickPerCellQC(df, percent_subsets=c(<span class="string">&quot;subsets_Mito_percent&quot;</span>,</span><br><span class="line">    <span class="string">&quot;altexps_ERCC_percent&quot;</span>), batch=batch)</span><br><span class="line">colSums(as.matrix(batch.reasons))</span><br></pre></td></tr></table></figure>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">##              low_lib_size            low_n_features high_subsets_Mito_percent </span></span><br><span class="line"><span class="comment">##                         4                         2                         2 </span></span><br><span class="line"><span class="comment">## high_altexps_ERCC_percent                   discard </span></span><br><span class="line"><span class="comment">##                         4                         7</span></span><br></pre></td></tr></table></figure>
<p><strong>但是，<code>batch</code>参数不是万能的</strong>，之前也说过，这种相对阈值需要一个<strong>假设：每个批次的大部分细胞都是高质量的</strong>。当某个批次的细胞整体质量偏低，离群点检测很有可能失败</p>
<p>例如，在Grun et al. (2016)的数据集中有两个donor的实验是失败的。它们的ERCC含量相对其他批次高，增加了中位数和MAD值，导致过滤低质量细胞失败。因此这种情况下，可以先算其他几个批次的中位数和MAD值，然后参考这些值去对有问题的批次进行过滤。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">library</span>(scRNAseq)</span><br><span class="line">sce.grun &lt;- GrunPancreasData()</span><br><span class="line">sce.grun &lt;- addPerCellQC(sce.grun)</span><br><span class="line"></span><br><span class="line"><span class="comment"># First attempt with batch-specific thresholds.</span></span><br><span class="line">discard.ercc &lt;- isOutlier(sce.grun$altexps_ERCC_percent,</span><br><span class="line">    type=<span class="string">&quot;higher&quot;</span>, batch=sce.grun$donor)</span><br><span class="line">with.blocking &lt;- plotColData(sce.grun, x=<span class="string">&quot;donor&quot;</span>, y=<span class="string">&quot;altexps_ERCC_percent&quot;</span>,</span><br><span class="line">    colour_by=I(discard.ercc))</span><br><span class="line"></span><br><span class="line"><span class="comment"># Second attempt, sharing information across batches</span></span><br><span class="line"><span class="comment"># to avoid dramatically different thresholds for unusual batches.</span></span><br><span class="line">discard.ercc2 &lt;- isOutlier(sce.grun$altexps_ERCC_percent,</span><br><span class="line">    type=<span class="string">&quot;higher&quot;</span>, batch=sce.grun$donor,</span><br><span class="line">    subset=sce.grun$donor %<span class="keyword">in</span>% c(<span class="string">&quot;D17&quot;</span>, <span class="string">&quot;D2&quot;</span>, <span class="string">&quot;D7&quot;</span>))</span><br><span class="line">without.blocking &lt;- plotColData(sce.grun, x=<span class="string">&quot;donor&quot;</span>, y=<span class="string">&quot;altexps_ERCC_percent&quot;</span>,</span><br><span class="line">    colour_by=I(discard.ercc2))</span><br><span class="line"></span><br><span class="line">gridExtra::grid.arrange(with.blocking, without.blocking, ncol=<span class="number">2</span>) </span><br></pre></td></tr></table></figure>
<p><img src="/images/quality-control/qc-plot-pancreas-1.png" /></p>
<blockquote>
<p>注：可以看到，左图是按照每个批次分别鉴定的离群点；右图是用质量好的批次计算的阈值，然后运用到差的批次上的结果</p>
</blockquote>
<p>为了鉴别有问题的批次，可以先将每个批次分别计算结果，然后比较它们的阈值，如果比同类批次超出太多，就要警觉。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">ercc.thresholds &lt;- attr(discard.ercc, <span class="string">&quot;thresholds&quot;</span>)[<span class="string">&quot;higher&quot;</span>,]</span><br><span class="line">ercc.thresholds</span><br><span class="line"><span class="comment">##        D10        D17         D2         D3         D7 </span></span><br><span class="line"><span class="comment">##  73.610696   7.599947   6.010975 113.105828  15.216956</span></span><br><span class="line"></span><br><span class="line">names(ercc.thresholds)[isOutlier(ercc.thresholds, type=<span class="string">&quot;higher&quot;</span>)]</span><br><span class="line"><span class="comment">## [1] &quot;D10&quot; &quot;D3&quot;</span></span><br></pre></td></tr></table></figure>
<blockquote>
<p>可以看到D10、D3的阈值就超过其他批次很多</p>
</blockquote>
<p>但是这么做的前提都是：我们认为批次中的细胞质量整体还不错。如果我们保证不了细胞质量，那么这种计算相对阈值的方法就不成立，还是要使用绝对阈值，手动去除。</p>
<h3 id="其他方法">3. 其他方法</h3>
<p>另一个策略是根据每个细胞的QC指标来在高维空间中识别异常值。利用<code>robustbase</code> 包，将细胞放在高维空间，根据他们的QC指标（文库大小、表达基因数、spike-in比例、线粒体比例），然后使用<code>isOutlier()</code>来识别表现出异常高outlylier水平的低质量细胞</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">stats &lt;- cbind(log10(df$sum), log10(df$detected),</span><br><span class="line">    df$subsets_Mito_percent, df$altexps_ERCC_percent)</span><br><span class="line"></span><br><span class="line"><span class="keyword">library</span>(robustbase)</span><br><span class="line">outlying &lt;- adjOutlyingness(stats, only.outlyingness = <span class="literal">TRUE</span>)</span><br><span class="line">multi.outlier &lt;- isOutlier(outlying, type = <span class="string">&quot;higher&quot;</span>)</span><br><span class="line">summary(multi.outlier)</span><br><span class="line"><span class="comment">##    Mode   FALSE    TRUE </span></span><br><span class="line"><span class="comment">## logical     182      10</span></span><br></pre></td></tr></table></figure>
<p>除此以外，有时<strong>还可以根据基因表达量进行过滤</strong>，不过在bulk转录组中用的比较多，但是在scRNA中这样操作可能会损失一些本身数量就比较少的高质量细胞群体（比如一些罕见细胞，本身基因表达量就不是很高）</p>
<h2 id="画图检查">画图检查</h2>
<p>检查QC度量的分布以确定可能的问题是一个很好的实践。在最理想的情况下，我们会看到正态分布，可以证明在离群值检测中使用的<code>3 MAD</code>阈值是合理的。另一种模式下的大量细胞表明QC指标可能与某些生物状态相关，可能导致过滤过程中不同细胞类型的丢失;或者有不一致的库准备为一个子集的细胞，一个非罕见的现象在板的协议。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 把QC指标和原来的sce.416b细胞信息整合起来</span></span><br><span class="line">colData(sce.416b) &lt;- cbind(colData(sce.416b), df)</span><br><span class="line"><span class="comment"># 修改一下整合后的信息</span></span><br><span class="line">sce.416b$block &lt;- factor(sce.416b$block)</span><br><span class="line">sce.416b$phenotype &lt;- ifelse(grepl(<span class="string">&quot;induced&quot;</span>, sce.416b$phenotype),</span><br><span class="line">    <span class="string">&quot;induced&quot;</span>, <span class="string">&quot;wild type&quot;</span>)</span><br><span class="line">sce.416b$discard &lt;- reasons$discard</span><br><span class="line"><span class="comment"># 绘图</span></span><br><span class="line">gridExtra::grid.arrange(</span><br><span class="line">    plotColData(sce.416b, x=<span class="string">&quot;block&quot;</span>, y=<span class="string">&quot;sum&quot;</span>, colour_by=<span class="string">&quot;discard&quot;</span>,</span><br><span class="line">        other_fields=<span class="string">&quot;phenotype&quot;</span>) + facet_wrap(~phenotype) + </span><br><span class="line">        scale_y_log10() + ggtitle(<span class="string">&quot;Total count&quot;</span>),</span><br><span class="line">    plotColData(sce.416b, x=<span class="string">&quot;block&quot;</span>, y=<span class="string">&quot;detected&quot;</span>, colour_by=<span class="string">&quot;discard&quot;</span>, </span><br><span class="line">        other_fields=<span class="string">&quot;phenotype&quot;</span>) + facet_wrap(~phenotype) + </span><br><span class="line">        scale_y_log10() + ggtitle(<span class="string">&quot;Detected features&quot;</span>),</span><br><span class="line">    plotColData(sce.416b, x=<span class="string">&quot;block&quot;</span>, y=<span class="string">&quot;subsets_Mito_percent&quot;</span>, </span><br><span class="line">        colour_by=<span class="string">&quot;discard&quot;</span>, other_fields=<span class="string">&quot;phenotype&quot;</span>) + </span><br><span class="line">        facet_wrap(~phenotype) + ggtitle(<span class="string">&quot;Mito percent&quot;</span>),</span><br><span class="line">    plotColData(sce.416b, x=<span class="string">&quot;block&quot;</span>, y=<span class="string">&quot;altexps_ERCC_percent&quot;</span>, </span><br><span class="line">        colour_by=<span class="string">&quot;discard&quot;</span>, other_fields=<span class="string">&quot;phenotype&quot;</span>) + </span><br><span class="line">        facet_wrap(~phenotype) + ggtitle(<span class="string">&quot;ERCC percent&quot;</span>),</span><br><span class="line">    ncol=<span class="number">1</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>
<p><img src="/images/quality-control/qc-dist-416b-1.png" /></p>
<blockquote>
<p>展示的是不同批次的QC指标</p>
</blockquote>
<p>另一种有用的诊断方法是绘制相对于其他QC指标的线粒体计数比例图。</p>
<p>为了确保不存在这样的细胞：虽然细胞文库大，但它的线粒体占比也高。另外也是为了避免意外过滤掉本来是高质量但同时具有高代谢活性（即高线粒体占比）的细胞（如肝脏细胞）</p>
<p>======未完</p>
<h2 id="针对droplet数据的细胞判断">针对Droplet数据的细胞判断</h2>
<h3 id="背景">背景</h3>
<p>由于这种建库方法的独特性，我们没办法事先知道某一个细胞文库（比如一个cell barcode）是真正包含一个细胞还是只是一个空的液滴（droplet）。因此，<strong>第一步是需要根据得到的表达量信息，来推断液滴是不是空的</strong>。但仅仅根据表达量判断还是不太靠谱，因为还有可能在空的液滴中依然包含外源RNA，最后的细胞文库依旧不为0，但确实不包含细胞。</p>
<p>这里为了说明这个问题，<strong>使用了一个未过滤的10X PBMC数据</strong></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 数据下载</span></span><br><span class="line"><span class="keyword">library</span>(BiocFileCache)</span><br><span class="line">bfc &lt;- BiocFileCache(<span class="string">&quot;raw_data&quot;</span>, ask = <span class="literal">FALSE</span>)</span><br><span class="line">raw.path &lt;- bfcrpath(bfc, file.path(<span class="string">&quot;http://cf.10xgenomics.com/samples&quot;</span>,</span><br><span class="line">    <span class="string">&quot;cell-exp/2.1.0/pbmc4k/pbmc4k_raw_gene_bc_matrices.tar.gz&quot;</span>))</span><br><span class="line"><span class="comment"># 解压数据</span></span><br><span class="line">untar(raw.path, exdir=file.path(tempdir(), <span class="string">&quot;pbmc4k&quot;</span>))</span><br><span class="line"></span><br><span class="line"><span class="keyword">library</span>(DropletUtils)</span><br><span class="line"><span class="keyword">library</span>(Matrix)</span><br><span class="line">fname &lt;- file.path(tempdir(), <span class="string">&quot;pbmc4k/raw_gene_bc_matrices/GRCh38&quot;</span>)</span><br><span class="line">sce.pbmc &lt;- read10xCounts(fname, col.names=<span class="literal">TRUE</span>)</span><br><span class="line">sce.pbmc</span><br></pre></td></tr></table></figure>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## class: SingleCellExperiment </span></span><br><span class="line"><span class="comment">## dim: 33694 737280 </span></span><br><span class="line"><span class="comment">## metadata(1): Samples</span></span><br><span class="line"><span class="comment">## assays(1): counts</span></span><br><span class="line"><span class="comment">## rownames(33694): ENSG00000243485 ENSG00000237613 ... ENSG00000277475</span></span><br><span class="line"><span class="comment">##   ENSG00000268674</span></span><br><span class="line"><span class="comment">## rowData names(2): ID Symbol</span></span><br><span class="line"><span class="comment">## colnames(737280): AAACCTGAGAAACCAT-1 AAACCTGAGAAACCGC-1 ...</span></span><br><span class="line"><span class="comment">##   TTTGTCATCTTTAGTC-1 TTTGTCATCTTTCCTC-1</span></span><br><span class="line"><span class="comment">## colData names(2): Sample Barcode</span></span><br><span class="line"><span class="comment">## reducedDimNames(0):</span></span><br><span class="line"><span class="comment">## altExpNames(0):</span></span><br></pre></td></tr></table></figure>
<p><strong>整体观察不同的barcodes（不一定都是真的细胞）的文库大小分布：</strong></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">bcrank &lt;- barcodeRanks(counts(sce.pbmc))</span><br><span class="line"></span><br><span class="line"><span class="comment"># Only showing unique points for plotting speed.</span></span><br><span class="line">uniq &lt;- !duplicated(bcrank$rank)</span><br><span class="line">plot(bcrank$rank[uniq], bcrank$total[uniq], log=<span class="string">&quot;xy&quot;</span>,</span><br><span class="line">    xlab=<span class="string">&quot;Rank&quot;</span>, ylab=<span class="string">&quot;Total UMI count&quot;</span>, cex.lab=<span class="number">1.2</span>)</span><br><span class="line"></span><br><span class="line">abline(h=metadata(bcrank)$inflection, col=<span class="string">&quot;darkgreen&quot;</span>, lty=<span class="number">2</span>)</span><br><span class="line">abline(h=metadata(bcrank)$knee, col=<span class="string">&quot;dodgerblue&quot;</span>, lty=<span class="number">2</span>)</span><br><span class="line"></span><br><span class="line">legend(<span class="string">&quot;bottomleft&quot;</span>, legend=c(<span class="string">&quot;Inflection&quot;</span>, <span class="string">&quot;Knee&quot;</span>), </span><br><span class="line">        col=c(<span class="string">&quot;darkgreen&quot;</span>, <span class="string">&quot;dodgerblue&quot;</span>), lty=<span class="number">2</span>, cex=<span class="number">1.2</span>)</span><br></pre></td></tr></table></figure>
<p><img src="/images/quality-control/rankplot-1.png" /></p>
<blockquote>
<p>看到各个barcodes的文库大小有高有低，并且相差较大，因此可能对应着真实存在的细胞和空液滴。当然最简单的办法还是”一刀切“，<strong>留下那些文库较大的细胞</strong>，不过还是有损失真实细胞类型的风险</p>
</blockquote>
<h3 id="检测空的液滴">检测空的液滴</h3>
<p>我们使用<code>emptyDrops()</code>函数，检查每个barcode的表达量是不是和外源RNA的表达量有显著差异(Lun et al. 2019)。如果存在显著差异，就说明barcode中更有可能是一个细胞。这种方法可<strong>以帮助区分：测序质量好的空液滴 和 包含细胞但RNA含量较少的液滴</strong>。尽管它们总体的表达量可能很相似，但本质不同，还是要区分的。</p>
<p><code>emptyDrops()</code> 使用Monte Carlo统计模拟计算p值，如果要重复结果，需要设置随机种子。另外设置 false discovery rate (FDR)为0.1%，意味着不超过0.1%的barcodes是空的。</p>
      </div>
      
        <div class="prev-or-next">
          <div class="post-foot-next">
            
              <a href="/2020/08/10/one-tail/" target="_self">
                <i class="iconfont icon-chevronleft"></i>
                <span>上一页</span>
              </a>
            
          </div>
          <div class="post-attach">
            <span class="post-pubtime">
              <i class="iconfont icon-updatetime" title="更新时间"></i>
              2020-08-12
            </span>
            
                  <span class="post-tags">
                    <i class="iconfont icon-tags" title="标签"></i>
                    
                    <span class="span--tag">
                      <a href="/tags/SingleCell/" title="SingleCell">
                        <b>#</b> SingleCell
                      </a>
                    </span>
                    
                  </span>
              
          </div>
          <div class="post-foot-prev">
            
              <a href="/2020/08/12/single-cell-i/" target="_self">
                <span>下一页</span>
                <i class="iconfont icon-chevronright"></i>
              </a>
            
          </div>
        </div>
      
    </div>
    
  <div class="post-catalog" id="catalog">
    <div class="title">目录</div>
    <div class="catalog-content">
      <ol class="toc"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%8A%A8%E6%9C%BA"><span class="toc-text">动机</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#qc%E6%8C%87%E6%A0%87%E7%9A%84%E9%80%89%E6%8B%A9"><span class="toc-text">QC指标的选择</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E8%AF%86%E5%88%AB%E4%BD%8E%E8%B4%A8%E9%87%8F%E7%BB%86%E8%83%9E"><span class="toc-text">识别低质量细胞</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#%E4%BD%BF%E7%94%A8%E5%9B%BA%E5%AE%9A%E7%9A%84%E9%98%88%E5%80%BC"><span class="toc-text">1. 使用固定的阈值</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E4%BD%BF%E7%94%A8%E7%9B%B8%E5%AF%B9%E9%98%88%E5%80%BC"><span class="toc-text">2. 使用相对阈值</span></a><ol class="toc-child"><li class="toc-item toc-level-4"><a class="toc-link" href="#%E9%89%B4%E5%AE%9A%E7%A6%BB%E7%BE%A4%E7%82%B9"><span class="toc-text">鉴定离群点</span></a></li></ol></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E7%A6%BB%E7%BE%A4%E7%82%B9%E6%A3%80%E6%B5%8B%E7%9A%84%E5%81%87%E8%AE%BE"><span class="toc-text">离群点检测的假设</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E8%80%83%E8%99%91%E5%AE%9E%E9%AA%8C%E7%9A%84%E5%9B%A0%E7%B4%A0"><span class="toc-text">考虑实验的因素</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E5%85%B6%E4%BB%96%E6%96%B9%E6%B3%95"><span class="toc-text">3. 其他方法</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E7%94%BB%E5%9B%BE%E6%A3%80%E6%9F%A5"><span class="toc-text">画图检查</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E9%92%88%E5%AF%B9droplet%E6%95%B0%E6%8D%AE%E7%9A%84%E7%BB%86%E8%83%9E%E5%88%A4%E6%96%AD"><span class="toc-text">针对Droplet数据的细胞判断</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#%E8%83%8C%E6%99%AF"><span class="toc-text">背景</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E6%A3%80%E6%B5%8B%E7%A9%BA%E7%9A%84%E6%B6%B2%E6%BB%B4"><span class="toc-text">检测空的液滴</span></a></li></ol></li></ol>
    </div>
  </div>

  
<script src="/js/catalog.js"></script>




    
      <div class="comments-container">
        


  <script src='//unpkg.com/valine/dist/Valine.min.js'></script>

  <div id="vcomments"></div>

  <script>
    new Valine({
      el: '#vcomments',
      appId: 'DdMocNMQxnEQcChwW38b0JVQ-gzGzoHsz',
      appKey: 'ETFXPtdWGEvlSfsH7RwSMzAf',
      placeholder: 'Welcome!',
      avatar: 'retro',
      lang: 'zh-CN'
    })
  </script>

    <style>
      .comments-container .v .vempty {
        display: none!important;
      }
    </style>




      </div>
    
  </div>


        <div class="footer">
  <div class="social">
    <ul>
      
        <li>
          <a title="github" target="_blank" rel="noopener" href="https://github.com/Moonerss">
            <i class="iconfont icon-github"></i>
          </a>
        </li>
      
        <li>
          <a title="rss" href="/atom.xml">
            <i class="iconfont icon-rss"></i>
          </a>
        </li>
      
    </ul>
  </div>
  
    <div class="footer-more">
      <a target="_blank" rel="noopener" href="https://github.com/Moonerss">Copyright © mooner 2020</a>
    </div>
  
    <div class="footer-more">
      <a target="_blank" rel="noopener" href="https://github.com/zchengsite/hexo-theme-oranges">Theme by Oranges | Powered by Hexo</a>
    </div>
  
</div>

      </div>

      <div class="back-to-top hidden">
  <a href="javascript: void(0)">
    <i class="iconfont icon-chevronup"></i>
  </a>
</div>


<script src="/js/backtotop.js"></script>



      
  <div class="search-icon" id="search-icon">
    <a href="javascript: void(0)">
      <i class="iconfont icon-search"></i>
    </a>
  </div>

  <div class="search-overlay hidden">
    <div class="search-content" tabindex="0">
      <div class="search-title">
        <span class="search-icon-input">
          <a href="javascript: void(0)">
            <i class="iconfont icon-search"></i>
          </a>
        </span>
        
          <input type="text" class="search-input" id="search-input" placeholder="Search ...">
        
        <span class="search-close-icon" id="search-close-icon">
          <a href="javascript: void(0)">
            <i class="iconfont icon-close"></i>
          </a>
        </span>
      </div>
      <div class="search-result" id="search-result"></div>
    </div>
  </div>

  <script type="text/javascript">
    var inputArea = document.querySelector("#search-input")
    var searchOverlayArea = document.querySelector(".search-overlay")

    inputArea.onclick = function() {
      getSearchFile()
      this.onclick = null
    }

    inputArea.onkeydown = function() {
      if(event.keyCode == 13)
        return false
    }

    function openOrHideSearchContent() {
      let isHidden = searchOverlayArea.classList.contains('hidden')
      if (isHidden) {
        searchOverlayArea.classList.remove('hidden')
        document.body.classList.add('hidden')
        // inputArea.focus()
      } else {
        searchOverlayArea.classList.add('hidden')
        document.body.classList.remove('hidden')
      }
    }

    function blurSearchContent(e) {
      if (e.target === searchOverlayArea) {
        openOrHideSearchContent()
      }
    }

    document.querySelector("#search-icon").addEventListener("click", openOrHideSearchContent, false)
    document.querySelector("#search-close-icon").addEventListener("click", openOrHideSearchContent, false)
    searchOverlayArea.addEventListener("click", blurSearchContent, false)

    var searchFunc = function (path, search_id, content_id) {
      'use strict';
      var $input = document.getElementById(search_id);
      var $resultContent = document.getElementById(content_id);
      $resultContent.innerHTML = "<ul><span class='local-search-empty'>首次搜索，正在载入索引文件，请稍后……<span></ul>";
      $.ajax({
        // 0x01. load xml file
        url: path,
        dataType: "xml",
        success: function (xmlResponse) {
          // 0x02. parse xml file
          var datas = $("entry", xmlResponse).map(function () {
            return {
              title: $("title", this).text(),
              content: $("content", this).text(),
              url: $("url", this).text()
            };
          }).get();
          $resultContent.innerHTML = "";

          $input.addEventListener('input', function () {
            // 0x03. parse query to keywords list
            var str = '<ul class=\"search-result-list\">';
            var keywords = this.value.trim().toLowerCase().split(/[\s\-]+/);
            $resultContent.innerHTML = "";
            if (this.value.trim().length <= 0) {
              return;
            }
            // 0x04. perform local searching
            datas.forEach(function (data) {
              var isMatch = true;
              var content_index = [];
              if (!data.title || data.title.trim() === '') {
                data.title = "Untitled";
              }
              var orig_data_title = data.title.trim();
              var data_title = orig_data_title.toLowerCase();
              var orig_data_content = data.content.trim().replace(/<[^>]+>/g, "");
              var data_content = orig_data_content.toLowerCase();
              var data_url = data.url;
              var index_title = -1;
              var index_content = -1;
              var first_occur = -1;
              // only match artiles with not empty contents
              if (data_content !== '') {
                keywords.forEach(function (keyword, i) {
                  index_title = data_title.indexOf(keyword);
                  index_content = data_content.indexOf(keyword);

                  if (index_title < 0 && index_content < 0) {
                    isMatch = false;
                  } else {
                    if (index_content < 0) {
                      index_content = 0;
                    }
                    if (i == 0) {
                      first_occur = index_content;
                    }
                    // content_index.push({index_content:index_content, keyword_len:keyword_len});
                  }
                });
              } else {
                isMatch = false;
              }
              // 0x05. show search results
              if (isMatch) {
                str += "<li><a href='" + data_url + "' class='search-result-title'>" + orig_data_title + "</a>";
                var content = orig_data_content;
                if (first_occur >= 0) {
                  // cut out 100 characters
                  var start = first_occur - 20;
                  var end = first_occur + 80;

                  if (start < 0) {
                    start = 0;
                  }

                  if (start == 0) {
                    end = 100;
                  }

                  if (end > content.length) {
                    end = content.length;
                  }

                  var match_content = content.substr(start, end);

                  // highlight all keywords
                  keywords.forEach(function (keyword) {
                    var regS = new RegExp(keyword, "gi");
                    match_content = match_content.replace(regS, "<span class=\"search-keyword\">" + keyword + "</span>");
                  });

                  str += "<p class=\"search-result-abstract\">" + match_content + "...</p>"
                }
                str += "</li>";
              }
            });
            str += "</ul>";
            if (str.indexOf('<li>') === -1) {
              return $resultContent.innerHTML = "<ul><span class='local-search-empty'>没有找到内容，请尝试更换检索词。<span></ul>";
            }
            $resultContent.innerHTML = str;
          });
        },
        error: function(xhr, status, error) {
          $resultContent.innerHTML = ""
          if (xhr.status === 404) {
            $resultContent.innerHTML = "<ul><span class='local-search-empty'>未找到search.xml文件，具体请参考：<a href='https://github.com/zchengsite/hexo-theme-oranges#configuration' target='_black'>configuration</a><span></ul>";
          } else {
            $resultContent.innerHTML = "<ul><span class='local-search-empty'>请求失败，尝试重新刷新页面或稍后重试。<span></ul>";
          }
        }
      });
      $(document).on('click', '#search-close-icon', function() {
        $('#search-input').val('');
        $('#search-result').html('');
      });
    }

    var getSearchFile = function() {
        var path = "/search.xml";
        searchFunc(path, 'search-input', 'search-result');
    }
  </script>




    </div>
  </body>
</html>
