<!DOCTYPE html>
<html lang="zh-CN">

  <head>
  <meta charset="utf-8">
  <meta name="author" content="zchengsite, 1451426471@qq.com" />
  <meta name="viewport" content="width=device-width, initial-scale=1" />
  
  
  <title>测序数据常见文件类型 | Hexo</title>

  
    <link rel="apple-touch-icon" href="/images/favicon.png">
    <link rel="icon" href="/images/favicon.png">
  

  <!-- Raleway-Font -->
  <link href="https://fonts.googleapis.com/css?family=Raleway&display=swap" rel="stylesheet">

  <!-- hexo site css -->
  
<link rel="stylesheet" href="/css/base.css">
<link rel="stylesheet" href="/iconfont/iconfont.css">
<link rel="stylesheet" href="/css/github-markdown.css">
<link rel="stylesheet" href="/css/highlight.css">


  <!-- jquery3.3.1 -->
  <script src="https://cdn.bootcss.com/jquery/3.3.1/jquery.min.js"></script>

  <!-- fancybox -->
  <link href="https://cdn.bootcss.com/fancybox/3.5.2/jquery.fancybox.min.css" rel="stylesheet">
  <script async src="https://cdn.bootcss.com/fancybox/3.5.2/jquery.fancybox.min.js"></script>
  
<script src="/js/fancybox.js"></script>


  

<meta name="generator" content="Hexo 5.2.0"><style>mjx-container[jax="SVG"] {
  direction: ltr;
}

mjx-container[jax="SVG"] > svg {
  overflow: visible;
}

mjx-container[jax="SVG"] > svg a {
  fill: blue;
  stroke: blue;
}

mjx-container[jax="SVG"][display="true"] {
  display: block;
  text-align: center;
  margin: 1em 0;
}

mjx-container[jax="SVG"][justify="left"] {
  text-align: left;
}

mjx-container[jax="SVG"][justify="right"] {
  text-align: right;
}

g[data-mml-node="merror"] > g {
  fill: red;
  stroke: red;
}

g[data-mml-node="merror"] > rect[data-background] {
  fill: yellow;
  stroke: none;
}

g[data-mml-node="mtable"] > line[data-line] {
  stroke-width: 70px;
  fill: none;
}

g[data-mml-node="mtable"] > rect[data-frame] {
  stroke-width: 70px;
  fill: none;
}

g[data-mml-node="mtable"] > .mjx-dashed {
  stroke-dasharray: 140;
}

g[data-mml-node="mtable"] > .mjx-dotted {
  stroke-linecap: round;
  stroke-dasharray: 0,140;
}

g[data-mml-node="mtable"] > svg {
  overflow: visible;
}

[jax="SVG"] mjx-tool {
  display: inline-block;
  position: relative;
  width: 0;
  height: 0;
}

[jax="SVG"] mjx-tool > mjx-tip {
  position: absolute;
  top: 0;
  left: 0;
}

mjx-tool > mjx-tip {
  display: inline-block;
  padding: .2em;
  border: 1px solid #888;
  font-size: 70%;
  background-color: #F8F8F8;
  color: black;
  box-shadow: 2px 2px 5px #AAAAAA;
}

g[data-mml-node="maction"][data-toggle] {
  cursor: pointer;
}

mjx-status {
  display: block;
  position: fixed;
  left: 1em;
  bottom: 1em;
  min-width: 25%;
  padding: .2em .4em;
  border: 1px solid #888;
  font-size: 90%;
  background-color: #F8F8F8;
  color: black;
}

foreignObject[data-mjx-xml] {
  font-family: initial;
  line-height: normal;
  overflow: visible;
}

.MathJax path {
  stroke-width: 3;
}

mjx-container[display="true"] {
  overflow: auto hidden;
}

mjx-container[display="true"] + br {
  display: none;
}
</style><link rel="alternate" href="/atom.xml" title="Hexo" type="application/atom+xml">
</head>


  <body>
    <div id="app">
      <div class="header">
  <div class="avatar">
    <a href="/">
      <!-- 头像取消懒加载，添加no-lazy -->
      
        <img src="/images/mooner.jpg" alt="">
      
    </a>
    <div class="nickname"><a href="/">mooner</a></div>
  </div>
  <div class="navbar">
    <ul>
      
        <li class="nav-item" data-path="/">
          <a href="/">主页</a>
        </li>
      
        <li class="nav-item" data-path="/archives/">
          <a href="/archives/">归档</a>
        </li>
      
        <li class="nav-item" data-path="/tags/">
          <a href="/tags/">标签</a>
        </li>
      
        <li class="nav-item" data-path="/about/">
          <a href="/about/">关于</a>
        </li>
      
    </ul>
  </div>
</div>


<script src="/js/activeNav.js"></script>



      <div class="flex-container">
        <!-- 文章详情页，展示文章具体内容，url形式：https://yoursite/文章标题/ -->
<!-- 同时为「标签tag」，「朋友friend」，「分类categories」，「关于about」页面的承载页面，具体展示取决于page.type -->



  

  

  

  
  <!-- 文章内容页 url形式：https://yoursite/文章标题/ -->
  <div class="container post-details" id="post-details">
    <div class="post-content">
      <div class="post-title">测序数据常见文件类型</div>
      <div class="post-attach">
        <span class="post-pubtime">
          <i class="iconfont icon-updatetime" title="更新时间"></i>
          2020-10-02
        </span>
        
              <span class="post-tags">
                <i class="iconfont icon-tags" title="标签"></i>
                
                <span class="span--tag">
                  <a href="/tags/NGS/" title="NGS">
                    <b>#</b> NGS
                  </a>
                </span>
                
              </span>
          
      </div>
      <div class="markdown-body">
        <p>各行各业都有在自己的标准体系，生物信息学数据分析也不例外，各个厂商出品的芯片系列，还有各种NGS组学分析，都会涉及到不同的分析步骤，有着丰富多样的中间文件。其中一些常用的文件就被规定成文件格式。 文件格式那么多，都可以了解一二，当然，不需要背诵它们所有的细节，不过对下面我们单独拿出来详细介绍的还是尽量要耳熟能详。</p>
<p>简单来说，测序得到的是带有质量值的碱基序列(fastq格式)；参考基因组是(fasta格式)；用比对工具把fastq格式的序列回帖到对应的fasta格式的参考基因组序列，就可以产生sam格式的比对文件；把sam格式的文本文件压缩成二进制的bam文件可以节省空间；如果对参考基因组上面的各个区段标记它们的性质，比如哪些区域是外显子，内含子，UTR等等，这就是gtf/gff格式； 如果只是为了单纯描述某个基因组区域，就是bed格式文件，记录染色体号以及起始终止坐标，正负链即可；如果是记录某些位点或者区域碱基的变化，就是VCF文件格式；如果仅仅是为了追踪参考基因组的各个区域的覆盖度，测序深度就可以用bigwig/wig格式。上面是对这些生信中的数据格式进行简介，下面将对这些数据格式进行详述。</p>
<h2 id="FASTQ">FASTQ</h2>
<p><strong>简介</strong></p>
<p>FASTQ用于保存生物序列（通常是核酸序列）和其测序质量信息的标准格式。 其序列以及质量信息都是使用一个ASCII字符标示，最初由Sanger开发。 目的是将FASTA序列与质量数据放到一起，目前已经成为高通量测序结果的实施标准。</p>
<p><strong>一、定义和示例</strong></p>
<p>FASTQ文件中每个序列通常有四行：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">第一行是序列标识以及相关的描述信息，以‘@’开头 </span><br><span class="line">第二行是序列</span><br><span class="line">第三行以‘+’开头，后面是序列标示符、描述信息，或者什么也不加，但是“+”不能少。</span><br><span class="line">第四行，是质量信息，和第二行的序列相对应，每一个序列都有一个质量评分，根据评分体系的不同，每个字符的含义表示的数字也不相同。</span><br></pre></td></tr></table></figure>
<p>一个简单的示例如下：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">@SEQ_ID</span><br><span class="line">GATTTGGGGTTCAAAGCAGTATCGATCAAATAGTAAATCCATTTGTTCAACTCACAGTTT</span><br><span class="line">+</span><br><span class="line">!&#39;&#39;*((((***+))%%%++)(%%%%).1***-+*&#39;&#39;))**55CCF&gt;&gt;&gt;&gt;&gt;&gt;CCCCCCC65</span><br></pre></td></tr></table></figure>
<p><strong>二、序列标识</strong></p>
<p>上面说到第一行是序列标识以及相关的描述信息，以‘@’开头。可以像上面的示例那么简单，但如果是正规测序仪下机的真实数据，通常会很复杂。比如：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">@EAS139:136:FC706VJ:2:2104:15343:197393 1:Y:18:ATCACG</span><br></pre></td></tr></table></figure>
<p>这个序列标识以及相关描述信息以冒号分割，每一个字段信息如下：</p>
<table>
<thead>
<tr>
<th>字段</th>
<th>解释</th>
</tr>
</thead>
<tbody>
<tr>
<td>EAS139</td>
<td>the unique instrument name</td>
</tr>
<tr>
<td>136</td>
<td>the run id</td>
</tr>
<tr>
<td>FC706VJ</td>
<td>the flowcell id</td>
</tr>
<tr>
<td>2</td>
<td>flowcell lane</td>
</tr>
<tr>
<td>2104</td>
<td>tile number within the flowcell lane</td>
</tr>
<tr>
<td>15343</td>
<td>‘x’-coordinate of the cluster within the tile</td>
</tr>
<tr>
<td>197393</td>
<td>‘y’-coordinate of the cluster within the tile</td>
</tr>
<tr>
<td>1</td>
<td>the member of a pair, 1 or 2 (paired-end or mate-pair reads only)</td>
</tr>
<tr>
<td>Y</td>
<td>Y if the read fails filter (read is bad), N otherwise</td>
</tr>
<tr>
<td>18</td>
<td>0 when none of the control bits are on, otherwise it is an even number</td>
</tr>
<tr>
<td>ATCACG</td>
<td>index sequence</td>
</tr>
</tbody>
</table>
<p>当然，上面的表格介绍的只是其中一个测序仪下机数据，如果是其它机器，产商可以自由定义标识符格式，因为fastq格式的第一行只需要以@符号开头即可。</p>
<p>不过，也有一些时候fastq数据并不是测序仪直接下机的，而且他人上传到了NCBI的SRA中心，我们下载下来解压后一般就没有了测序仪相关的标识，例子如下：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">@SRR001666.1 071112_SLXA-EAS1_s_7:5:1:817:345 length&#x3D;36</span><br><span class="line">GGGTGATGGCCGCTGCCGATGGCGTCAAATCCCACC</span><br><span class="line">+SRR001666.1 071112_SLXA-EAS1_s_7:5:1:817:345 length&#x3D;36</span><br><span class="line">IIIIIIIIIIIIIIIIIIIIIIIIIIIIII9IG9IC</span><br></pre></td></tr></table></figure>
<p><strong>三、质量编码格式</strong></p>
<p>质量评分指的是一个碱基的错误概率的对数值。 其最初在Phred拼接软件中定义与使用，其后在许多软件中得到使用。 其质量得分与错误概率的对应关系见下表：</p>
<table>
<thead>
<tr>
<th>PHRED QUALITY SCORE</th>
<th>PROBABILITY OF INCORRECT BASE CALL</th>
<th>BASE CALL ACCURACY</th>
</tr>
</thead>
<tbody>
<tr>
<td>10</td>
<td>1 in 10</td>
<td>90 %</td>
</tr>
<tr>
<td>20</td>
<td>1 in 100</td>
<td>99%</td>
</tr>
<tr>
<td>30</td>
<td>1 in 1000</td>
<td>99.9%</td>
</tr>
<tr>
<td>40</td>
<td>1 in 10000</td>
<td>99.99%</td>
</tr>
<tr>
<td>50</td>
<td>1 in 100000</td>
<td>99.999%</td>
</tr>
</tbody>
</table>
<blockquote>
<p>Phred quality scores Q are defined as a property which is logarithmically related to the base-calling error probabilities P.<br>
Q=-10lgP</p>
</blockquote>
<p>除了Phred质量得分换算标准，还有就是Solexa标准：是把P换成p/(1-p)</p>
<p>对于每个碱基的质量编码标示，不同的软件采用不同的方案，目前有5种方案：</p>
<ol>
<li>Sanger，Phred quality score，值的范围从0到92，对应的ASCII码从33到126，但是对于测序数据（raw read data）质量得分通常小于60，序列拼接或者mapping可能用到更大的分数。</li>
<li>Solexa/Illumina 1.0, Solexa/Illumina quality score，值的范围从-5到63，对应的ASCII码从59到126，对于测序数据，得分一般在-5到40之间。</li>
<li>Illumina 1.3+，Phred quality score，值的范围从0到62对应的ASCII码从64到126，低于测序数据，得分在0到40之间；</li>
<li>Illumina 1.5+，Phred quality score，但是0到2作为另外的标示。</li>
<li>Illumina 1.8+</li>
</ol>
<p>下面是更为直观的表示：</p>
<p><img src="/images/filtype/wiki-fastq-format.png" alt=""></p>
<p><strong>四、文件后缀</strong></p>
<p>没有特别的规定，通常使用.fq, .fastq, .txt等。 但是要注意，这个文件格式主要指的是文本文件里面的每行每列的内容规则，并不是我们常见的计算机领域的mp3,mp4,avi,xls,doc等等。</p>
<p>其它注意事项:</p>
<ul>
<li>双端测序一般有两个文件（也可通过某种规则把两个文件合并成一个）。</li>
<li>第一个文件与第二个文件的行数完全一样，且测序序列的排列顺序完全一致。</li>
<li>在第一个文件中，描述信息的结尾是“/1”，表示是双端测序的一端；第二个文件中同样位置/行数的相对应的测序序列的描述信息则以“/2”结尾，表示是双端测序的另一端。（2.2.2的表2-5中有叙述）</li>
</ul>
<h2 id="FASTA">FASTA</h2>
<p><strong>简介</strong></p>
<p>FASTA格式用于表示核苷酸序列或氨基酸序列的格式。 在这种格式中碱基对或氨基酸用单个字母来编码，且允许在序列前添加序列名及注释。 fasta序列格式是blast组织数据的基本格式，无论是数据库还是查询序列，大多数情况都使用fasta序列格式。 它要比上一小节介绍的FASTQ格式简明很多。</p>
<p><strong>一、定义和示例</strong></p>
<p>总的来说，Fasta格式开始于一个标识符：“&gt;”，然后是一行描述，下面是的序列，直到下一个&quot;&gt;&quot;,表示下一条。</p>
<p>下面是一个来源于NCBI的fasta格式序列：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">&gt;gi|187608668|ref|NM_001043364.2| Bombyx mori moricin (Mor), mRNA</span><br><span class="line">AAACCGCGCAGTTATTTAAAATATGAATATTTTAAAACTTTTCTTTGTTTTTA</span><br><span class="line">TTGTGGCAATGTCTCTGGTGTCATGTAGTACAGCCGCTCCAGCAAAAATACCT</span><br><span class="line">ATCAAGGCCATTAAGACTGTAGGAAAGGCAGTCGGTAAAGGTCTAAGAGCCAT </span><br><span class="line">ATCAAGGCCATTAA</span><br></pre></td></tr></table></figure>
<p>Fasta格式首先以大于号“&gt;”开头，接着是序列的标识符“gi|187608668|ref|NM_001043364.2|”，然后是序列的描述信息。 换行后是序列信息，序列中允许空格，换行，空行，直到下一个大于号，表示该序列的结束。 下面简单给一个表格说明序列来源的数据库与对应的标识符</p>
<table>
<thead>
<tr>
<th>Database Name数据库名称</th>
<th>Identifier Syntax 标识符</th>
</tr>
</thead>
<tbody>
<tr>
<td>GenBank</td>
<td>```gb</td>
</tr>
<tr>
<td>EMBL Data Library</td>
<td>```emb</td>
</tr>
<tr>
<td>DDBJ, DNA Database of Japan</td>
<td>``` dbj</td>
</tr>
<tr>
<td>NBRF PIR</td>
<td>```pir</td>
</tr>
<tr>
<td>Protein Research Foundation</td>
<td>``` prf</td>
</tr>
<tr>
<td>SWISS-PROT</td>
<td>```sp</td>
</tr>
<tr>
<td>Brookhaven Protein Data Bank</td>
<td>```pdb</td>
</tr>
<tr>
<td>Patents</td>
<td>```pat</td>
</tr>
<tr>
<td>GenInfo Backbone Id</td>
<td>``` bbs</td>
</tr>
<tr>
<td>General database identifier</td>
<td>``` gnl</td>
</tr>
<tr>
<td>NCBI Reference Sequence</td>
<td>``` ref</td>
</tr>
<tr>
<td>Local Sequence identifier</td>
<td>```lcl</td>
</tr>
</tbody>
</table>
<p>通常情况下序列的标识符不会像上面的例子那样复杂，再复杂的标识符也是有规则的，上面的标识符是NCBI定义的，可以去其官网了解详情。</p>
<p><strong>二、序列中字母代表的含义</strong></p>
<p>FASTA格式支持的核苷酸代码如下：</p>
<table>
<thead>
<tr>
<th>核苷酸代码</th>
<th>意义</th>
</tr>
</thead>
<tbody>
<tr>
<td>A</td>
<td>Adenosine</td>
</tr>
<tr>
<td>C</td>
<td>Cytosine</td>
</tr>
<tr>
<td>G</td>
<td>Guanine</td>
</tr>
<tr>
<td>T</td>
<td>Thymidine</td>
</tr>
<tr>
<td>U</td>
<td>Uracil</td>
</tr>
<tr>
<td>R</td>
<td>G A (puRine)</td>
</tr>
<tr>
<td>Y</td>
<td>T C (pYrimidine)</td>
</tr>
<tr>
<td>K</td>
<td>G T (Ketone)</td>
</tr>
<tr>
<td>M</td>
<td>A C (aMino group)</td>
</tr>
<tr>
<td>S</td>
<td>G C (Strong interaction)</td>
</tr>
<tr>
<td>W</td>
<td>A T (Weak interaction)</td>
</tr>
<tr>
<td>B</td>
<td>G T C (not A) (B comes after A)</td>
</tr>
<tr>
<td>D</td>
<td>G A T (not C) (D comes after C)</td>
</tr>
<tr>
<td>H</td>
<td>A C T (not G) (H comes after G)</td>
</tr>
<tr>
<td>V</td>
<td>G C A (not T, not U) (V comes after U)</td>
</tr>
<tr>
<td>N</td>
<td>A G C T (aNy)</td>
</tr>
<tr>
<td>X</td>
<td>masked</td>
</tr>
<tr>
<td>-</td>
<td>gap of indeterminate length</td>
</tr>
</tbody>
</table>
<p>FASTA格式支持的氨基酸代码如下：</p>
<table>
<thead>
<tr>
<th>氨基酸代码</th>
<th>意义</th>
</tr>
</thead>
<tbody>
<tr>
<td>A</td>
<td>Alanine</td>
</tr>
<tr>
<td>B</td>
<td>Aspartic acid or Asparagine</td>
</tr>
<tr>
<td>C</td>
<td>Cysteine</td>
</tr>
<tr>
<td>D</td>
<td>Aspartic acid</td>
</tr>
<tr>
<td>E</td>
<td>Glutamic acid</td>
</tr>
<tr>
<td>F</td>
<td>Phenylalanine</td>
</tr>
<tr>
<td>G</td>
<td>Glycine</td>
</tr>
<tr>
<td>H</td>
<td>Histidine</td>
</tr>
<tr>
<td>I</td>
<td>Isoleucine</td>
</tr>
<tr>
<td>K</td>
<td>Lysine</td>
</tr>
<tr>
<td>L</td>
<td>Leucine</td>
</tr>
<tr>
<td>M</td>
<td>Methionine</td>
</tr>
<tr>
<td>N</td>
<td>Asparagine</td>
</tr>
<tr>
<td>O</td>
<td>Pyrrolysine</td>
</tr>
<tr>
<td>P</td>
<td>Proline</td>
</tr>
<tr>
<td>Q</td>
<td>Glutamine</td>
</tr>
<tr>
<td>R</td>
<td>Arginine</td>
</tr>
<tr>
<td>S</td>
<td>Serine</td>
</tr>
<tr>
<td>T</td>
<td>Threonine</td>
</tr>
<tr>
<td>U</td>
<td>Selenocysteine</td>
</tr>
<tr>
<td>V</td>
<td>Valine</td>
</tr>
<tr>
<td>W</td>
<td>Tryptophan</td>
</tr>
<tr>
<td>Y</td>
<td>Tyrosine</td>
</tr>
<tr>
<td>Z</td>
<td>Glutamic acid or Glutamine</td>
</tr>
<tr>
<td>X</td>
<td>any</td>
</tr>
<tr>
<td>*</td>
<td>translation stop</td>
</tr>
<tr>
<td>-</td>
<td>gap of indeterminate length</td>
</tr>
</tbody>
</table>
<p><strong>注意事项：</strong></p>
<p>对于自己构建的序列数据库（序列不是来源与NCBI或其他数据），可以采用“gnl|database|identifier”或者“lcl|identifier”格式，以保证可以使用blast的所有功能。 database或者identifier是需要指定的数据库的标识和序列标识，指定的名称可以用大小写字母、数字、下划线“_”、破折号“-”或者点号“.”。 注意名称是区分大小写的，同时不能出现空格，空格表示序列标识符结束。</p>
<p>数据库中的序列标识符必须保证唯一，许多时候格式数据库是formatdb报告错误，就是因为标示符重复，还有一点需要强调的是序列不能为空，否则也会报错。</p>
<h2 id="SAM格式">SAM格式</h2>
<p><strong>简介</strong></p>
<p>SAM格式主要应用于测序序列mapping到基因组上的结果表示，当然也可以表示任意的多重比对结果。 SAM是一种序列比对格式标准，由sanger制定，是以TAB为分割符的文本格式。 SAM的全称是sequence alignment/map format。</p>
<p><strong>一、定义和示例</strong></p>
<p>SAM分为两部分，注释信息（header section ）和比对结果部分 （alignment section）。 通常是把FASTQ文件格式的测序数据比对到对应的参考基因组版本得到的。 注释信息并不是SAM文件的重点，是该SAM文件产生以及被处理过程的一个记录，规定以@开头，用不同的tag表示不同的信息，主要有：</p>
<ul>
<li>@HD，说明符合标准的版本、对比序列的排列顺序；</li>
<li>@SQ，参考序列说明；</li>
<li>@RG，比对上的序列（read）说明；</li>
<li>@PG，使用的程序说明；</li>
<li>@CO，任意的说明信息。</li>
</ul>
<p>一个简单的SAM文件例子如下：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">@HD VN:1.0 SO:unsorted</span><br><span class="line">@SQ SN:chr1 LN:249250621</span><br><span class="line">@SQ SN:chr2 LN:243199373</span><br><span class="line">@PG ID:Bowtie VN:1.0.0 CL:&quot;bowtie genome&#x2F;hg19 -q reads&#x2F;SRR3101251.fastq -m 1 -p 4 -S&quot;</span><br><span class="line">SRR3101251.1 0 chr19 9486878 255 49M * 0 0 NTACTCCCACTACTCTCAGATTCAAGCAATCCTCCCACCCTAGCCCACC #1&#x3D;DDDFFHHHHHIHHIJJJHIJIIJIHIFHJIIJJJJJJJIIJJJJJJ XA:i:1 MD:Z:0A48 NM:i:1</span><br><span class="line">SRR3101251.5 16 chr2 240279787 255 49M * 0 0 CCTGAATCCATCAGAGCAGCCGGGCTGTGACACTCACTGTCATGATGTT JIJJIHIIIIJJJJJJJJJGHJJJJIIHJHICJIGCHHHHHFFFFFCCC XA:i:0 MD:Z:49 NM:i:0</span><br><span class="line">SRR3101251.6 4 * 0 0 * * 0 0 NATTCCCACCTATGAGTGAGAATATGCGGTGTTTGGTTTTTTGTTCTTG #1&#x3D;DDDFFHHHHHJJJGHIJJJJJJJJJJCGGIIJJIIJJJIJHJIIJJ XM:i:1</span><br></pre></td></tr></table></figure>
<p>前四行是注释信息，其后是比对结果，下面对比对结果进行解释，它是SAM格式文件的精华部分。</p>
<p><strong>二、比对结果详解</strong></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">SRR035022.2621862 163 16 59999 37 22S54M &#x3D; 60102 179 CCAACCCAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCGACCCTCACCCTCACCC &gt;AAA&#x3D;&gt;?AA&gt;@@B@B?AABAB?AABAB?AAC@B?@AB@A?A&gt;A@A?AAAAB??ABAB?79A?AAB;B?@?@&lt;&#x3D;8:8 XT:A:M XN:i:2 SM:i:37 AM:i:37 XM:i:0 XO:i:0 XG:i:0 RG:Z:SRR035022 NM:i:2 MD:Z:0N0N52 OQ:Z:CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCBCCCCCCBBCC@CCCCCCCCCCACCCCC;CCCBBC?CCCACCACA@</span><br></pre></td></tr></table></figure>
<p>所以在我们的例子中，每一个字段的说明如下：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><span class="line">QNAME	SRR035022.2621862</span><br><span class="line">FLAG	163</span><br><span class="line">RNAME	16</span><br><span class="line">POS	59999</span><br><span class="line">MAQ	37</span><br><span class="line">CIGAR	22S54M</span><br><span class="line">MRNM	&#x3D;</span><br><span class="line">MPOS	60102</span><br><span class="line">ISIZE	179</span><br><span class="line">SEQ	CCAACCCAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCGACCCTCACCCTCACCC</span><br><span class="line">QUAL	&gt;AAA&#x3D;&gt;?AA&gt;@@B@B?AABAB?AABAB?AAC@B?@AB@A?A&gt;A@A?AAAAB??ABAB?79A?AAB;B?@?@&lt;&#x3D;8:8</span><br><span class="line">TAG	XT:A:M</span><br><span class="line">TAG	XN:i:2</span><br><span class="line">TAG	SM:i:37</span><br><span class="line">TAG	AM:i:37</span><br><span class="line">TAG	XM:i:0</span><br><span class="line">TAG	XO:i:0</span><br><span class="line">TAG	XG:i:0</span><br><span class="line">TAG	RG:Z:SRR035022</span><br><span class="line">TAG	NM:i:2</span><br><span class="line">TAG	MD:Z:0N0N52</span><br><span class="line">TAG	OQ:Z:CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCBCCCCCCBBCC@CCCCCCCCCCACCCCC;CCCBBC?CCCACCACA</span><br></pre></td></tr></table></figure>
<p>比对结果部分（alignment section），每一行表示一个片段（segment）的比对信息，包括11个必须的字段（mandatory fields）和一个可选的字段，字段之间用tag分割。</p>
<p>必须的字段有11个，顺序固定，不可自行改动，根据字段定义，可以为’0‘或者’*‘，这是11个字段包括：</p>
<ol>
<li>QNAME，比对片段的（template）的编号；</li>
<li>FLAG，位标识，template mapping情况的数字表示，每一个数字代表一种比对情况，这里的值是符合情况的数字相加总和； （picard专门有一个工具解读sam的flag:<a target="_blank" rel="noopener" href="http://broadinstitute.github.io/picard/explain-flags.html%EF%BC%89">http://broadinstitute.github.io/picard/explain-flags.html）</a></li>
</ol>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">1	The read is one of a pair  read是pair中的一条（read表示本条read，mate表示pair中的另一条read）</span><br><span class="line">2	The alignment is one end of a proper paired-end alignment  pair一正一负完美的比对上</span><br><span class="line">4	The read has no reported alignments  这条read没有比对上</span><br><span class="line">8	The read is one of a pair and has no reported alignments  mate没有比对上</span><br><span class="line">16	The alignment is to the reverse reference strand  这条read反向比对</span><br><span class="line">32	The other mate in the paired-end alignment is aligned to the reverse reference strand   mate反向比对</span><br><span class="line">64	The read is the first (#1) mate in a pair  这条read是read1</span><br><span class="line">128	The read is the second (#2) mate in a pair  这条read是read2</span><br></pre></td></tr></table></figure>
<ol>
<li>RNAME，参考序列的编号，如果注释中对SQ-SN进行了定义，这里必须和其保持一致，另外对于没有mapping上的序列，这里是’*‘；</li>
<li>POS，比对上的位置，注意是从1开始计数，没有比对上，此处为0；</li>
<li>MAPQ，mappint的质量；</li>
<li>CIGAR，简要比对信息表达式（Compact Idiosyncratic Gapped Alignment Report），其以参考序列为基础，使用数字加字母表示比对结果，比如3S6M1P1I4M，前三个碱基被剪切去除了，然后6个比对上了，然后打开了一个缺口，有一个碱基插入，最后是4个比对上了，是按照顺序的；</li>
</ol>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">“M”表示 match或 mismatch；</span><br><span class="line">“I”表示 insert；</span><br><span class="line">“D”表示 deletion；</span><br><span class="line">“N”表示 skipped（跳过这段区域）；</span><br><span class="line">“S”表示 soft clipping（被剪切的序列存在于序列中）；</span><br><span class="line">“H”表示 hard clipping（被剪切的序列不存在于序列中）；</span><br><span class="line">“P”表示 padding；</span><br><span class="line">“&#x3D;”表示 match；</span><br><span class="line">“X”表示 mismatch（错配，位置是一一对应的）；</span><br></pre></td></tr></table></figure>
<ol>
<li>RNEXT，下一个片段比对上的参考序列的编号，没有另外的片段，这里是’*‘，同一个片段，用’=‘；</li>
<li>PNEXT，下一个片段比对上的位置，如果不可用，此处为0；</li>
<li>TLEN，Template的长度，最左边得为正，最右边的为负，中间的不用定义正负，不分区段（single-segment)的比对上，或者不可用时，此处为0；</li>
<li>SEQ，序列片段的序列信息，如果不存储此类信息，此处为’*‘，注意CIGAR中M/I/S/=/X对应数字的和要等于序列长度；</li>
<li>QUAL，序列的质量信息，格式同FASTQ一样。read质量的ASCII编码。</li>
<li>可选字段（optional fields)，格式如：TAG:TYPE:VALUE，其中TAG有两个大写字母组成，每个TAG代表一类信息，每一行一个TAG只能出现一次，TYPE表示TAG对应值的类型，可以是字符串、整数、字节、数组等。</li>
</ol>
<p><strong>三、SAM要处理好的问题</strong></p>
<ul>
<li>非常多序列（read)，mapping到多个参考基因组（reference）上</li>
<li>同一条序列，分多段（segment）比对到参考基因组上</li>
<li>无限量的，结构化信息表示，包括错配、删除、插入等比对信息</li>
</ul>
<p>要注意的几个概念，以及与之对应的模型：</p>
<ul>
<li>reference</li>
<li>read</li>
<li>segment</li>
<li>template（参考序列和比对上的序列共同组成的序列为template）</li>
<li>alignment</li>
<li>seq</li>
</ul>
<h2 id="BAM-格式">BAM 格式</h2>
<p><strong>简介</strong></p>
<p>本质上就是二进制压缩的SAM文件，大部分生物信息学流程都需要这个格式，为了节省存储空间以及方便索引。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"># BiocInstaller::biocLite(&#39;Rsamtools&#39;)</span><br><span class="line">library(Rsamtools) </span><br><span class="line">test_bam_file &lt;- &#39;data&#x2F;CHIP-seq.bam&#39; </span><br><span class="line">#fileter bam</span><br><span class="line">filter &lt;- FilterRules(list(MinWidth &#x3D; function(x) width(x$seq) &gt; 35))</span><br><span class="line">res &lt;- scanBam(test_bam_file, filter&#x3D;filter)[[1]]</span><br><span class="line">sapply(res, head)</span><br></pre></td></tr></table></figure>
<p>从上面的例子（需要在R里面运行）可以看到BAM文件需要用特殊的方法来读取，可以是R里面的Rsamtools包，也可以是linux环境下安装好的samtools软件，因为它是二进制文件，不能像普通的文本文件那样来打开。</p>
<p>我们用R里面的head函数查看了该BAM文件的前6行，比对的flag分别是<code>16 0 16 16 0 0</code>,说明有3条序列没有成功比对到基因组。width信息说明该序列长度都是36bp。序列的碱基以及对应的碱基质量也如上所述。</p>
<h2 id="VCF">VCF</h2>
<p><strong>简介</strong></p>
<p>Variant Call Format（VCF）是一个用于存储基因序列突变信息的文本格式。 可以表示单碱基突变, 插入/缺失, 拷贝数变异和结构变异等。 通常是对BAM文件格式的比对结果进行处理得到的。 BCF格式文件是VCF格式的二进制文件。</p>
<p><strong>一、定义和示例</strong></p>
<p><img src="/images/filtype/vcf_variation_class.jpg" alt=""></p>
<p>如上图所示，vcf记录的即为各类型的变异。例如：点突变，拷贝数变异，插入，缺失等结构变异。</p>
<p>VCF分为两部分，注释信息和变异位点记录信息。</p>
<p>注释信息通常以#开头，会描述该VCF版本，目前以4.2居多，然后会一行行记录变异位点信息里面会出现的所有TAG。</p>
<p>下面这个是NCBI的dbSNP数据库里面的人类的vcf文件的部分截取：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><span class="line">##fileformat&#x3D;VCFv4.0</span><br><span class="line">##fileDate&#x3D;20160601</span><br><span class="line">##source&#x3D;dbSNP</span><br><span class="line">##dbSNP_BUILD_ID&#x3D;147</span><br><span class="line">##reference&#x3D;GRCh37.p13</span><br><span class="line">##phasing&#x3D;partial</span><br><span class="line">##variationPropertyDocumentationUrl&#x3D;ftp:&#x2F;&#x2F;ftp.ncbi.nlm.nih.gov&#x2F;snp&#x2F;specs&#x2F;dbSNP_BitField_latest.pdf      </span><br><span class="line">##INFO&#x3D;&lt;ID&#x3D;RS,Number&#x3D;1,Type&#x3D;Integer,Description&#x3D;&quot;dbSNP ID (i.e. rs number)&quot;&gt;</span><br><span class="line">##INFO&#x3D;&lt;ID&#x3D;RSPOS,Number&#x3D;1,Type&#x3D;Integer,Description&#x3D;&quot;Chr position reported in dbSNP&quot;&gt;</span><br><span class="line">##INFO&#x3D;&lt;ID&#x3D;RV,Number&#x3D;0,Type&#x3D;Flag,Description&#x3D;&quot;RS orientation is reversed&quot;&gt;</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO</span><br><span class="line">1       10177   rs201752861     A       C       .       .       RS&#x3D;201752861;RSPOS&#x3D;10177;dbSNPBuildID&#x3D;137;SSR&#x3D;0;SAO&#x3D;0;VP&#x3D;0x050000020005000002000100;GENEINFO&#x3D;DDX11L1:100287102;WGT&#x3D;1;VC&#x3D;SNV;R5;ASP</span><br><span class="line">1       10177   rs367896724     A       AC      .       .       RS&#x3D;367896724;RSPOS&#x3D;10177;dbSNPBuildID&#x3D;138;SSR&#x3D;0;SAO&#x3D;0;VP&#x3D;0x050000020005170026000200;GENEINFO&#x3D;DDX11L1:100287102;WGT&#x3D;1;VC&#x3D;DIV;R5;ASP;VLD;G5A;G5;KGPhase3;CAF&#x3D;0.5747,0.4253;COMMON&#x3D;1</span><br><span class="line">1       10228   rs143255646     TA      T       .       .       RS&#x3D;143255646;RSPOS&#x3D;10229;dbSNPBuildID&#x3D;134;SSR&#x3D;0;SAO&#x3D;0;VP&#x3D;0x050000020005000002000200;GENEINFO&#x3D;DDX11L1:100287102;WGT&#x3D;1;VC&#x3D;DIV;R5;ASP</span><br><span class="line">1       10228   rs200462216     TAACCCCTAACCCTAACCCTAAACCCTA    T       .       .       RS&#x3D;200462216;RSPOS&#x3D;10229;dbSNPBuildID&#x3D;137;SSR&#x3D;0;SAO&#x3D;0;VP&#x3D;0x050000020005000002000200;GENEINFO&#x3D;DDX11L1:100287102;WGT&#x3D;1;VC&#x3D;DIV;R5;ASP</span><br><span class="line">1       10230   rs775928745     AC      A       .       .       RS&#x3D;775928745;RSPOS&#x3D;10231;dbSNPBuildID&#x3D;144;SSR&#x3D;0;SAO&#x3D;0;VP&#x3D;0x050000020005000002000200;GENEINFO&#x3D;DDX11L1:100287102;WGT&#x3D;1;VC&#x3D;DIV;R5;ASP</span><br><span class="line">1       10231   rs200279319     C       A       .       .       RS&#x3D;200279319;RSPOS&#x3D;10231;dbSNPBuildID&#x3D;137;SSR&#x3D;0;SAO&#x3D;0;VP&#x3D;0x050000020005000002000100;GENEINFO&#x3D;DDX11L1:100287102;WGT&#x3D;1;VC&#x3D;SNV;R5;ASP</span><br><span class="line">1       10234   rs145599635     C       T       .       .       RS&#x3D;145599635;RSPOS&#x3D;10234;dbSNPBuildID&#x3D;134;SSR&#x3D;0;SAO&#x3D;0;VP&#x3D;0x050100020005000002000100;GENEINFO&#x3D;DDX11L1:100287102;WGT&#x3D;1;VC&#x3D;SNV;SLO;R5;ASP</span><br><span class="line">1       10235   rs540431307     T       TA      .       .       RS&#x3D;540431307;RSPOS&#x3D;10235;dbSNPBuildID&#x3D;142;SSR&#x3D;0;SAO&#x3D;0;VP&#x3D;0x050000020005040024000200;GENEINFO&#x3D;DDX11L1:100287102;WGT&#x3D;1;VC&#x3D;DIV;R5;ASP;VLD;KGPhase3;CAF&#x3D;0.9988,0.001198;COMMON&#x3D;0</span><br><span class="line">1       10247   rs796996180     T       C       .       .       RS&#x3D;796996180;RSPOS&#x3D;10247;dbSNPBuildID&#x3D;146;SSR&#x3D;0;SAO&#x3D;0;VP&#x3D;0x050100020005000002000100;GENEINFO&#x3D;DDX11L1:100287102;WGT&#x3D;1;VC&#x3D;SNV;SLO;R5;ASP</span><br></pre></td></tr></table></figure>
<p>限于文章篇幅限制，我只是截取了该VCF文件的部分注释信息，很明显可以看到注释信息刚刚开始的几行其实是没有规则的，只需要以##开头即可，描述一些必备信息，包括参考基因组版本，得到该VCF文件的命令是什么等等。</p>
<p>后面的都是以INFO=&lt;ID=······&gt;的形式来介绍一个个TAG，这些TAG都是会在VCF的正文，变异位点记录里面用到的。而且每个tag都很容易理解，就是对应的英文描述而已。</p>
<p>接下来我们看看比较复杂的正文部分，就是变异位点记录信息。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  sample1</span><br><span class="line">1       858691  .       TG      T       222     .       INDEL;IDV&#x3D;37;IMF&#x3D;0.486842;DP&#x3D;76;VDB&#x3D;0.110516;SGB&#x3D;-0.693139;MQSB&#x3D;1;MQ0F&#x3D;0;ICB&#x3D;1;HOB&#x3D;0.5;AC&#x3D;1;AN&#x3D;2;DP4&#x3D;12,24,14,22;MQ&#x3D;60  GT:PL   0&#x2F;1:255,0,255</span><br><span class="line">1       858801  .       A       G       222     .       DP&#x3D;59;VDB&#x3D;0.728126;SGB&#x3D;-0.692717;RPB&#x3D;0.748623;MQB&#x3D;1;MQSB&#x3D;1;BQB&#x3D;0.963908;MQ0F&#x3D;0;ICB&#x3D;1;HOB&#x3D;0.5;AC&#x3D;1;AN&#x3D;2;DP4&#x3D;14,17,11,12;MQ&#x3D;60    GT:PL   0&#x2F;1:255,0,255</span><br><span class="line">1       859404  .       C       G       222     .       DP&#x3D;81;VDB&#x3D;0.0896228;SGB&#x3D;-0.693132;RPB&#x3D;0.849598;MQB&#x3D;1;MQSB&#x3D;1;BQB&#x3D;0.486963;MQ0F&#x3D;0;ICB&#x3D;1;HOB&#x3D;0.5;AC&#x3D;1;AN&#x3D;2;DP4&#x3D;17,15,18,16;MQ&#x3D;60   GT:PL   0&#x2F;1:255,0,255</span><br><span class="line">1       859690  .       C       G       222     .       DP&#x3D;75;VDB&#x3D;0.0662538;SGB&#x3D;-0.69312;RPB&#x3D;0.959181;MQB&#x3D;1;MQSB&#x3D;1;BQB&#x3D;0.962588;MQ0F&#x3D;0;ICB&#x3D;1;HOB&#x3D;0.5;AC&#x3D;1;AN&#x3D;2;DP4&#x3D;20,15,18,14;MQ&#x3D;60    GT:PL   0&#x2F;1:255,0,255</span><br><span class="line">1       859701  .       C       G       222     .       DP&#x3D;74;VDB&#x3D;0.274853;SGB&#x3D;-0.693127;RPB&#x3D;0.97201;MQB&#x3D;1;MQSB&#x3D;1;BQB&#x3D;0.717302;MQ0F&#x3D;0;ICB&#x3D;1;HOB&#x3D;0.5;AC&#x3D;1;AN&#x3D;2;DP4&#x3D;19,15,19,14;MQ&#x3D;60     GT:PL   0&#x2F;1:255,0,255</span><br><span class="line">1       859913  .       A       G       222     .       DP&#x3D;67;VDB&#x3D;0.756546;SGB&#x3D;-0.693139;RPB&#x3D;0.950685;MQB&#x3D;1;MQSB&#x3D;1;BQB&#x3D;0.662934;MQ0F&#x3D;0;ICB&#x3D;1;HOB&#x3D;0.5;AC&#x3D;1;AN&#x3D;2;DP4&#x3D;18,10,19,17;MQ&#x3D;60    GT:PL   0&#x2F;1:255,0,255</span><br><span class="line">1       860416  .       G       A       222     .       DP&#x3D;79;VDB&#x3D;0.673886;SGB&#x3D;-0.693144;RPB&#x3D;0.11919;MQB&#x3D;1;MQSB&#x3D;1;BQB&#x3D;0.992984;MQ0F&#x3D;0;ICB&#x3D;1;HOB&#x3D;0.5;AC&#x3D;1;AN&#x3D;2;DP4&#x3D;18,15,24,15;MQ&#x3D;60     GT:PL   0&#x2F;1:255,0,255</span><br></pre></td></tr></table></figure>
<p>每一行代表一个Variant的信息。</p>
<p>CHROM 和 POS：代表参考序列名和variant的位置；如果是INDEL的话，位置是INDEL的第一个碱基位置。</p>
<p>ID：variant的ID。比如在dbSNP中有该SNP的id，则会在此行给出(这个需要自己下载dbSNP数据库文件进行注释才有的)。 若没有或者注释不上，则用’.'表示其为一个novel variant。</p>
<p>REF 和 ALT：参考序列的碱基 和 Variant的碱基。</p>
<p>QUAL：Phred格式(Phred_scaled)的质量值，表示在该位点存在variant的可能性；该值越高，则variant的可能性越大； 计算方法：Phred值 = -10 * log (1-p) p为variant存在的概率; 通过计算公式可以看出值为10的表示错误概率为0.1，该位点为variant的概率为90%。</p>
<p>FILTER：使用上一个QUAL值来进行过滤的话，是不够的。GATK能使用其它的方法来进行过滤，过滤结果中通过则该值为”PASS”;若variant不可靠，则该项不为”PASS”或”.”。</p>
<p>INFO： 这一行是variant的详细信息，内容很多，以下再具体详述。</p>
<p>FORMAT 和 sample1 ：这两行合起来提供了 sample1 这个sample的基因型的信息。’sample1′代表这该名称的样品，是由SAM/BAM文件中的@RG下的 SM 标签决定的。 (当然并不是所有的VCF都是由一个BAM文件产生，比如数据库dbSNP提供的vcf文件，就没有样本信息啦)</p>
<p><strong>1、第8列的INFO</strong></p>
<p>该列信息最多了，都是以 “TAG=Value”, 并使用”;”分隔的形式 。 其中 很多的注释信息在VCF文件的头部注释中给出。以下是这些TAG的解释</p>
<p>AC，AF 和 AN：AC(Allele Count) 表示该Allele的数目；AF(Allele Frequency) 表示Allele的频率； AN(Allele Number) 表示Allele的总数目。对于1个diploid sample而言：则基因型 0/1 表示sample为杂合子，Allele数为1(双倍体的sample在该位点只有1个等位基因发生了突变)，Allele的频率为0.5(双倍体的 sample在该位点只有50%的等位基因发生了突变)，总的Allele为2； 基因型 1/1 则表示sample为纯合的，Allele数为2，Allele的频率为1，总的Allele为2。</p>
<p>DP：reads覆盖度。是一些reads被过滤掉后的覆盖度。</p>
<p>Dels：Fraction of Reads Containing Spanning Deletions。进行SNP和INDEL calling的结果中，有该TAG并且值为0表示该位点为SNP，没有则为INDEL。</p>
<p>FS：使用Fisher’s精确检验来检测strand bias而得到的Fhred格式的p值。该值越小越好。一般进行filter的时候，可以设置 FS &lt; 10～20。</p>
<p>HaplotypeScore：Consistency of the site with at most two segregating haplotypes</p>
<p>InbreedingCoeff：Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hard-Weinberg expectation</p>
<p>MLEAC：Maximum likelihood expectation (MLE) for the allele counts (not necessarily the same as the AC), for each ALT allele, in the same order as listed</p>
<p>MLEAF：Maximum likelihood expectation (MLE) for the allele frequency (not necessarily the same as the AF), for each ALT alle in the same order as listed</p>
<p>MQ：RMS Mapping Quality</p>
<p>MQ0：Total Mapping Quality Zero Reads</p>
<p>MQRankSum：Z-score From Wilcoxon rank sum test of Alt vs. Ref read mapping qualities</p>
<p>QD：Variant Confidence/Quality by Depth</p>
<p>RPA：Number of times tandem repeat unit is repeated, for each allele (including reference)</p>
<p>RU：Tandem repeat unit (bases)</p>
<p>ReadPosRankSum：Z-score from Wilcoxon rank sum test of Alt vs. Ref read position bias</p>
<p>STR：Variant is a short tandem repeat</p>
<p><strong>2、9和10列代表基因型</strong></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">GT:AD:DP:GQ:PL 0&#x2F;1:173,141:282:99:255,0,255 </span><br><span class="line">GT:AD:DP:GQ:PL 0&#x2F;1:1,3:4:25.92:103,0,26</span><br></pre></td></tr></table></figure>
<p>看上面最后两列数据，这两列数据是对应的，前者为格式，后者为格式对应的数据。这些TAG也是可以在VCF的头文件找到的</p>
<p>GT：样品的基因型（genotype）。两个数字中间用’/'分开，这两个数字表示双倍体的sample的基因型。0 表示样品中有ref的allele； 1 表示样品中variant的allele； 2表示有第二个variant的allele。因此： 0/0 表示sample中该位点为纯合的，和ref一致； 0/1 表示sample中该位点为杂合的，有ref和variant两个基因型； 1/1 表示sample中该位点为纯合的，和variant一致。</p>
<p>AD 和 DP：AD(Allele Depth)为sample中每一种allele的reads覆盖度，在diploid中则是用逗号分割的两个值，前者对应ref基因型，后者对应variant基因型。</p>
<p>DP（Depth）为sample中该位点的测序深度。</p>
<p>GQ：基因型的质量值(Genotype Quality)。Phred格式(Phred_scaled)的质量值，表示在该位点该基因型存在的可能性；该值越高，则Genotype的可能性越大；计算方法：Phred值 = -10 * log (1-p) p为基因型存在的概率。</p>
<p>PL：指定的三种基因型的质量值(provieds the likelihoods of the given genotypes)。这三种指定的基因型为(0/0,0/1,1/1)，这三种基因型的概率总和为1。和之前不一致，该值越大，表明为该种基因型的可能性越小。 Phred值 = -10 * log § p为基因型存在的概率。</p>
<p><img src="/images/filtype/vcf_file_format.jpg" alt=""></p>
<h2 id="GTF和GFF">GTF和GFF</h2>
<p><strong>简介</strong></p>
<p>GFF全称为general feature format，这种格式主要是用来注释基因组。 GTF全称为gene transfer format，主要是用来对基因进行注释。 GTF和GFF格式是Sanger研究所定义，是一种简单的、方便的对于DNA、RNA以及蛋白质序列的特征进行描述的一种数据格式. 比如序列的哪里到哪里是基因，是转录本，是外显子，内含子或者CDS等等，已经成为序列注释的通用格式，许多软件都支持输入或者输出gff格式。</p>
<p><strong>一、定义和示例</strong></p>
<p>gff由tab键隔开的9列组成，以下是各列的说明：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">Column 1: “seqid”</span><br><span class="line">序列的编号，编号的有效字符[a-zA-Z0-9.:^*$@!+_?-|]</span><br><span class="line">Column 2: “source”</span><br><span class="line">注释信息的来源，比如”Genescan”、”Genbank” 等，可以为空，为空用”.”点号代替</span><br><span class="line">Column 3: “type”</span><br><span class="line">注释信息的类型，比如Gene、cDNA、mRNA等，或者是SO对应的编号</span><br><span class="line">Columns 4 &amp; 5: “start” and “end”</span><br><span class="line">开始与结束的位置，注意计数是从1开始的。结束位置不能大于序列的长度</span><br><span class="line">Column 6: “score”</span><br><span class="line">得分，数字，是注释信息可能性的说明，可以是序列相似性比对时的E-values值或者基因预测是的P-values值。”.”表示为空。</span><br><span class="line">Column 7: “strand”</span><br><span class="line">序列的方向， +表示正义链, -反义链 , ? 表示未知.</span><br><span class="line">Column 8: “phase”</span><br><span class="line">仅对注释类型为 “CDS”有效，表示起始编码的位置，有效值为0、1、2。</span><br><span class="line">Column 9: “attributes”</span><br><span class="line">以多个键值对组成的注释信息描述，键与值之间用”&#x3D;“，不同的键值用”;“隔开，一个键可以有多个值，不同值用”,“分割。注意如果描述中包括tab键以及”,&#x3D;;”，要用URL转义规则进行转义，如tab键用 %09代替。键是区分大小写的，以大写字母开头的键是预先定义好的，在后面可能被其他注释信息所调用。</span><br></pre></td></tr></table></figure>
<p>预先定义的键包括：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">ID 注释信息的编号，在一个GFF文件中必须唯一；</span><br><span class="line">Name 注释信息的名称，可以重复；</span><br><span class="line">Alias 别名</span><br><span class="line">Parent Indicates 该注释所属的注释，值为注释信息的编号，比如外显子所属的转录组编号，转录组所属的基因的编号。值可以为多个。</span><br><span class="line">Target Indicates： the target of a nucleotide-to-nucleotide or protein-to-nucleotide alignment.（核苷酸对核苷酸或蛋白质至核苷酸比对的靶点。）</span><br><span class="line">Gap：The alignment of the feature to the target if the two are not collinear (e.g. contain gaps).（如果两者不共线（例如包含间隙），则该特征与目标的对准）</span><br><span class="line">Derives_from：Used to disambiguate the relationship between one feature and another when the relationship is a temporal one rather than a purely structural “part of” one. This is needed for polycistronic genes.（用于消除一个特征与另一个特征之间的关系，当关系是一个时间的关系，而不是纯粹的结构“一部分”时。 这是多顺反子基因所必需的。）</span><br><span class="line">Note： 备注</span><br><span class="line">Dbxref ：数据库索引</span><br><span class="line">Ontology_term： A cross reference to an ontology term.</span><br></pre></td></tr></table></figure>
<p>下面是一个简单的实例：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br></pre></td><td class="code"><pre><span class="line">##gff-version 3</span><br><span class="line">##sequence-region ctg123 1 1497228</span><br><span class="line">ctg123 . gene 1000 9000 . + . ID&#x3D;gene00001;Name&#x3D;EDEN</span><br><span class="line">ctg123 . TF_binding_site 1000 1012 . + . Parent&#x3D;gene00001</span><br><span class="line">ctg123 . mRNA 1050 9000 . + . ID&#x3D;mRNA00001;Parent&#x3D;gene00001</span><br><span class="line">ctg123 . mRNA 1050 9000 . + . ID&#x3D;mRNA00002;Parent&#x3D;gene00001 ctg123 . mRNA 1300 9000 . + . ID&#x3D;mRNA00003;Parent&#x3D;gene00001 ctg123 . exon 1300 1500 . + . Parent&#x3D;mRNA00003</span><br><span class="line">ctg123 . exon 1050 1500 . + . Parent&#x3D;mRNA00001,mRNA00002</span><br><span class="line">ctg123 . exon 3000 3902 . + . Parent&#x3D;mRNA00001,mRNA00003</span><br><span class="line">ctg123 . exon 5000 5500 . + . Parent&#x3D;mRNA00001,mRNA00002,mRNA00003</span><br><span class="line">ctg123 . exon 7000 9000 . + . Parent&#x3D;mRNA00001,mRNA00002,mRNA00003</span><br><span class="line">ctg123 . CDS 1201 1500 . + 0 ID&#x3D;cds00001;Parent&#x3D;mRNA00001</span><br><span class="line">ctg123 . CDS 3000 3902 . + 0 ID&#x3D;cds00001;Parent&#x3D;mRNA00001</span><br><span class="line">ctg123 . CDS 5000 5500 . + 0 ID&#x3D;cds00001;Parent&#x3D;mRNA00001</span><br><span class="line">ctg123 . CDS 7000 7600 . + 0 ID&#x3D;cds00001;Parent&#x3D;mRNA00001</span><br><span class="line">ctg123 . CDS 1201 1500 . + 0 ID&#x3D;cds00002;Parent&#x3D;mRNA00002</span><br><span class="line">ctg123 . CDS 5000 5500 . + 0 ID&#x3D;cds00002;Parent&#x3D;mRNA00002</span><br><span class="line">ctg123 . CDS 7000 7600 . + 0 ID&#x3D;cds00002;Parent&#x3D;mRNA00002</span><br><span class="line">ctg123 . CDS 3301 3902 . + 0 ID&#x3D;cds00003;Parent&#x3D;mRNA00003</span><br><span class="line">ctg123 . CDS 5000 5500 . + 2 ID&#x3D;cds00003;Parent&#x3D;mRNA00003</span><br><span class="line">ctg123 . CDS 7000 7600 . + 2 ID&#x3D;cds00003;Parent&#x3D;mRNA00003</span><br><span class="line">ctg123 . CDS 3391 3902 . + 0 ID&#x3D;cds00004;Parent&#x3D;mRNA00003</span><br><span class="line">ctg123 . CDS 5000 5500 . + 2 ID&#x3D;cds00004;Parent&#x3D;mRNA00003</span><br><span class="line">ctg123 . CDS 7000 7600 . + 2 ID&#x3D;cds00004;Parent&#x3D;mRNA00003</span><br></pre></td></tr></table></figure>
<p>可以看到第9列其实是可以无限扩展的，只需要以封号进行分割即可。</p>
<p><strong>二、GTF 与GFF的差异</strong></p>
<p>GTF文件以及GFF文件都由9列数据组成，这两种文件的前8列都是相同的（一些小的差别），它们的差别重点在第9列。</p>
<p>GTF文件的第9列同GFF文件不同，虽然同样是标签与值配对的情况，但标签与值之间以空格分开,而不是GFF里面的=号 且每个特征之后都要有分号，（包括最后一个特征）. 下面看一个GTF的实例：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">17	havana	five_prime_utr	7687377	7687427	.	-	.	gene_id &quot;ENSG00000141510&quot;; gene_version &quot;16&quot;; transcript_id &quot;ENST00000503591&quot;; transcript_version &quot;1&quot;; gene_name &quot;TP53&quot;; gene_source &quot;ensembl_havana&quot;; gene_biotype &quot;protein_coding&quot;; havana_gene &quot;OTTHUMG00000162125&quot;; havana_gene_version &quot;10&quot;; transcript_name &quot;TP53-003&quot;; transcript_source &quot;havana&quot;; transcript_biotype &quot;protein_coding&quot;; havana_transcript &quot;OTTHUMT00000367399&quot;; havana_transcript_version &quot;2&quot;; tag &quot;cds_end_NF&quot;; tag &quot;mRNA_end_NF&quot;; transcript_support_level &quot;5&quot;;</span><br><span class="line">17	havana	five_prime_utr	7677325	7677427	.	-	.	gene_id &quot;ENSG00000141510&quot;; gene_version &quot;16&quot;; transcript_id &quot;ENST00000503591&quot;; transcript_version &quot;1&quot;; gene_name &quot;TP53&quot;; gene_source &quot;ensembl_havana&quot;; gene_biotype &quot;protein_coding&quot;; havana_gene &quot;OTTHUMG00000162125&quot;; havana_gene_version &quot;10&quot;; transcript_name &quot;TP53-003&quot;; transcript_source &quot;havana&quot;; transcript_biotype &quot;protein_coding&quot;; havana_transcript &quot;OTTHUMT00000367399&quot;; havana_transcript_version &quot;2&quot;; tag &quot;cds_end_NF&quot;; tag &quot;mRNA_end_NF&quot;; transcript_support_level &quot;5&quot;;</span><br><span class="line">17	havana	five_prime_utr	7676595	7676622	.	-	.	gene_id &quot;ENSG00000141510&quot;; gene_version &quot;16&quot;; transcript_id &quot;ENST00000503591&quot;; transcript_version &quot;1&quot;; gene_name &quot;TP53&quot;; gene_source &quot;ensembl_havana&quot;; gene_biotype &quot;protein_coding&quot;; havana_gene &quot;OTTHUMG00000162125&quot;; havana_gene_version &quot;10&quot;; transcript_name &quot;TP53-003&quot;; transcript_source &quot;havana&quot;; transcript_biotype &quot;protein_coding&quot;; havana_transcript &quot;OTTHUMT00000367399&quot;; havana_transcript_version &quot;2&quot;; tag &quot;cds_end_NF&quot;; tag &quot;mRNA_end_NF&quot;; transcript_support_level &quot;5&quot;;</span><br></pre></td></tr></table></figure>
<h2 id="GenePred">GenePred</h2>
<p><strong>简介</strong></p>
<p>GenePred格式主要用于基因浏览器中基因预测的track。 如果有可变剪切的情况，那表格的每一行就是一个 transcript 的全部信息。</p>
<p><strong>一、定义和示例</strong></p>
<p>每一行的具体解释如下</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line">table genePred</span><br><span class="line">&quot;A gene prediction.&quot;</span><br><span class="line">    (</span><br><span class="line">    string  name;               &quot;Name of gene&quot;</span><br><span class="line">    string  chrom;              &quot;Chromosome name&quot;</span><br><span class="line">    char[1] strand;             &quot;+ or - for strand&quot;</span><br><span class="line">    uint    txStart;            &quot;Transcription start position&quot;</span><br><span class="line">    uint    txEnd;              &quot;Transcription end position&quot;</span><br><span class="line">    uint    cdsStart;           &quot;Coding region start&quot;</span><br><span class="line">    uint    cdsEnd;             &quot;Coding region end&quot;</span><br><span class="line">    uint    exonCount;          &quot;Number of exons&quot;</span><br><span class="line">    uint[exonCount] exonStarts; &quot;Exon start positions&quot;</span><br><span class="line">    uint[exonCount] exonEnds;   &quot;Exon end positions&quot;</span><br><span class="line">    )</span><br></pre></td></tr></table></figure>
<p>如果觉得抽象，我们可以用示例来进行一下对比。小编在这里首先将模式植物拟南芥的gtf文件转化为gpd格式。<code>head -n 1</code> 看一下gpd文件第一行的样式。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">#gpd</span><br><span class="line">AT1G01010.1	Chr1	+	3630	5899	3759	5630	6	3630,3995,4485,4705,5173,5438,	3913,4276,4605,5095,5326,5899</span><br></pre></td></tr></table></figure>
<p>再来<code>grep</code>一下gtf文件中有<strong>AT1G01010.1</strong>信息的那些行是什么样</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">#gtf</span><br><span class="line">Chr1	Araport11	5UTR	3631	3759	.	+	.	transcript_id &quot;AT1G01010.1&quot;; gene_id &quot;AT1G01010&quot;</span><br><span class="line">Chr1	Araport11	exon	3631	3913	.	+	.	transcript_id &quot;AT1G01010.1&quot;; gene_id &quot;AT1G01010&quot;</span><br><span class="line">Chr1	Araport11	start_codon	3760	3762	.	+	.	transcript_id &quot;AT1G01010.1&quot;; gene_id &quot;AT1G01010&quot;</span><br><span class="line">Chr1	Araport11	CDS	3760	3913	.	+	0	transcript_id &quot;AT1G01010.1&quot;; gene_id &quot;AT1G01010&quot;</span><br><span class="line">Chr1	Araport11	exon	3996	4276	.	+	.	transcript_id &quot;AT1G01010.1&quot;; gene_id &quot;AT1G01010&quot;</span><br><span class="line">Chr1	Araport11	CDS	3996	4276	.	+	2	transcript_id &quot;AT1G01010.1&quot;; gene_id &quot;AT1G01010&quot;</span><br><span class="line">Chr1	Araport11	exon	4486	4605	.	+	.	transcript_id &quot;AT1G01010.1&quot;; gene_id &quot;AT1G01010&quot;</span><br><span class="line">Chr1	Araport11	CDS	4486	4605	.	+	0	transcript_id &quot;AT1G01010.1&quot;; gene_id &quot;AT1G01010&quot;</span><br><span class="line">Chr1	Araport11	exon	4706	5095	.	+	.	transcript_id &quot;AT1G01010.1&quot;; gene_id &quot;AT1G01010&quot;</span><br><span class="line">Chr1	Araport11	CDS	4706	5095	.	+	0	transcript_id &quot;AT1G01010.1&quot;; gene_id &quot;AT1G01010&quot;</span><br><span class="line">Chr1	Araport11	exon	5174	5326	.	+	.	transcript_id &quot;AT1G01010.1&quot;; gene_id &quot;AT1G01010&quot;</span><br><span class="line">Chr1	Araport11	CDS	5174	5326	.	+	0	transcript_id &quot;AT1G01010.1&quot;; gene_id &quot;AT1G01010&quot;</span><br><span class="line">Chr1	Araport11	CDS	5439	5630	.	+	0	transcript_id &quot;AT1G01010.1&quot;; gene_id &quot;AT1G01010&quot;</span><br><span class="line">Chr1	Araport11	exon	5439	5899	.	+	.	transcript_id &quot;AT1G01010.1&quot;; gene_id &quot;AT1G01010&quot;</span><br><span class="line">Chr1	Araport11	stop_codon	5628	5630	.	+	.	transcript_id &quot;AT1G01010.1&quot;; gene_id &quot;AT1G01010&quot;</span><br><span class="line">Chr1	Araport11	3UTR	5631	5899	.	+	.	transcript_id &quot;AT1G01010.1&quot;; gene_id &quot;AT1G01010&quot;</span><br></pre></td></tr></table></figure>
<p>我们可以看出，在GTF中，AT1G01010.1这个transcript共有6个CDS，那么对应到相应gpd文件AT1G01010.1这一行的第8列就是6，而第9列和第10列则是这6个CDS对应的起始和终止位置。</p>
<p>细心的朋友可能会发现，GTF文件中CDS起始位置在GenePred table中统统少了1，这其实就是两种文件的起始位置从1开始还是从0开始计数的区别。</p>
<p>GTF的产生和流行有其历史的原因。但是从技术角度来讲，这个文件格式是个非常差劲的格式。</p>
<p>GTF格式非常冗余。以人类转录组为例，Gencode V22的GTF文件为1.2G，压缩之后只有40M。大家知道压缩软件的压缩比是和软件的冗余程度。很少有文件能够压缩到1/30的大小。可见GTF格式多么冗余。GTF格式（及其早期版本GFF等）有很好的替代格式。从信息量上来讲：GTF 等价于 GenePred （Bed12) + Gene_Anno_table。</p>
<p>GenePred是Jimmy Kent创建UCSC genome browser的时候建立的文件格式。UCSC的文件格式定义是非常smart的，包括之后我可能会讲到的2bit，bigwig格式。</p>
<p><strong>二、GTF vs GenePred</strong></p>
<ul>
<li>从文件大小上来看，压缩前：GTF（1.2G) &gt;&gt; Genepred(23M) + Gene_Anno_table (2.8M)。压缩后：GTF(40M) &gt;&gt; GenePred(7.8M) +Gene_Anno_table (580K)</li>
<li>从可读性来讲，GTF是以gene interval 为单位（行），每行可以是gene，transcript，exon，intron，utr等各种信息，看起来什么都在里面，很全面，其实可读性非常差，而且容易产生各种错误。GenePred格式是以transcript为单位，每行就是一个transcript，简洁直观。</li>
<li>从程序处理的角度来讲：以GTF文件作为输入的程序，如果换成以GenePred格式为输入，编程的难度会降低一个数量级，运算时间会快很多，代码的可读性强很多。</li>
</ul>
<p>GTF 转换成GenePred：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">gtfToGenePred -genePredExt -ignoreGroupsWithoutExons -geneNameAsName2 test.gtf test.gpd</span><br></pre></td></tr></table></figure>
<p>Gene_Anno_table: 其实就是把GTF的所有transcript行的第9列转换变成一个表格。</p>
<h2 id="BED">BED</h2>
<p><strong>简介</strong></p>
<p>BED 文件格式提供了一种灵活的方式来定义的数据行，用于描述注释的信息。 跟GTF/GFF格式一样，也可以用来描述基因组特征。但没有GTF/GFF格式那么正规，通常用来描述 任何人为定义的区间。 但没有GTF/GFF格式那么正规，通常用来描述任何人为定义的区间。 所以BED格式最重要的就是染色体加上起始终止坐标这3列。</p>
<p><strong>一、定义和示例</strong></p>
<p>BED行有3个必须的列和9个额外可选的列。 每行的数据格式要求一致。</p>
<p><strong>1、必须包含的3列是：</strong></p>
<ol>
<li>chrom, 染色体或scafflold 的名字(eg chr3， chrY, chr2_random, scaffold0671 )</li>
<li>chromStart 染色体或scaffold的起始位置，染色体第一个碱基的位置是0</li>
<li>chromEnd 染色体或scaffold的结束位置，染色体的末端位置没有包含到显示信息里面。例如，首先得100个碱基的染色体定义为chromStart =0 . chromEnd=100, 碱基的数目是0-99</li>
</ol>
<p><strong>2、9 个额外的可选列:</strong></p>
<ol>
<li>name 指定BED行的名字，这个名字标签会展示在基因组浏览器中的bed行的左侧。</li>
<li>score 0到1000的分值，如果在注释数据的设定中将原始基线设置为１，那么这个分值会决定现示灰度水平（数字越大，灰度越高）</li>
<li>strand 定义链的方向，‘’+” 或者”-”</li>
<li>thickStart 起始位置（The starting position at which the feature is drawn thickly）(例如，基因起始编码位置）</li>
<li>thickEnd 终止位置（The ending position at which the feature is drawn thickly）（例如：基因终止编码位置）</li>
<li>itemRGB 是一个RGB值的形式, R, G, B (eg. 255, 0, 0), 如果itemRgb设置为’On”, 这个RBG值将决定数据的显示的颜色。</li>
<li>blockCount BED行中的block数目，也就是外显子数目</li>
<li>blockSize 用逗号分割的外显子的大小,这个item的数目对应于BlockCount的数目</li>
<li>blockStarts- 用逗号分割的列表, 所有外显子的起始位置，数目也与blockCount数目对应.</li>
</ol>
<p><strong>3、一个简单的示例如下：</strong></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">track name&#x3D;pairedReads description&#x3D;&quot;Clone Paired Reads&quot; useScore&#x3D;1</span><br><span class="line">chr22 1000 5000 cloneA 960 + 1000 5000 0 2 567,488, 0,3512</span><br><span class="line">chr22 2000 6000 cloneB 900 - 2000 6000 0 2 433,399, 0,3601</span><br></pre></td></tr></table></figure>
<p>bed格式有相应的软件来处理这类格式的文件，如bedtools。</p>
<ul>
<li>
<p>注意：用于在GBrowse上展示相关注释的bed格式通常第一行有一个关于track的描述信息。</p>
<p><strong>速记：</strong></p>
</li>
<li>
<p>bed不是床，缺了主要3列就得黄~</p>
</li>
<li>
<p>9列可选列，看了不会胡略略~</p>
</li>
</ul>
<p><strong>二、BED 与GFF的差异</strong></p>
<p>BED文件中起始坐标为0，结束坐标至少是1,</p>
<p>GFF中起始坐标是1而结束坐标至少是1。</p>
<h2 id="MAF">MAF</h2>
<p><strong>简介</strong></p>
<p>MAF格式通常用于记录体细胞突变。 MAF本来并不是一个常见的文本文件格式，只是因为癌症研究实在是太热门了，对它的理解也变得需求旺盛起来了。</p>
<p><strong>一、MAF的说明</strong></p>
<p>这些文件应该使用下面描述的突变注释格式（MAF）进行格式化。另外下文中有文件命名规范。</p>
<p>以下几种类型的体细胞突变会在MAF文件中出现：</p>
<ul>
<li>错义突变及无义突变 *剪接位点，其定义为剪接位点2 bp以内的SNP *沉默突变 *与基因的编码区、剪接位点或遗传元件目标区域重叠的引物。 *移码突变 *调控区突变</li>
</ul>
<p>大部分MAF提交提交的是原始数据。这些原始数据中在体细胞中标记的位点与已知的变异类型相重合的。为避免有可能出现的细胞系污染，MAF规定了一定的下细胞过滤标准。根据现行政策，可开放获取MAF资料应满足：</p>
<ul>
<li>包括所有已验证的体细胞突变名称 *包括与编码区域或剪接位点重叠的所有未验证的体细胞突变名称 *排除所有其他类型的突变（即非体细胞突变、不在编码区域或剪接位点的未验证体的细胞突变以及未在dbSNP、COSMIC或OMIM中注释为体细胞的dbSNP位点）</li>
</ul>
<p>我们提交给DCC MAF存档的数据包括两种：Somatic MAF（named .somatic.maf）的开放访问数据以及不经过筛选的包含原始数据的Protected MAF（named.protected.maf）。所有数据将使用MAF标准进行格式化。</p>
<p><strong>二、MAF文件字段</strong></p>
<p>MAF文件的格式是以制表符分隔的列。这些列都在表1中进行了详细注释，每个MAF文件中都必须按照相应格式进行处理，DCC将验证每列的顺序是否符合标注。每列的标题和值有时需要区分大小写。列中允许出现空值（即空白单元格）或枚举值。验证器将查找是否存在一个符合规范（例如#version 2.4）的标题，如果没有，验证将会失败。除了出现在表1中的列外，也可以选择附加其他列。可选列不经过DCC验证，可以按任何顺序进行。</p>
<p>MAF文件可能有两种格式 ，可能是47列，或者120列，第一行一般都是 头文件，注释着每一列的信息，的确，信息量有点略大。如下：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br><span class="line">109</span><br><span class="line">110</span><br><span class="line">111</span><br><span class="line">112</span><br><span class="line">113</span><br><span class="line">114</span><br><span class="line">115</span><br><span class="line">116</span><br><span class="line">117</span><br><span class="line">118</span><br><span class="line">119</span><br><span class="line">120</span><br><span class="line">121</span><br><span class="line">122</span><br><span class="line">123</span><br><span class="line">124</span><br><span class="line">125</span><br><span class="line">126</span><br><span class="line">127</span><br><span class="line">128</span><br><span class="line">129</span><br><span class="line">130</span><br><span class="line">131</span><br><span class="line">132</span><br><span class="line">133</span><br><span class="line">134</span><br><span class="line">135</span><br><span class="line">136</span><br><span class="line">137</span><br><span class="line">138</span><br><span class="line">139</span><br><span class="line">140</span><br><span class="line">141</span><br><span class="line">142</span><br><span class="line">143</span><br><span class="line">144</span><br><span class="line">145</span><br><span class="line">146</span><br><span class="line">147</span><br><span class="line">148</span><br><span class="line">149</span><br><span class="line">150</span><br><span class="line">151</span><br><span class="line">152</span><br><span class="line">153</span><br><span class="line">154</span><br><span class="line">155</span><br><span class="line">156</span><br><span class="line">157</span><br><span class="line">158</span><br><span class="line">159</span><br><span class="line">160</span><br><span class="line">161</span><br><span class="line">162</span><br><span class="line">163</span><br><span class="line">164</span><br><span class="line">165</span><br><span class="line">166</span><br><span class="line">167</span><br><span class="line">168</span><br><span class="line">169</span><br><span class="line">170</span><br><span class="line">171</span><br><span class="line">172</span><br></pre></td><td class="code"><pre><span class="line">     1	Hugo_Symbol</span><br><span class="line">     2	Entrez_Gene_Id</span><br><span class="line">     3	Center</span><br><span class="line">     4	NCBI_Build</span><br><span class="line">     5	Chromosome</span><br><span class="line">     6	Start_Position</span><br><span class="line">     7	End_Position</span><br><span class="line">     8	Strand</span><br><span class="line">     9	Consequence</span><br><span class="line">    10	Variant_Classification</span><br><span class="line">    11	Variant_Type</span><br><span class="line">    12	Reference_Allele</span><br><span class="line">    13	Tumor_Seq_Allele1</span><br><span class="line">    14	Tumor_Seq_Allele2</span><br><span class="line">    15	dbSNP_RS</span><br><span class="line">    16	dbSNP_Val_Status</span><br><span class="line">    17	Tumor_Sample_Barcode</span><br><span class="line">    18	Matched_Norm_Sample_Barcode</span><br><span class="line">    19	Match_Norm_Seq_Allele1</span><br><span class="line">    20	Match_Norm_Seq_Allele2</span><br><span class="line">    21	Tumor_Validation_Allele1</span><br><span class="line">    22	Tumor_Validation_Allele2</span><br><span class="line">    23	Match_Norm_Validation_Allele1</span><br><span class="line">    24	Match_Norm_Validation_Allele2</span><br><span class="line">    25	Verification_Status</span><br><span class="line">    26	Validation_Status</span><br><span class="line">    27	Mutation_Status</span><br><span class="line">    28	Sequencing_Phase</span><br><span class="line">    29	Sequence_Source</span><br><span class="line">    30	Validation_Method</span><br><span class="line">    31	Score</span><br><span class="line">    32	BAM_File</span><br><span class="line">    33	Sequencer</span><br><span class="line">    34	t_ref_count</span><br><span class="line">    35	t_alt_count</span><br><span class="line">    36	n_ref_count</span><br><span class="line">    37	n_alt_count</span><br><span class="line">    38	HGVSc</span><br><span class="line">    39	HGVSp</span><br><span class="line">    40	HGVSp_Short</span><br><span class="line">    41	Transcript_ID</span><br><span class="line">    42	RefSeq</span><br><span class="line">    43	Protein_position</span><br><span class="line">    44	Codons</span><br><span class="line">    45	Hotspot</span><br><span class="line">    46	cDNA_change</span><br><span class="line">    47	Amino_Acid_Change</span><br><span class="line">     1	Hugo_Symbol</span><br><span class="line">     2	Entrez_Gene_Id</span><br><span class="line">     3	Center</span><br><span class="line">     4	NCBI_Build</span><br><span class="line">     5	Chromosome</span><br><span class="line">     6	Start_Position</span><br><span class="line">     7	End_Position</span><br><span class="line">     8	Strand</span><br><span class="line">     9	Variant_Classification</span><br><span class="line">    10	Variant_Type</span><br><span class="line">    11	Reference_Allele</span><br><span class="line">    12	Tumor_Seq_Allele1</span><br><span class="line">    13	Tumor_Seq_Allele2</span><br><span class="line">    14	dbSNP_RS</span><br><span class="line">    15	dbSNP_Val_Status</span><br><span class="line">    16	Tumor_Sample_Barcode</span><br><span class="line">    17	Matched_Norm_Sample_Barcode</span><br><span class="line">    18	Match_Norm_Seq_Allele1</span><br><span class="line">    19	Match_Norm_Seq_Allele2</span><br><span class="line">    20	Tumor_Validation_Allele1</span><br><span class="line">    21	Tumor_Validation_Allele2</span><br><span class="line">    22	Match_Norm_Validation_Allele1</span><br><span class="line">    23	Match_Norm_Validation_Allele2</span><br><span class="line">    24	Verification_Status</span><br><span class="line">    25	Validation_Status</span><br><span class="line">    26	Mutation_Status</span><br><span class="line">    27	Sequencing_Phase</span><br><span class="line">    28	Sequence_Source</span><br><span class="line">    29	Validation_Method</span><br><span class="line">    30	Score</span><br><span class="line">    31	BAM_File</span><br><span class="line">    32	Sequencer</span><br><span class="line">    33	Tumor_Sample_UUID</span><br><span class="line">    34	Matched_Norm_Sample_UUID</span><br><span class="line">    35	HGVSc</span><br><span class="line">    36	HGVSp</span><br><span class="line">    37	HGVSp_Short</span><br><span class="line">    38	Transcript_ID</span><br><span class="line">    39	Exon_Number</span><br><span class="line">    40	t_depth</span><br><span class="line">    41	t_ref_count</span><br><span class="line">    42	t_alt_count</span><br><span class="line">    43	n_depth</span><br><span class="line">    44	n_ref_count</span><br><span class="line">    45	n_alt_count</span><br><span class="line">    46	all_effects</span><br><span class="line">    47	Allele</span><br><span class="line">    48	Gene</span><br><span class="line">    49	Feature</span><br><span class="line">    50	Feature_type</span><br><span class="line">    51	One_Consequence</span><br><span class="line">    52	Consequence</span><br><span class="line">    53	cDNA_position</span><br><span class="line">    54	CDS_position</span><br><span class="line">    55	Protein_position</span><br><span class="line">    56	Amino_acids</span><br><span class="line">    57	Codons</span><br><span class="line">    58	Existing_variation</span><br><span class="line">    59	ALLELE_NUM</span><br><span class="line">    60	DISTANCE</span><br><span class="line">    61	TRANSCRIPT_STRAND</span><br><span class="line">    62	SYMBOL</span><br><span class="line">    63	SYMBOL_SOURCE</span><br><span class="line">    64	HGNC_ID</span><br><span class="line">    65	BIOTYPE</span><br><span class="line">    66	CANONICAL</span><br><span class="line">    67	CCDS</span><br><span class="line">    68	ENSP</span><br><span class="line">    69	SWISSPROT</span><br><span class="line">    70	TREMBL</span><br><span class="line">    71	UNIPARC</span><br><span class="line">    72	RefSeq</span><br><span class="line">    73	SIFT</span><br><span class="line">    74	PolyPhen</span><br><span class="line">    75	EXON</span><br><span class="line">    76	INTRON</span><br><span class="line">    77	DOMAINS</span><br><span class="line">    78	GMAF</span><br><span class="line">    79	AFR_MAF</span><br><span class="line">    80	AMR_MAF</span><br><span class="line">    81	ASN_MAF</span><br><span class="line">    82	EAS_MAF</span><br><span class="line">    83	EUR_MAF</span><br><span class="line">    84	SAS_MAF</span><br><span class="line">    85	AA_MAF</span><br><span class="line">    86	EA_MAF</span><br><span class="line">    87	CLIN_SIG</span><br><span class="line">    88	SOMATIC</span><br><span class="line">    89	PUBMED</span><br><span class="line">    90	MOTIF_NAME</span><br><span class="line">    91	MOTIF_POS</span><br><span class="line">    92	HIGH_INF_POS</span><br><span class="line">    93	MOTIF_SCORE_CHANGE</span><br><span class="line">    94	IMPACT</span><br><span class="line">    95	PICK</span><br><span class="line">    96	VARIANT_CLASS</span><br><span class="line">    97	TSL</span><br><span class="line">    98	HGVS_OFFSET</span><br><span class="line">    99	PHENO</span><br><span class="line">   100	MINIMISED</span><br><span class="line">   101	ExAC_AF</span><br><span class="line">   102	ExAC_AF_Adj</span><br><span class="line">   103	ExAC_AF_AFR</span><br><span class="line">   104	ExAC_AF_AMR</span><br><span class="line">   105	ExAC_AF_EAS</span><br><span class="line">   106	ExAC_AF_FIN</span><br><span class="line">   107	ExAC_AF_NFE</span><br><span class="line">   108	ExAC_AF_OTH</span><br><span class="line">   109	ExAC_AF_SAS</span><br><span class="line">   110	GENE_PHENO</span><br><span class="line">   111	FILTER</span><br><span class="line">   112	CONTEXT</span><br><span class="line">   113	src_vcf_id</span><br><span class="line">   114	tumor_bam_uuid</span><br><span class="line">   115	normal_bam_uuid</span><br><span class="line">   116	case_id</span><br><span class="line">   117	GDC_FILTER</span><br><span class="line">   118	COSMIC</span><br><span class="line">   119	MC3_Overlap</span><br><span class="line">   120	GDC_Validation_Status</span><br><span class="line">重要标准</span><br><span class="line">列表中每列的顺序最好与索引列相同。</span><br><span class="line">标有“Case Sensitive“的列所有标题都需要区分大小写。</span><br><span class="line">标有”Null“的列表示允许具有空值。</span><br><span class="line">标有“Enumerated column”的列表示具有指定的值，比如“Enumerated value” 是&quot;No&quot;表示该列没有指定的值；其他值表示允许列出的具体值; “Set”表示该列的值来自指定的一组已知值（例如HUGO基因符号）</span><br></pre></td></tr></table></figure>
<p><strong>三、MAF文件检查</strong></p>
<p>DCC 档案验证器将检查MAF文件的完整性。如果MAF文件中的任何一项出现错误，验证将失败：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br></pre></td><td class="code"><pre><span class="line">1.	列标题文本（包括大小写）和顺序必须与表1完全一致</span><br><span class="line">2.	表1中列出的列标题下的值不是空值时必须具有相应的值</span><br><span class="line">3.	表1中指定为“Case Sensitive”的值必须区分大小写。</span><br><span class="line">4.	如果列标题在规范中列出为具有枚举值（即“Enumerated”列的“Yes”），则这些列中的值必须来自“Enumerated”下列出的枚举值。</span><br><span class="line">5.	如果规范中列标题具有设置值（即“Enumerated”列的“Set”），则那些列下的值必须来自该域的枚举值（例如，HUGO基因符号）。</span><br><span class="line">6.	所有Allele-based列必须包含 –(deletion)或由以下大写字母组成的字符串：A，T，G，C。</span><br><span class="line">7.	如果Validation_Status &#x3D;&#x3D; &quot;Untested&quot; 那么Tumor_Validation_Allele1, Tumor_Validation_Allele2, Match_Norm_Validation_Allele1, Match_Norm_Validation_Allele2 可以是空值(由 Validation_Status决定).</span><br><span class="line">a)	如果Validation_Status &#x3D;&#x3D; &quot;Inconclusive&quot; 那么Tumor_Validation_Allele1, Tumor_Validation_Allele2, Match_Norm_Validation_Allele1, Match_Norm_Validation_Allele2 可以是空值(由 Validation_Status决定)</span><br><span class="line">8.	如果Validation_Status &#x3D;&#x3D; Valid, 那么Validated_Tumor_Allele1 and Validated_Tumor_Allele2一定要是 A, C, G, T, 或“-“中的一种</span><br><span class="line"></span><br><span class="line">a)	如果Validation_Status  &#x3D;&#x3D; &quot;Valid&quot; 那么Tumor_Validation_Allele1, Tumor_Validation_Allele2, Match_Norm_Validation_Allele1, Match_Norm_Validation_Allele2 不可以是空值</span><br><span class="line"></span><br><span class="line">b)	 如果Validation_Status &#x3D;&#x3D; &quot;Invalid&quot; 那么Tumor_Validation_Allele1, Tumor_Validation_Allele2, Match_Norm_Validation_Allele1, Match_Norm_Validation_Allele2 不可以是空值Tumor_Validation_Allelle1 &#x3D;&#x3D; Match_Norm_Validation_Allele1 </span><br><span class="line">Tumor_Validation_Allelle2 &#x3D;&#x3D; Match_Norm_Validation_Allele2  (出现错误时，增加以替代8a)</span><br><span class="line"></span><br><span class="line">9.	检查Mutation_Status的等位基因值:</span><br><span class="line">检查Validation_status的等位基因值:</span><br><span class="line"></span><br><span class="line">a)	如果Mutation_Status &#x3D;&#x3D; &quot;Germline&quot; and Validation_Status &#x3D;&#x3D; &quot;Valid&quot;, 那么Tumor_Validation_Allele1 &#x3D;&#x3D; Match_Norm_Validation_Allele1 Tumor_Validation_Allele2 &#x3D;&#x3D; Match_Norm_Validation_Allele2.</span><br><span class="line"></span><br><span class="line">b)	如果Mutation_Status &#x3D;&#x3D; &quot;Somatic&quot; ，Validation_Status &#x3D;&#x3D; &quot;Valid&quot;, 那么Match_Norm_Validation_Allele1 &#x3D;&#x3D; Match_Norm_Validation_Allele2 &#x3D;&#x3D; Reference_Allele 且(Tumor_Validation_Allele1 or Tumor_Validation_Allele2) !&#x3D; Reference_Allele</span><br><span class="line"></span><br><span class="line">c)	如果Mutation_Status &#x3D;&#x3D; &quot;LOH&quot; and Validation_Status&#x3D;&#x3D;&quot;Valid&quot;, 那么Tumor_Validation_Allele1 &#x3D;&#x3D; Tumor_Validation_Allele2 Match_Norm_Validation_Allele1 !&#x3D; Match_Norm_Validation_Allele2 and Tumor_Validation_Allele1 &#x3D;&#x3D; (Match_Norm_Validation_Allele1 or Match_Norm_Validation_Allele2).</span><br><span class="line">10.	检查 Start_position &lt;&#x3D; End_position</span><br><span class="line">11.	根据Variant_Type检查Start_position和End_position: </span><br><span class="line">a)	如果Variant_Type &#x3D;&#x3D; &quot;INS&quot;, 那么(End_position - Start_position + 1 &#x3D;&#x3D; length (Reference_Allele) 或End_position - Start_position &#x3D;&#x3D; 1) 且length(Reference_Allele) &lt;&#x3D; length(Tumor_Seq_Allele1 and Tumor_Seq_Allele2)</span><br><span class="line">b)	如果Variant_Type &#x3D;&#x3D; &quot;DEL&quot;, 那么End_position - Start_position + 1 &#x3D;&#x3D; length (Reference_Allele), </span><br><span class="line">且length(Reference_Allele) &gt;&#x3D; length(Tumor_Seq_Allele1 and Tumor_Seq_Allele2)</span><br><span class="line">c)	如果Variant_Type &#x3D;&#x3D; &quot;SNP&quot;, 那么length(Reference_Allele and Tumor_Seq_Allele1 and Tumor_Seq_Allele2) &#x3D;&#x3D;  1 且(Reference_Allele and Tumor_Seq_Allele1 and Tumor_Seq_Allele2) !&#x3D; &quot;-&quot;</span><br><span class="line">d)	如果Variant_Type &#x3D;&#x3D; &quot;DNP&quot;, 那么length(Reference_Allele 和Tumor_Seq_Allele1 and Tumor_Seq_Allele2) &#x3D;&#x3D;  2 且(Reference_Allele and Tumor_Seq_Allele1 andTumor_Seq_Allele2) !contain &quot;-&quot;</span><br><span class="line">e)	如果Variant_Type &#x3D;&#x3D; &quot;TNP&quot;, 那么length(Reference_Allele and Tumor_Seq_Allele1 and Tumor_Seq_Allele2) &#x3D;&#x3D;  3 且(Reference_Allele and Tumor_Seq_Allele1 and Tumor_Seq_Allele2) !contain &quot;-&quot;</span><br><span class="line">f)	如果Variant_Type &#x3D;&#x3D; &quot;ONP&quot;, 那么length(Reference_Allele) &#x3D;&#x3D; length(Tumor_Seq_Allele1) &#x3D;&#x3D; length(Tumor_Seq_Allele2) &gt; 3 且(Reference_Allele and Tumor_Seq_Allele1 and Tumor_Seq_Allele2) !contain &quot;-&quot;</span><br><span class="line">12.	基于UUID的文件的验证: </span><br><span class="line">a)	列＃33必须包含肿瘤样本的BCR等分试样的UUID的Tumor_Sample_UUID</span><br><span class="line">b)	列＃34必须是Matched_Norm_Sample_UUID，其中包含用于匹配正常样本的BCR等份的UUID</span><br><span class="line">c)	由Tumor_Sample_Barcode和Matched_Norm_Sample_Barcode表示的元数据应分别对应于分配给Tumor_Sample_UUID和Matched_Norm_Sample_UUID的UUID</span><br></pre></td></tr></table></figure>
<p><strong>三、MAF命名协议</strong></p>
<p>在上传到DCC的档案中，MAF文件名应与以下方式与包含档案名称相关： 如果存档名称是：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">&lt;domain&gt;_&lt;disease_abbrev&gt;.&lt;platform&gt;.Level_2.&lt;serial_index&gt;.&lt;revision&gt;.0.tar.gz</span><br><span class="line">那么具有存档的公开的MAF文件命名依据为：</span><br><span class="line">&lt;domain&gt;_&lt;disease_abbrev&gt;.&lt;platform&gt;.Level_2.&lt;serial_index&gt;[.&lt;optional_tag&gt;].somatic.maf</span><br><span class="line">具有存档的受保护的MAF文件命名依据为：</span><br><span class="line">&lt;domain&gt;_&lt;disease_abbrev&gt;.&lt;platform&gt;.Level_2.&lt;serial_index&gt;[.&lt;optional_tag&gt;].protected.maf</span><br><span class="line">&lt;optional_tag&gt;可能包含字母数字字符、破折号或下划线，但不能有空格或句号。&lt;optional_tag&gt;可以省略。使用&lt;optional_tag&gt;的目的是给出一些简短的注释。</span><br><span class="line">例如，对于文件</span><br><span class="line">genome.wustl.edu_OV.IlluminaGA_DNASeq.Level_2.7.6.0.tar.gz</span><br></pre></td></tr></table></figure>
<p>有效的MAF名称的为：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">genome.wustl.edu_OV.IlluminaGA_DNASeq.Level_2.7.preliminary.somatic.maf</span><br><span class="line">genome.wustl.edu_OV.IlluminaGA_DNASeq.Level_2.7.protected.maf</span><br></pre></td></tr></table></figure>
<h1>Wiggle、BigWig和bedgraph</h1>
<p><strong>简介</strong></p>
<p>Wiggle、BigWig和bedgraph仅仅用于追踪参考基因组的各个区域的覆盖度，测序深度。与sam/bam格式文件不同，bam或者bed格式的文件主要是为了追踪我们的reads到底比对到了参考基因组的哪些区域。注意这几者的差别。 Wiggle、BigWig和bedgraph均由UCSC规定的文件格式，可以无缝连接到UCSC的Genome Browser工具里面进行可视化。 wig和bigWig文件的优势在于可以体现出数据大小的变化和高低，例如组蛋白修饰的峰值等。</p>
<p><strong>一、定义</strong></p>
<p>**Bigwig：**简写为bw，它规定了全基因组数据的每个坐标区间的测序深度，bigWig是通过wig格式的文件转换的二进制压缩文件，是一种全基因组计算或实验信号数据的压缩的，索引的二进制格式，使用该格式更加节省空间。 <strong>Wiggle</strong>：简写为wig，表示基因组上一个区域的信号，可以上传至UCSC上进行可视化。而wig文件也是和BED文件类似的包含区域信息的文件，一般使用MACS峰值探测后可以产生wig格式的文件。</p>
<p><strong>二、Wiggle文件示例</strong></p>
<p>Wig是一种比较老的格式，展示连续值的数据，比如GC百分比，转录组数据等，Wig的数据被压缩的，储存在128个独特的仓中，当导出为其他格式时，会有很小的损失。</p>
<p>在UCSC中具体使用哪种数据格式，细节见：<a target="_blank" rel="noopener" href="http://genomewiki.ucsc.edu/index.php/Selecting_a_graphing_track_data_format%E3%80%82%E4%BB%A5UCSC%E7%BB%99%E7%9A%84Wig%E5%8F%82%E8%80%83%E6%96%87%E4%BB%B6%E4%B8%BA%E4%BE%8B%EF%BC%8CWig%E7%9A%84%E6%95%B0%E6%8D%AE%E6%98%AF%E9%9D%A2%E5%90%91%E8%A1%8C%E7%9A%84%EF%BC%8C%E7%AC%AC%E4%B8%80%E8%A1%8C%E5%AE%9A%E4%B9%89%E4%BA%86track%E7%9A%84%E5%B1%9E%E6%80%A7%EF%BC%8C%E6%AF%94%E5%A6%82track">http://genomewiki.ucsc.edu/index.php/Selecting_a_graphing_track_data_format。以UCSC给的Wig参考文件为例，Wig的数据是面向行的，第一行定义了track的属性，比如track</a> type=wiggle_0，指定track为Wig track，为默认选项。其余为可选。一般不用管这些参数，除非你已经很熟悉UCSC的Genome Browser工具。下面是wiggle文件示例。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br></pre></td><td class="code"><pre><span class="line">track type&#x3D;wiggle_0 name&#x3D;&quot;variableStep&quot; description&#x3D;&quot;variableStep format&quot; visibility&#x3D;full autoScale&#x3D;off viewLimits&#x3D;0.0:25.0 color&#x3D;50,150,255 yLineMark&#x3D;11.76 yLineOnOff&#x3D;on priority&#x3D;10</span><br><span class="line">variableStep chrom&#x3D;chr19 span&#x3D;150</span><br><span class="line">49304701 10.0</span><br><span class="line">49304901 12.5</span><br><span class="line">49305401 15.0</span><br><span class="line">49305601 17.5</span><br><span class="line">49305901 20.0</span><br><span class="line">49306081 17.5</span><br><span class="line">49306301 15.0</span><br><span class="line">49306691 12.5</span><br><span class="line">49307871 10.0</span><br><span class="line">#	200 base wide points graph at every 300 bases, 50 pixel high graph</span><br><span class="line">#	autoScale off and viewing range set to [0:1000]</span><br><span class="line">#	priority &#x3D; 20 positions this as the second graph</span><br><span class="line">#	Note, one-relative coordinate system in use for this format</span><br><span class="line">track type&#x3D;wiggle_0 name&#x3D;&quot;fixedStep&quot; description&#x3D;&quot;fixedStep format&quot; visibility&#x3D;full autoScale&#x3D;off viewLimits&#x3D;0:1000 color&#x3D;0,200,100 maxHeightPixels&#x3D;100:50:20 graphType&#x3D;points priority&#x3D;20</span><br><span class="line">fixedStep chrom&#x3D;chr19 start&#x3D;49307401 step&#x3D;300 span&#x3D;200</span><br><span class="line">1000</span><br><span class="line"> 900</span><br><span class="line"> 800</span><br><span class="line"> 700</span><br><span class="line"> 600</span><br><span class="line"> 500</span><br><span class="line"> 400</span><br><span class="line"> 300</span><br><span class="line"> 200</span><br><span class="line"> 100</span><br></pre></td></tr></table></figure>
<p><strong>三、Wig文件详解</strong></p>
<p>Wig文件主要由两部分格式组成，**variableStep format*<em>和*<em>fixedStep format</em></em>。<strong>variableStep format</strong>以一个声明开始，明确了染色体的序号，跨度(span)。后面跟两列数据，染色体开始的碱基位置，数据的值value（可以理解为覆盖度）。span参数可以将含有相同value的连续碱基包含在一起，使数据更加简洁。如下，variableStep format span=150，包含的第一行数据49304701 10.0表示49304701-49304851有相同的value，为10.0。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">variableStep chrom&#x3D;chr2</span><br><span class="line">300701 12.5</span><br><span class="line">300702 12.5</span><br><span class="line">300703 12.5</span><br><span class="line">300704 12.5</span><br><span class="line">300705 12.5 </span><br></pre></td></tr></table></figure>
<p>等价于：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">variableStep chrom&#x3D;chr2 span&#x3D;5</span><br><span class="line">300701 12.5</span><br></pre></td></tr></table></figure>
<p>都表示在2号染色体300701-300705位置上有相同的value，且value=12.5。</p>
<p>第二部分为<strong>fixedStep format，</strong> 由声明和单列数据组成。声明部分和variableStep format中各变量的意义一样。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">fixedStep chrom&#x3D;chr3 start&#x3D;400601 step&#x3D;100 span&#x3D;5</span><br><span class="line">11</span><br><span class="line">22</span><br><span class="line">33 </span><br></pre></td></tr></table></figure>
<p>表示在3号染色体400601-400605, 400701-400705, and 400801-400805三个区域，value值分别为11, 22, 和 33。Wig中的value值可以是整数，实数，正数或者负数。只有指定的位置有value值，没有制定的位置则没有value，且不会在UCSU Genome Browser中作出图。</p>
<p>其实对我们的bam文件，用samtools软件也可以很容易得到基因组区域的覆盖度和测序深度，比如：</p>
<p>samtools depth -r <strong>chr12:126073855-126073965</strong> Ip.sorted.bam</p>
<blockquote>
<p>chr12 126073855 5</p>
<p>chr12 126073856 15</p>
<p>chr12 126073857 31</p>
<p>chr12 126073858 40</p>
<p>chr12 126073859 44</p>
<p>chr12 126073860 52</p>
<p>…其余省略输出…</p>
</blockquote>
<p>这其实就是wig文件的雏形，但是wig文件会更复杂一点！</p>
<p>首先它不需要第一列了，因为全部是重复字段，只需要在每个染色体的第一行定义好染色体即可。</p>
<p><strong>四、文件转换示例</strong></p>
<p>BigWig格式是wig格式文件的二进制压缩版本，用于密集连续的数据，并在基因组浏览器中进行可视化，是UCSC推荐的一种格式。BigWig文件是由原始的Wig格式通过wigToBigWig工具转换过来， 转换工具命令示例：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"># create the chrom.sizes file for the UCSC database (e.g., hg19).</span><br><span class="line">fetchChromSizes  hg19 &gt; chrSize.txt</span><br><span class="line"># Convert wig to big wig:  </span><br><span class="line">wigToBigWig input.wig chrSize.txt myBigWig.bw</span><br></pre></td></tr></table></figure>
<p>输出的BigWig文件大约比原始Wig文件多占用50%的内存（UCSC数据格式转换代码见：<a target="_blank" rel="noopener" href="http://barcwiki.wi.mit.edu/wiki/SOPs/coordinates%EF%BC%89%E3%80%82%E4%B9%9F%E5%8F%AF%E4%BB%A5%E7%94%B1bedGraph">http://barcwiki.wi.mit.edu/wiki/SOPs/coordinates）。也可以由bedGraph</a> 格式通过bedGraphToBigWig转换，这里不作展开。BigWig的主要优点是仅将显示特定区域所需的部分传输到UCSC，所以对于大数据集的文件，BigWig比Wig文件速度快。BigWig主要存在用户可以获得的网页上(http, https or ftp)，而不在UCSC服务器上，只有当前正在查看的染色体位置所需的文件部分在本地缓存为一个“稀疏文件”。</p>
<p><strong>五、BedGraph文件示例</strong></p>
<p>最后顺便讲讲BedGraph ，它的trace type和Wig文件很像，不过后面的数据和bed文件很类似，ChIPseq数据做完peak calling后的bed文件最短只有三列，染色体序号，染色体起始位置和结束位置。如下图，前面的声明和Wig类似，后面的四列分别表示染色体序号，起始位置，结束位置和value值。相当于为bed文件的延伸格式。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line">browser position chr19:49302001-49304701</span><br><span class="line">browser hide all</span><br><span class="line">browser pack refGene encodeRegions</span><br><span class="line">browser full altGraph</span><br><span class="line">#	300 base wide bar graph, autoScale is on by default &#x3D;&#x3D; graphing</span><br><span class="line">#	limits will dynamically change to always show full range of data</span><br><span class="line">#	in viewing window, priority &#x3D; 20 positions this as the second graph</span><br><span class="line">#	Note, zero-relative, half-open coordinate system in use for bedGraph format</span><br><span class="line">track type&#x3D;bedGraph name&#x3D;&quot;BedGraph Format&quot; description&#x3D;&quot;BedGraph format&quot; visibility&#x3D;full color&#x3D;200,100,0 altColor&#x3D;0,100,200 priority&#x3D;20</span><br><span class="line">chr19 49302000 49302300 -1.0</span><br><span class="line">chr19 49302300 49302600 -0.75</span><br><span class="line">chr19 49302600 49302900 -0.50</span><br><span class="line">chr19 49302900 49303200 -0.25</span><br><span class="line">chr19 49303200 49303500 0.0</span><br><span class="line">chr19 49303500 49303800 0.25</span><br><span class="line">chr19 49303800 49304100 0.50</span><br><span class="line">chr19 49304100 49304400 0.75</span><br><span class="line">chr19 49304400 49304700 1.00</span><br></pre></td></tr></table></figure>
<p><strong>文件格式选择</strong></p>
<p>Wig数据的元素大小必须是一样的。</p>
<p>如果数据大小不一样，应该使用bedGraph格式，如果数据过大，就转换为bigWig。</p>

      </div>
      
        <div class="prev-or-next">
          <div class="post-foot-next">
            
              <a href="/2020/09/28/hemberg-lab5/" target="_self">
                <i class="iconfont icon-chevronleft"></i>
                <span>上一页</span>
              </a>
            
          </div>
          <div class="post-attach">
            <span class="post-pubtime">
              <i class="iconfont icon-updatetime" title="更新时间"></i>
              2020-10-02
            </span>
            
                  <span class="post-tags">
                    <i class="iconfont icon-tags" title="标签"></i>
                    
                    <span class="span--tag">
                      <a href="/tags/NGS/" title="NGS">
                        <b>#</b> NGS
                      </a>
                    </span>
                    
                  </span>
              
          </div>
          <div class="post-foot-prev">
            
              <a href="/2020/10/09/aspera/" target="_self">
                <span>下一页</span>
                <i class="iconfont icon-chevronright"></i>
              </a>
            
          </div>
        </div>
      
    </div>
    
  <div class="post-catalog" id="catalog">
    <div class="title">目录</div>
    <div class="catalog-content">
      <ol class="toc"><li class="toc-item toc-level-2"><a class="toc-link" href="#FASTQ"><span class="toc-text">FASTQ</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#FASTA"><span class="toc-text">FASTA</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#SAM%E6%A0%BC%E5%BC%8F"><span class="toc-text">SAM格式</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#BAM-%E6%A0%BC%E5%BC%8F"><span class="toc-text">BAM 格式</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#VCF"><span class="toc-text">VCF</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#GTF%E5%92%8CGFF"><span class="toc-text">GTF和GFF</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#GenePred"><span class="toc-text">GenePred</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#BED"><span class="toc-text">BED</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#MAF"><span class="toc-text">MAF</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link"><span class="toc-text">Wiggle、BigWig和bedgraph</span></a>
    </div>
  </div>

  
<script src="/js/catalog.js"></script>




    
      <div class="comments-container">
        


  <script src='//unpkg.com/valine/dist/Valine.min.js'></script>

  <div id="vcomments"></div>

  <script>
    new Valine({
      el: '#vcomments',
      appId: 'DdMocNMQxnEQcChwW38b0JVQ-gzGzoHsz',
      appKey: 'ETFXPtdWGEvlSfsH7RwSMzAf',
      placeholder: 'Welcome!',
      avatar: 'retro',
      lang: 'zh-CN'
    })
  </script>

    <style>
      .comments-container .v .vempty {
        display: none!important;
      }
    </style>




      </div>
    
  </div>


        <div class="footer">
  <div class="social">
    <ul>
      
        <li>
          <a title="github" target="_blank" rel="noopener" href="https://github.com/Moonerss">
            <i class="iconfont icon-github"></i>
          </a>
        </li>
      
        <li>
          <a title="rss" href="/atom.xml">
            <i class="iconfont icon-rss"></i>
          </a>
        </li>
      
    </ul>
  </div>
  
    <div class="footer-more">
      <a target="_blank" rel="noopener" href="https://github.com/Moonerss">Copyright © mooner 2020</a>
    </div>
  
    <div class="footer-more">
      <a target="_blank" rel="noopener" href="https://github.com/zchengsite/hexo-theme-oranges">Theme by Oranges | Powered by Hexo</a>
    </div>
  
</div>

      </div>

      <div class="back-to-top hidden">
  <a href="javascript: void(0)">
    <i class="iconfont icon-chevronup"></i>
  </a>
</div>


<script src="/js/backtotop.js"></script>



      
  <div class="search-icon" id="search-icon">
    <a href="javascript: void(0)">
      <i class="iconfont icon-search"></i>
    </a>
  </div>

  <div class="search-overlay hidden">
    <div class="search-content" tabindex="0">
      <div class="search-title">
        <span class="search-icon-input">
          <a href="javascript: void(0)">
            <i class="iconfont icon-search"></i>
          </a>
        </span>
        
          <input type="text" class="search-input" id="search-input" placeholder="Search ...">
        
        <span class="search-close-icon" id="search-close-icon">
          <a href="javascript: void(0)">
            <i class="iconfont icon-close"></i>
          </a>
        </span>
      </div>
      <div class="search-result" id="search-result"></div>
    </div>
  </div>

  <script type="text/javascript">
    var inputArea = document.querySelector("#search-input")
    var searchOverlayArea = document.querySelector(".search-overlay")

    inputArea.onclick = function() {
      getSearchFile()
      this.onclick = null
    }

    inputArea.onkeydown = function() {
      if(event.keyCode == 13)
        return false
    }

    function openOrHideSearchContent() {
      let isHidden = searchOverlayArea.classList.contains('hidden')
      if (isHidden) {
        searchOverlayArea.classList.remove('hidden')
        document.body.classList.add('hidden')
        // inputArea.focus()
      } else {
        searchOverlayArea.classList.add('hidden')
        document.body.classList.remove('hidden')
      }
    }

    function blurSearchContent(e) {
      if (e.target === searchOverlayArea) {
        openOrHideSearchContent()
      }
    }

    document.querySelector("#search-icon").addEventListener("click", openOrHideSearchContent, false)
    document.querySelector("#search-close-icon").addEventListener("click", openOrHideSearchContent, false)
    searchOverlayArea.addEventListener("click", blurSearchContent, false)

    var searchFunc = function (path, search_id, content_id) {
      'use strict';
      var $input = document.getElementById(search_id);
      var $resultContent = document.getElementById(content_id);
      $resultContent.innerHTML = "<ul><span class='local-search-empty'>首次搜索，正在载入索引文件，请稍后……<span></ul>";
      $.ajax({
        // 0x01. load xml file
        url: path,
        dataType: "xml",
        success: function (xmlResponse) {
          // 0x02. parse xml file
          var datas = $("entry", xmlResponse).map(function () {
            return {
              title: $("title", this).text(),
              content: $("content", this).text(),
              url: $("url", this).text()
            };
          }).get();
          $resultContent.innerHTML = "";

          $input.addEventListener('input', function () {
            // 0x03. parse query to keywords list
            var str = '<ul class=\"search-result-list\">';
            var keywords = this.value.trim().toLowerCase().split(/[\s\-]+/);
            $resultContent.innerHTML = "";
            if (this.value.trim().length <= 0) {
              return;
            }
            // 0x04. perform local searching
            datas.forEach(function (data) {
              var isMatch = true;
              var content_index = [];
              if (!data.title || data.title.trim() === '') {
                data.title = "Untitled";
              }
              var orig_data_title = data.title.trim();
              var data_title = orig_data_title.toLowerCase();
              var orig_data_content = data.content.trim().replace(/<[^>]+>/g, "");
              var data_content = orig_data_content.toLowerCase();
              var data_url = data.url;
              var index_title = -1;
              var index_content = -1;
              var first_occur = -1;
              // only match artiles with not empty contents
              if (data_content !== '') {
                keywords.forEach(function (keyword, i) {
                  index_title = data_title.indexOf(keyword);
                  index_content = data_content.indexOf(keyword);

                  if (index_title < 0 && index_content < 0) {
                    isMatch = false;
                  } else {
                    if (index_content < 0) {
                      index_content = 0;
                    }
                    if (i == 0) {
                      first_occur = index_content;
                    }
                    // content_index.push({index_content:index_content, keyword_len:keyword_len});
                  }
                });
              } else {
                isMatch = false;
              }
              // 0x05. show search results
              if (isMatch) {
                str += "<li><a href='" + data_url + "' class='search-result-title'>" + orig_data_title + "</a>";
                var content = orig_data_content;
                if (first_occur >= 0) {
                  // cut out 100 characters
                  var start = first_occur - 20;
                  var end = first_occur + 80;

                  if (start < 0) {
                    start = 0;
                  }

                  if (start == 0) {
                    end = 100;
                  }

                  if (end > content.length) {
                    end = content.length;
                  }

                  var match_content = content.substr(start, end);

                  // highlight all keywords
                  keywords.forEach(function (keyword) {
                    var regS = new RegExp(keyword, "gi");
                    match_content = match_content.replace(regS, "<span class=\"search-keyword\">" + keyword + "</span>");
                  });

                  str += "<p class=\"search-result-abstract\">" + match_content + "...</p>"
                }
                str += "</li>";
              }
            });
            str += "</ul>";
            if (str.indexOf('<li>') === -1) {
              return $resultContent.innerHTML = "<ul><span class='local-search-empty'>没有找到内容，请尝试更换检索词。<span></ul>";
            }
            $resultContent.innerHTML = str;
          });
        },
        error: function(xhr, status, error) {
          $resultContent.innerHTML = ""
          if (xhr.status === 404) {
            $resultContent.innerHTML = "<ul><span class='local-search-empty'>未找到search.xml文件，具体请参考：<a href='https://github.com/zchengsite/hexo-theme-oranges#configuration' target='_black'>configuration</a><span></ul>";
          } else {
            $resultContent.innerHTML = "<ul><span class='local-search-empty'>请求失败，尝试重新刷新页面或稍后重试。<span></ul>";
          }
        }
      });
      $(document).on('click', '#search-close-icon', function() {
        $('#search-input').val('');
        $('#search-result').html('');
      });
    }

    var getSearchFile = function() {
        var path = "/search.xml";
        searchFunc(path, 'search-input', 'search-result');
    }
  </script>




    </div>
  </body>
</html>
