<!DOCTYPE html>
<html lang="zh-CN">

  <head>
  <meta charset="utf-8">
  <meta name="author" content="zchengsite, 1451426471@qq.com" />
  <meta name="viewport" content="width=device-width, initial-scale=1" />
  
  
  <title>WGCNA教程 | Hexo</title>

  
    <link rel="apple-touch-icon" href="/images/favicon.png">
    <link rel="icon" href="/images/favicon.png">
  

  <!-- Raleway-Font -->
  <link href="https://fonts.googleapis.com/css?family=Raleway&display=swap" rel="stylesheet">

  <!-- hexo site css -->
  
<link rel="stylesheet" href="/css/base.css">
<link rel="stylesheet" href="/iconfont/iconfont.css">
<link rel="stylesheet" href="/css/github-markdown.css">
<link rel="stylesheet" href="/css/highlight.css">


  <!-- jquery3.3.1 -->
  <script src="https://cdn.bootcss.com/jquery/3.3.1/jquery.min.js"></script>

  <!-- fancybox -->
  <link href="https://cdn.bootcss.com/fancybox/3.5.2/jquery.fancybox.min.css" rel="stylesheet">
  <script async src="https://cdn.bootcss.com/fancybox/3.5.2/jquery.fancybox.min.js"></script>
  
<script src="/js/fancybox.js"></script>


  

<meta name="generator" content="Hexo 5.2.0"><link rel="alternate" href="/atom.xml" title="Hexo" type="application/atom+xml">
</head>


  <body>
    <div id="app">
      <div class="header">
  <div class="avatar">
    <a href="/">
      <!-- 头像取消懒加载，添加no-lazy -->
      
        <img src="/images/mooner.jpg" alt="">
      
    </a>
    <div class="nickname"><a href="/">mooner</a></div>
  </div>
  <div class="navbar">
    <ul>
      
        <li class="nav-item" data-path="/">
          <a href="/">主页</a>
        </li>
      
        <li class="nav-item" data-path="/archives/">
          <a href="/archives/">归档</a>
        </li>
      
        <li class="nav-item" data-path="/tags/">
          <a href="/tags/">标签</a>
        </li>
      
        <li class="nav-item" data-path="/about/">
          <a href="/about/">关于</a>
        </li>
      
    </ul>
  </div>
</div>


<script src="/js/activeNav.js"></script>



      <div class="flex-container">
        <!-- 文章详情页，展示文章具体内容，url形式：https://yoursite/文章标题/ -->
<!-- 同时为「标签tag」，「朋友friend」，「分类categories」，「关于about」页面的承载页面，具体展示取决于page.type -->



  

  

  

  
  <!-- 文章内容页 url形式：https://yoursite/文章标题/ -->
  <div class="container post-details" id="post-details">
    <div class="post-content">
      <div class="post-title">WGCNA教程</div>
      <div class="post-attach">
        <span class="post-pubtime">
          <i class="iconfont icon-updatetime" title="更新时间"></i>
          2020-08-15
        </span>
        
              <span class="post-tags">
                <i class="iconfont icon-tags" title="标签"></i>
                
                <span class="span--tag">
                  <a href="/tags/WGCNA/" title="WGCNA">
                    <b>#</b> WGCNA
                  </a>
                </span>
                
              </span>
          
      </div>
      <div class="markdown-body">
        <h1 id="WGCNA-分析"><a href="#WGCNA-分析" class="headerlink" title="WGCNA 分析"></a>WGCNA 分析</h1><h2 id="WGCNA基本概念"><a href="#WGCNA基本概念" class="headerlink" title="WGCNA基本概念"></a>WGCNA基本概念</h2><p>加权基因共表达网络分析 (WGCNA, Weighted correlation network analysis)是用来描述不同样品之间基因关联模式的系统生物学方法，可以用来鉴定高度<strong>协同变化</strong>的基因集。并根据基因集的内连性和基因集与表型之间的关联鉴定候补生物标记基因或治疗靶点。  </p>
<p>相比于只关注差异表达的基因，WGCNA利用数千或近万个变化最大的基因或全部基因的信息识别感兴趣的基因集，并与表型进行显著性关联分析。一是充分利用了信息，二是把数千个基因与表型的关联转换为数个基因集与表型的关联，免去了多重假设检验校正的问题。  </p>
<h2 id="WGCNA常见术语"><a href="#WGCNA常见术语" class="headerlink" title="WGCNA常见术语"></a>WGCNA常见术语</h2><ul>
<li>共表达网络（Co-expression network）：定义为加权基因网络。点代表基因，边代表基因表达相关性。加权是指对<strong>相关性值进行冥次运算</strong>。 (冥次的值也就是<strong>软阈值</strong> (<code>power</code>, <code>/post/2020-08-15-wgcna_fileskSoftThreshold</code>这个函数所做的就是确定合适的<code>power</code>))。无向网络的边属性计算方式为 <code>abs(cor(genex, geney)) ^ power</code>；有向网络的边属性计算方式为<code>(1+cor(genex, geney)/2) ^ power</code>; <code>sign hybrid</code>的边属性计算方式为<code>cor(genex, geney)^power if cor&gt;0 else 0</code>。这种处理方式强化了强相关，弱化了弱相关或负相关，使得相关性数值更符合<code>无标度网络</code>特征，更具有生物意义。如果没有合适的<strong>power</strong>，一般是由于部分样品与其它样品因为某种原因差别太大导致的，可根据具体问题移除部分样品或查看后面的<code>经验值</code>。  </li>
<li>模块（Module）：高度內连的基因集。在无向网络中，模块内是高度<strong>相关</strong>的基因。在有向网络中，模块内是高度<strong>正相关</strong>的基因。把基因聚类成模块后，可以对每个模块进行三个层次的分析：<ol>
<li>功能富集分析查看其功能特征是否与研究目的相符；</li>
<li>模块与性状进行关联分析，找出与关注性状相关度最高的模块；</li>
<li>模块与样本进行关联分析，找到样品特异高表达的模块。</li>
</ol>
</li>
<li>连接度（Connectivity）：类似于网络中 “度”(degree)的概念。每个基因的连接度是与其他网络基因的连接度。 在共表达网络中，连接度衡量的是一个基因与其他的网络基因有多相关。 </li>
<li>模块内连接度（Intramodular connectivity）: 模块内连接度衡量的是给定基因相对于特定模块的基因是如何连接或共同表达的。模内连接性可以解释为模块成员关系的度量。  </li>
<li>Module eigengene E: 给定模型的第一主成分，代表<strong>整个模型的基因表达谱</strong>。  </li>
<li>Module membership: 给定基因表达谱与给定模型的eigengene的相关性。  </li>
<li>Hub gene: 关键基因 (连接度最多或连接多个模块的基因)。  </li>
<li>Adjacency matrix ：基因和基因之间的加权相关性值构成的矩阵。  </li>
<li>TOM (Topological overlap matrix)：把邻接矩阵转换为拓扑重叠矩阵，以降低噪音和假相关，获得的新距离矩阵，这个信息可拿来构建网络或绘制TOM图。  </li>
<li>Gene signifcance GS：基因表达模式与某样本特质之间的相关，一般相关值越大，表示该基因越重要，有时候也用<code>-log（p）</code>来量化  </li>
<li>Module signifcance：该模块中所有基因相关绝对值的均值。当基因相关是基于基因表达模式与样本特质之间的相关值时，该测量与模块特征基因和样本特质之间存在高相关。  </li>
</ul>
<h2 id="基本分析流程"><a href="#基本分析流程" class="headerlink" title="基本分析流程"></a>基本分析流程</h2><p><img src="/images/wgcna/wgcna.png" alt="WGCNA">  </p>
<ol>
<li>构建基因共表达网络：使用加权的表达相关性。基本目的是充分利用基因间的交互模式；采用相关系数作为共表达的测量。  </li>
<li>识别基因集：基于加权相关性，进行层级聚类分析，并根据设定标准切分聚类结果，获得不同的基因模块，用聚类树的分枝和不同颜色表示。可以采用层次聚类（hierarchical clustering）或动态树切分（dynamic tree cut）实现。  </li>
<li>将模块与外部信息关联。其中外部信息包括基于阵列的临床数据（clinical data），多态性数据（SNPs）和蛋白组数据（proteomics）；基因信息包括ontology和功能富集（functional enrichment）。基本目的是发现具有生物学意义的模块。  </li>
<li>研究模型之间的关系，从系统层面查看不同模型的互作网络。基本目的是生物数据降维，并提供系统水平的视角（system-level view）；采用eigengene网络实现。  </li>
<li>最后发现感兴趣模块的关键驱动子（key drivers）。基本目的是实验验证并明确生物标记（biomarker）；采用的手段可以是模块内连接（intramodular connectivity）以及因果检验（causality testing）。  </li>
</ol>
<h2 id="WGCNA包实战"><a href="#WGCNA包实战" class="headerlink" title="WGCNA包实战"></a>WGCNA包实战</h2><p>R包<code>WGCNA</code>是用于计算各种加权关联分析的功能集合，可用于网络构建，基因筛选，基因簇鉴定，拓扑特征计算，数据模拟和可视化等。 </p>
<h3 id="准备工作"><a href="#准备工作" class="headerlink" title="准备工作"></a>准备工作</h3><ol>
<li>WGCNA本质是基于相关系数的网络分析方法，适用于多样品数据模式，一般要求样本数多于<strong>15</strong>个。样本数多于<strong>20</strong>时效果更好，样本越多，结果越稳定。  </li>
<li>基因表达矩阵: 常规表达矩阵即可，即基因在行，样品在列，进入分析前做一个转置。RPKM、FPKM或其它标准化方法影响不大。  </li>
<li>安装WGCNA   </li>
</ol>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">library</span>(WGCNA)</span><br><span class="line">Loading required package: dynamicTreeCut</span><br><span class="line">Loading required package: fastcluster</span><br><span class="line"></span><br><span class="line">Attaching package: ‘fastcluster’</span><br><span class="line"></span><br><span class="line">The following object is masked from ‘package:stats’:</span><br><span class="line"></span><br><span class="line">    hclust</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line">Attaching package: ‘WGCNA’</span><br><span class="line"></span><br><span class="line">The following object is masked from ‘package:stats’:</span><br><span class="line"></span><br><span class="line">    cor</span><br><span class="line"></span><br><span class="line">options(stringsAsFactors = <span class="literal">FALSE</span>)</span><br><span class="line"><span class="comment"># 开启多线程</span></span><br><span class="line">enableWGCNAThreads()</span><br><span class="line">Allowing parallel execution with up to <span class="number">7</span> working processes.</span><br><span class="line"></span><br><span class="line">load(<span class="string">&quot;./data.RData&quot;</span>)</span><br><span class="line">head(datTraits)  </span><br><span class="line">                  gsm cellline       subtype</span><br><span class="line">GSM1172844 GSM1172844    184A1 Non-malignant</span><br><span class="line">GSM1172845 GSM1172845    184B5 Non-malignant</span><br><span class="line">GSM1172846 GSM1172846    21MT1         Basal</span><br><span class="line">GSM1172847 GSM1172847    21MT2         Basal</span><br><span class="line">GSM1172848 GSM1172848     21NT         Basal</span><br><span class="line">GSM1172849 GSM1172849     21PT         Basal</span><br><span class="line">fpkm[<span class="number">1</span>:<span class="number">4</span>,<span class="number">1</span>:<span class="number">4</span>]  </span><br><span class="line">                GSM1172844 GSM1172845 GSM1172846  GSM1172847</span><br><span class="line">ENSG00000000003   <span class="number">95.21255</span>   <span class="number">95.69868</span>   <span class="number">19.99467</span>  <span class="number">65.6863763</span></span><br><span class="line">ENSG00000000005    <span class="number">0.00000</span>    <span class="number">0.00000</span>    <span class="number">0.00000</span>   <span class="number">0.1492021</span></span><br><span class="line">ENSG00000000419  <span class="number">453.20831</span>  <span class="number">243.64804</span>  <span class="number">142.05818</span> <span class="number">200.4131493</span></span><br><span class="line">ENSG00000000457   <span class="number">18.10439</span>   <span class="number">26.56661</span>   <span class="number">16.12776</span>  <span class="number">12.0873135</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## 选取前5000高变异的基因计算WGCNA</span></span><br><span class="line">datExpr &lt;- t(fpkm[order(apply(fpkm,<span class="number">1</span>,mad), decreasing = <span class="literal">T</span>)[<span class="number">1</span>:<span class="number">5000</span>],])</span><br><span class="line"></span><br></pre></td></tr></table></figure>



<h3 id="1-确定最佳beta值"><a href="#1-确定最佳beta值" class="headerlink" title="1. 确定最佳beta值"></a>1. 确定最佳beta值</h3><blockquote>
<p>在这里选取的软阈值是为了构建最有效的无尺度网络。</p>
</blockquote>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">powers = c(c(<span class="number">1</span>:<span class="number">10</span>), seq(from = <span class="number">12</span>, to=<span class="number">20</span>, by=<span class="number">2</span>))</span><br><span class="line"><span class="comment">## 批量计算多个power值的R方	</span></span><br><span class="line">sft = pickSoftThreshold(datExpr, powerVector = powers, RsquaredCut = <span class="number">0.85</span>, verbose = <span class="number">5</span>)</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><span class="line">pickSoftThreshold: will use block size <span class="number">5000.</span></span><br><span class="line"> pickSoftThreshold: calculating connectivity <span class="keyword">for</span> given powers...</span><br><span class="line">   ..working on genes <span class="number">1</span> through <span class="number">5000</span> of <span class="number">5000</span></span><br><span class="line">   Power SFT.R.sq  slope truncated.R.sq  mean.k. median.k.  max.k.</span><br><span class="line"><span class="number">1</span>      <span class="number">1</span>   <span class="number">0.0944</span> -<span class="number">0.904</span>          <span class="number">0.885</span> <span class="number">1040.000</span>  <span class="number">1.02e+03</span> <span class="number">1810.00</span></span><br><span class="line"><span class="number">2</span>      <span class="number">2</span>   <span class="number">0.4910</span> -<span class="number">1.580</span>          <span class="number">0.952</span>  <span class="number">328.000</span>  <span class="number">3.03e+02</span>  <span class="number">866.00</span></span><br><span class="line"><span class="number">3</span>      <span class="number">3</span>   <span class="number">0.7030</span> -<span class="number">1.860</span>          <span class="number">0.983</span>  <span class="number">128.000</span>  <span class="number">1.08e+02</span>  <span class="number">474.00</span></span><br><span class="line"><span class="number">4</span>      <span class="number">4</span>   <span class="number">0.7920</span> -<span class="number">2.000</span>          <span class="number">0.991</span>   <span class="number">57.300</span>  <span class="number">4.38e+01</span>  <span class="number">283.00</span></span><br><span class="line"><span class="number">5</span>      <span class="number">5</span>   <span class="number">0.8490</span> -<span class="number">2.060</span>          <span class="number">0.996</span>   <span class="number">28.400</span>  <span class="number">1.95e+01</span>  <span class="number">179.00</span></span><br><span class="line"><span class="number">6</span>      <span class="number">6</span>   <span class="number">0.8810</span> -<span class="number">2.090</span>          <span class="number">0.991</span>   <span class="number">15.200</span>  <span class="number">9.45e+00</span>  <span class="number">118.00</span></span><br><span class="line"><span class="number">7</span>      <span class="number">7</span>   <span class="number">0.9040</span> -<span class="number">2.070</span>          <span class="number">0.990</span>    <span class="number">8.640</span>  <span class="number">4.89e+00</span>   <span class="number">80.60</span></span><br><span class="line"><span class="number">8</span>      <span class="number">8</span>   <span class="number">0.9220</span> -<span class="number">2.040</span>          <span class="number">0.994</span>    <span class="number">5.170</span>  <span class="number">2.67e+00</span>   <span class="number">56.40</span></span><br><span class="line"><span class="number">9</span>      <span class="number">9</span>   <span class="number">0.9330</span> -<span class="number">2.030</span>          <span class="number">0.995</span>    <span class="number">3.240</span>  <span class="number">1.54e+00</span>   <span class="number">40.50</span></span><br><span class="line"><span class="number">10</span>    <span class="number">10</span>   <span class="number">0.9350</span> -<span class="number">2.020</span>          <span class="number">0.989</span>    <span class="number">2.100</span>  <span class="number">9.29e-01</span>   <span class="number">30.00</span></span><br><span class="line"><span class="number">11</span>    <span class="number">12</span>   <span class="number">0.9250</span> -<span class="number">2.030</span>          <span class="number">0.977</span>    <span class="number">0.971</span>  <span class="number">3.63e-01</span>   <span class="number">17.30</span></span><br><span class="line"><span class="number">12</span>    <span class="number">14</span>   <span class="number">0.9210</span> -<span class="number">2.020</span>          <span class="number">0.982</span>    <span class="number">0.496</span>  <span class="number">1.56e-01</span>   <span class="number">10.50</span></span><br><span class="line"><span class="number">13</span>    <span class="number">16</span>   <span class="number">0.9250</span> -<span class="number">1.970</span>          <span class="number">0.992</span>    <span class="number">0.275</span>  <span class="number">7.04e-02</span>    <span class="number">6.61</span></span><br><span class="line"><span class="number">14</span>    <span class="number">18</span>   <span class="number">0.8940</span> -<span class="number">1.960</span>          <span class="number">0.973</span>    <span class="number">0.163</span>  <span class="number">3.31e-02</span>    <span class="number">4.31</span></span><br><span class="line"><span class="number">15</span>    <span class="number">20</span>   <span class="number">0.9220</span> -<span class="number">1.820</span>          <span class="number">0.986</span>    <span class="number">0.102</span>  <span class="number">1.63e-02</span>    <span class="number">2.89</span></span><br><span class="line">power = sft$powerEstimate</span><br><span class="line">power</span><br><span class="line">[<span class="number">1</span>] <span class="number">6</span></span><br></pre></td></tr></table></figure>

<blockquote>
<p>由以上结果可以看出在power = 6时首次达到R方的cut阈值0.85，因此选取6作为合适的power值</p>
</blockquote>
<p>通过绘图更加直观的观察阈值的选取：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">par(mfrow = c(<span class="number">1</span>,<span class="number">2</span>));</span><br><span class="line">cex1 = <span class="number">0.9</span>;</span><br><span class="line"><span class="comment"># Scale-free topology fit index as a function of the soft-thresholding power</span></span><br><span class="line">plot(sft$fitIndices[,<span class="number">1</span>], -sign(sft$fitIndices[,<span class="number">3</span>])*sft$fitIndices[,<span class="number">2</span>],</span><br><span class="line">      xlab=<span class="string">&quot;Soft Threshold (power)&quot;</span>,ylab=<span class="string">&quot;Scale Free Topology Model Fit,signed R^2&quot;</span>,type=<span class="string">&quot;n&quot;</span>,</span><br><span class="line">      main = paste(<span class="string">&quot;Scale independence&quot;</span>));</span><br><span class="line">text(sft$fitIndices[,<span class="number">1</span>], -sign(sft$fitIndices[,<span class="number">3</span>])*sft$fitIndices[,<span class="number">2</span>],</span><br><span class="line">      labels=powers,cex=cex1,col=<span class="string">&quot;red&quot;</span>);</span><br><span class="line"><span class="comment"># this line corresponds to using an R^2 cut-off of h</span></span><br><span class="line">abline(h=<span class="number">0.85</span>,col=<span class="string">&quot;red&quot;</span>)</span><br><span class="line"><span class="comment"># Mean connectivity as a function of the soft-thresholding power</span></span><br><span class="line">plot(sft$fitIndices[,<span class="number">1</span>], sft$fitIndices[,<span class="number">5</span>],</span><br><span class="line">      xlab=<span class="string">&quot;Soft Threshold (power)&quot;</span>,ylab=<span class="string">&quot;Mean Connectivity&quot;</span>, type=<span class="string">&quot;n&quot;</span>,</span><br><span class="line">      main = paste(<span class="string">&quot;Mean connectivity&quot;</span>))</span><br><span class="line">text(sft$fitIndices[,<span class="number">1</span>], sft$fitIndices[,<span class="number">5</span>], labels=powers, cex=cex1,col=<span class="string">&quot;red&quot;</span>)</span><br></pre></td></tr></table></figure>

<p><img src="/images/wgcna/soft.png" alt="soft"></p>
<blockquote>
<p>假如没有合适的阈值，在官方文档中说，如果是unsigned的就选6，如果是signed的就选12.  </p>
</blockquote>
<h3 id="2-网络构建"><a href="#2-网络构建" class="headerlink" title="2. 网络构建"></a>2. 网络构建</h3><h4 id="一步法构建"><a href="#一步法构建" class="headerlink" title="一步法构建"></a>一步法构建</h4><p><code>WGCNA</code>提供了<code>blockwiseModules</code>函数可以一步计算出所有的模块</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">##一步法网络构建：One-step network construction and module detection</span></span><br><span class="line"><span class="comment"># numericLabels: 返回数字而不是颜色作为模块的名字，后面可以再转换为颜色</span></span><br><span class="line"><span class="comment"># saveTOMs：最耗费时间的计算，存储起来，供后续使用</span></span><br><span class="line"><span class="comment"># mergeCutHeight: 合并模块的阈值，越大模块越</span></span><br><span class="line">net = blockwiseModules(datExpr, power = power, maxBlockSize = <span class="number">6000</span>,</span><br><span class="line">                       TOMType = type, minModuleSize = <span class="number">30</span>,</span><br><span class="line">                       reassignThreshold = <span class="number">0</span>, mergeCutHeight = <span class="number">0.25</span>,</span><br><span class="line">                       numericLabels = <span class="literal">TRUE</span>, pamRespectsDendro = <span class="literal">FALSE</span>,</span><br><span class="line">                       saveTOMs=<span class="literal">TRUE</span>, corType = corType, </span><br><span class="line">                       maxPOutliers=maxPOutliers, loadTOMs=<span class="literal">TRUE</span>,</span><br><span class="line">                       saveTOMFileBase = paste0(exprMat, <span class="string">&quot;.tom&quot;</span>), <span class="comment">## 保存tom矩阵</span></span><br><span class="line">                       verbose = <span class="number">3</span>)</span><br></pre></td></tr></table></figure>

<p>展示最终层级聚类的结果  </p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## 灰色的为未分到模块的基因。</span></span><br><span class="line"><span class="comment"># Convert labels to colors for plotting</span></span><br><span class="line">moduleLabels = net$colors</span><br><span class="line">moduleColors = labels2colors(moduleLabels)</span><br><span class="line"><span class="comment"># Plot the dendrogram and the module colors underneath</span></span><br><span class="line">plotDendroAndColors(net$dendrograms[[<span class="number">1</span>]], moduleColors[net$blockGenes[[<span class="number">1</span>]]],</span><br><span class="line">                    <span class="string">&quot;Module colors&quot;</span>,</span><br><span class="line">                    dendroLabels = <span class="literal">FALSE</span>, hang = <span class="number">0.03</span>,</span><br><span class="line">                    addGuide = <span class="literal">TRUE</span>, guideHang = <span class="number">0.05</span>)</span><br></pre></td></tr></table></figure>

<p><img src="/images/wgcna/cluster.png" alt="cluster"></p>
<h4 id="分步构建"><a href="#分步构建" class="headerlink" title="分步构建"></a>分步构建</h4><p>首先计算邻接矩阵  </p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># calculate the adjacency</span></span><br><span class="line">softPower = <span class="number">6</span>;</span><br><span class="line">adjacency = adjacency(datExpr, power = softPower);</span><br></pre></td></tr></table></figure>

<p>将邻接矩阵转换成拓扑重叠矩阵  </p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Turn adjacency into topological overlap</span></span><br><span class="line">TOM = TOMsimilarity(adjacency);</span><br><span class="line">dissTOM = <span class="number">1</span>-TOM</span><br></pre></td></tr></table></figure>

<p>对TOM进行聚类  </p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Call the hierarchical clustering function</span></span><br><span class="line">geneTree = hclust(as.dist(dissTOM), method = <span class="string">&quot;average&quot;</span>);</span><br><span class="line"><span class="comment"># Plot the resulting clustering tree (dendrogram)</span></span><br><span class="line">plot(geneTree, xlab=<span class="string">&quot;&quot;</span>, sub=<span class="string">&quot;&quot;</span>, main = <span class="string">&quot;Gene clustering on TOM-based dissimilarity&quot;</span>,</span><br><span class="line">     labels = <span class="literal">FALSE</span>, hang = <span class="number">0.04</span>);</span><br></pre></td></tr></table></figure>

<p><img src="/images/wgcna/tree.png" alt="tree"></p>
<p>上面得到了基因聚类的树，我们对这棵树进行剪枝得到不同的模块  </p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># We like large modules, so we set the minimum module size relatively high:</span></span><br><span class="line">minModuleSize = <span class="number">30</span>;</span><br><span class="line"><span class="comment"># Module identification using dynamic tree cut:</span></span><br><span class="line">dynamicMods = cutreeDynamic(dendro = geneTree, distM = dissTOM,</span><br><span class="line">                            deepSplit = <span class="number">2</span>, pamRespectsDendro = <span class="literal">FALSE</span>,</span><br><span class="line">                            minClusterSize = minModuleSize);</span><br><span class="line">table(dynamicMods)</span><br></pre></td></tr></table></figure>

<p>得到模块后我们就可以看到不同的模块表示 </p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Convert numeric lables into colors</span></span><br><span class="line">dynamicColors = labels2colors(dynamicMods)</span><br><span class="line">table(dynamicColors)</span><br><span class="line"><span class="comment"># Plot the dendrogram and colors underneath</span></span><br><span class="line">plotDendroAndColors(geneTree, dynamicColors, <span class="string">&quot;Dynamic Tree Cut&quot;</span>,</span><br><span class="line">                    dendroLabels = <span class="literal">FALSE</span>, hang = <span class="number">0.03</span>,</span><br><span class="line">                    addGuide = <span class="literal">TRUE</span>, guideHang = <span class="number">0.05</span>,</span><br><span class="line">                    main = <span class="string">&quot;Gene dendrogram and module colors&quot;</span>)</span><br></pre></td></tr></table></figure>

<p><img src="/images/wgcna/cuttree.png" alt="cuttree">  </p>
<p>这样我们就得到了不同的基因表达模块。然而在这时某些模块之间也有可能具有很高的相似性，为了进一步衡量这些模块之间的共表达相似性，我们计算了eigengenes然后计算他们之间的相似性。  </p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">MEList = moduleEigengenes(datExpr, colors = dynamicColors)</span><br><span class="line">MEs = MEList$eigengenes</span><br><span class="line"><span class="comment"># Calculate dissimilarity of module eigengenes</span></span><br><span class="line">MEDiss = <span class="number">1</span>-cor(MEs);</span><br><span class="line"><span class="comment"># Cluster module eigengenes</span></span><br><span class="line">METree = hclust(as.dist(MEDiss), method = <span class="string">&quot;average&quot;</span>);</span><br><span class="line"><span class="comment"># Plot the result</span></span><br><span class="line">plot(METree, main = <span class="string">&quot;Clustering of module eigengenes&quot;</span>,</span><br><span class="line">xlab = <span class="string">&quot;&quot;</span>, sub = <span class="string">&quot;&quot;</span>)</span><br><span class="line"><span class="comment"># 在这里选取0.25作为阈值对eigengenes进行合并  </span></span><br><span class="line">MEDissThres = <span class="number">0.25</span></span><br><span class="line"><span class="comment"># Plot the cut line into the dendrogram</span></span><br><span class="line">abline(h=MEDissThres, col = <span class="string">&quot;red&quot;</span>)</span><br></pre></td></tr></table></figure>

<p><img src="/images/wgcna/eigengenes.png" alt="eigengenes">  </p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Call an automatic merging function</span></span><br><span class="line">merge = mergeCloseModules(datExpr, dynamicColors, cutHeight = MEDissThres, verbose = <span class="number">3</span>)</span><br><span class="line"><span class="comment"># The merged module colors</span></span><br><span class="line">mergedColors = merge$colors;</span><br><span class="line"><span class="comment"># Eigengenes of the new merged modules:</span></span><br><span class="line">mergedMEs = merge$newMEs;</span><br><span class="line">plotDendroAndColors(geneTree, cbind(dynamicColors, mergedColors),</span><br><span class="line">                    c(<span class="string">&quot;Dynamic Tree Cut&quot;</span>, <span class="string">&quot;Merged dynamic&quot;</span>),</span><br><span class="line">                    dendroLabels = <span class="literal">FALSE</span>, hang = <span class="number">0.03</span>,</span><br><span class="line">                    addGuide = <span class="literal">TRUE</span>, guideHang = <span class="number">0.05</span>)</span><br></pre></td></tr></table></figure>

<p><img src="/images/wgcna/mergetree.png" alt="mergetree">  </p>
<blockquote>
<p>这个结果其实和一步计算的结果基本一致,得到模块数目相同  </p>
</blockquote>
<h3 id="3-模块之间的相关性"><a href="#3-模块之间的相关性" class="headerlink" title="3. 模块之间的相关性"></a>3. 模块之间的相关性</h3><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># module eigengene, 可以绘制线图，作为每个模块的基因表达趋势的展示</span></span><br><span class="line">MEs = net$MEs</span><br><span class="line">MEs_col = MEs</span><br><span class="line">colnames(MEs_col) = paste0(<span class="string">&quot;ME&quot;</span>, labels2colors(</span><br><span class="line">  as.numeric(str_replace_all(colnames(MEs),<span class="string">&quot;ME&quot;</span>,<span class="string">&quot;&quot;</span>))))</span><br><span class="line">MEs_col = orderMEs(MEs_col)</span><br><span class="line"><span class="comment"># marDendro/marHeatmap 设置下、左、上、右的边距</span></span><br><span class="line">plotEigengeneNetworks(MEs_col, <span class="string">&quot;Eigengene adjacency heatmap&quot;</span>, </span><br><span class="line">                      marDendro = c(<span class="number">3</span>,<span class="number">3</span>,<span class="number">2</span>,<span class="number">4</span>),</span><br><span class="line">                      marHeatmap = c(<span class="number">3</span>,<span class="number">4</span>,<span class="number">2</span>,<span class="number">2</span>), plotDendrograms = <span class="literal">T</span>, </span><br><span class="line">                      xLabelsAngle = <span class="number">90</span>)</span><br></pre></td></tr></table></figure>

<p><img src="/images/wgcna/eiheatmap.png" alt="eiheatmap">  </p>
<h3 id="4-样本和性状的聚类关系"><a href="#4-样本和性状的聚类关系" class="headerlink" title="4. 样本和性状的聚类关系"></a>4. 样本和性状的聚类关系</h3><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">nGenes = ncol(datExpr)</span><br><span class="line">nSamples = nrow(datExpr)</span><br><span class="line"><span class="comment">#首先针对样本做个系统聚类</span></span><br><span class="line">datExpr_tree&lt;-hclust(dist(datExpr), method = <span class="string">&quot;average&quot;</span>)</span><br><span class="line">sample_colors &lt;- numbers2colors(as.numeric(factor(datTraits$subtype)), </span><br><span class="line">                                colors = c(<span class="string">&quot;white&quot;</span>,<span class="string">&quot;blue&quot;</span>,<span class="string">&quot;red&quot;</span>,<span class="string">&quot;green&quot;</span>),signed = <span class="literal">FALSE</span>)</span><br><span class="line">par(mar = c(<span class="number">1</span>,<span class="number">4</span>,<span class="number">3</span>,<span class="number">1</span>),cex=<span class="number">0.8</span>)</span><br><span class="line">plotDendroAndColors(datExpr_tree, sample_colors,</span><br><span class="line">                      groupLabels = colnames(sample),</span><br><span class="line">                      cex.dendroLabels = <span class="number">0.8</span>,</span><br><span class="line">                      marAll = c(<span class="number">1</span>, <span class="number">4</span>, <span class="number">3</span>, <span class="number">1</span>),</span><br><span class="line">                      cex.rowText = <span class="number">0.01</span>,</span><br><span class="line">                      main = <span class="string">&quot;Sample dendrogram and trait heatmap&quot;</span>)</span><br></pre></td></tr></table></figure>

<p><img src="/images/wgcna/sample-subtype-cluster.png" alt="sample_subtype">  </p>
<p>条形图展示  </p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">Luminal = as.data.frame(design[,<span class="number">3</span>]);</span><br><span class="line">names(Luminal) = <span class="string">&quot;Luminal&quot;</span></span><br><span class="line">y=Luminal</span><br><span class="line">GS1=as.numeric(cor(y,datExpr, use=<span class="string">&quot;p&quot;</span>))</span><br><span class="line">GeneSignificance=abs(GS1)</span><br><span class="line"><span class="comment"># Next module significance is defined as average gene significance.</span></span><br><span class="line">ModuleSignificance=tapply(GeneSignificance,</span><br><span class="line">                          moduleColors, mean, na.rm=<span class="literal">T</span>)</span><br><span class="line">sizeGrWindow(<span class="number">8</span>,<span class="number">7</span>)</span><br><span class="line">par(mfrow = c(<span class="number">1</span>,<span class="number">1</span>))</span><br><span class="line"><span class="comment"># 如果模块太多，下面的展示就不友好</span></span><br><span class="line"><span class="comment"># 不过，我们可以自定义出图。</span></span><br><span class="line">plotModuleSignificance(GeneSignificance,moduleColors)</span><br></pre></td></tr></table></figure>

<p><img src="/images/wgcna/bar.png" alt="bar">  </p>
<h3 id="5-可视化基因网络"><a href="#5-可视化基因网络" class="headerlink" title="5. 可视化基因网络"></a>5. 可视化基因网络</h3><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">nSelect = <span class="number">400</span></span><br><span class="line"><span class="comment"># For reproducibility, we set the random seed</span></span><br><span class="line">set.seed(<span class="number">10</span>);</span><br><span class="line">select = sample(nGenes, size = nSelect);</span><br><span class="line">selectTOM = dissTOM[select, select];</span><br><span class="line"><span class="comment"># There’s no simple way of restricting a clustering tree to a subset of genes, so we must re-cluster.</span></span><br><span class="line">selectTree = hclust(as.dist(selectTOM), method = <span class="string">&quot;average&quot;</span>)</span><br><span class="line">selectColors = moduleColors[select];</span><br><span class="line">plotDiss = selectTOM^<span class="number">7</span>;</span><br><span class="line">diag(plotDiss) = <span class="literal">NA</span>;</span><br><span class="line">TOMplot(plotDiss, selectTree, selectColors, main = <span class="string">&quot;Network heatmap plot, selected genes&quot;</span>)</span><br></pre></td></tr></table></figure>

<p><img src="/images/wgcna/geneheatmap.png" alt="geneheatmap">  </p>
<h3 id="6-模块和性状的关系"><a href="#6-模块和性状的关系" class="headerlink" title="6. 模块和性状的关系"></a>6. 模块和性状的关系</h3><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br></pre></td><td class="code"><pre><span class="line">nGenes = ncol(datExpr)</span><br><span class="line">nSamples = nrow(datExpr)</span><br><span class="line">design=model.matrix(~<span class="number">0</span>+ datTraits$subtype)</span><br><span class="line">colnames(design)=levels(datTraits$subtype)</span><br><span class="line">moduleColors &lt;- labels2colors(net$colors)</span><br><span class="line"><span class="comment"># Recalculate MEs with color labels</span></span><br><span class="line">MEs0 = moduleEigengenes(datExpr, moduleColors)$eigengenes</span><br><span class="line">MEs = orderMEs(MEs0); <span class="comment">##不同颜色的模块的ME值矩 (样本vs模块)</span></span><br><span class="line">moduleTraitCor = cor(MEs, design , use = <span class="string">&quot;p&quot;</span>);</span><br><span class="line">moduleTraitPvalue = corPvalueStudent(moduleTraitCor, nSamples)</span><br><span class="line"></span><br><span class="line">sizeGrWindow(<span class="number">10</span>,<span class="number">6</span>)</span><br><span class="line"><span class="comment"># Will display correlations and their p-values</span></span><br><span class="line">textMatrix = paste(signif(moduleTraitCor, <span class="number">2</span>), <span class="string">&quot;\n(&quot;</span>,</span><br><span class="line">                   signif(moduleTraitPvalue, <span class="number">1</span>), <span class="string">&quot;)&quot;</span>, sep = <span class="string">&quot;&quot;</span>);</span><br><span class="line">dim(textMatrix) = dim(moduleTraitCor)</span><br><span class="line">par(mar = c(<span class="number">6</span>, <span class="number">8.5</span>, <span class="number">3</span>, <span class="number">3</span>));</span><br><span class="line"><span class="comment"># Display the correlation values within a heatmap plot</span></span><br><span class="line">labeledHeatmap(Matrix = moduleTraitCor,</span><br><span class="line">               xLabels = colnames(design),</span><br><span class="line">               yLabels = names(MEs),</span><br><span class="line">               ySymbols = names(MEs),</span><br><span class="line">               colorLabels = <span class="literal">FALSE</span>,</span><br><span class="line">               colors = greenWhiteRed(<span class="number">50</span>),</span><br><span class="line">               textMatrix = textMatrix,</span><br><span class="line">               setStdMargins = <span class="literal">FALSE</span>,</span><br><span class="line">               cex.text = <span class="number">0.5</span>,</span><br><span class="line">               zlim = c(-<span class="number">1</span>,<span class="number">1</span>),</span><br><span class="line">               main = paste(<span class="string">&quot;Module-trait relationships&quot;</span>))</span><br></pre></td></tr></table></figure>

<p><img src="/images/wgcna/Module-trait-relationships.png" alt="moduletrait">  </p>
<p>从上图已经可以看到跟乳腺癌分类相关的基因模块了，包括<code>Basal</code> 、<code>Claudin-low</code> 、<code>Luminal</code> 、<code>Non-malignant</code>、<code>unknown</code> 这5类所对应的不同模块的基因列表。可以看到每一种乳腺癌都有跟它强烈相关的模块，可以作为它的表达signature，模块里面的基因可以拿去做下游分析。我们看到<strong>Luminal表型</strong>跟<strong>棕色的模块</strong>相关性高达0.86，而且极其显著的相关，所以值得我们挖掘，这<strong>个模块里面的基因是什么，为什么如此的相关呢？</strong>  </p>
<h3 id="7-感兴趣模块的具体分析"><a href="#7-感兴趣模块的具体分析" class="headerlink" title="7. 感兴趣模块的具体分析"></a>7. 感兴趣模块的具体分析</h3><p>性状跟模块虽然求出了相关性，可以挑选最相关的那些模块来分析，但是模块本身仍然包含非常多的基因，还需进一步的寻找最重要的基因。所有的模块都可以跟基因算出相关系数，所有的连续型性状也可以跟基因的表达值算出相关系数。 <strong>如果跟性状显著相关基因也跟某个模块显著相关，那么这些基因可能就非常重要</strong>。  </p>
<h4 id="计算模块与基因的相关性矩阵"><a href="#计算模块与基因的相关性矩阵" class="headerlink" title="计算模块与基因的相关性矩阵"></a>计算模块与基因的相关性矩阵</h4><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># names (colors) of the modules</span></span><br><span class="line">modNames = substring(names(MEs), <span class="number">3</span>)</span><br><span class="line"><span class="comment">## 算出每个模块跟基因的皮尔森相关系数矩阵</span></span><br><span class="line">geneModuleMembership = as.data.frame(cor(datExpr, MEs, use = <span class="string">&quot;p&quot;</span>));</span><br><span class="line">MMPvalue = as.data.frame(corPvalueStudent(as.matrix(geneModuleMembership), nSamples));</span><br><span class="line">names(geneModuleMembership) = paste(<span class="string">&quot;MM&quot;</span>, modNames, sep=<span class="string">&quot;&quot;</span>);</span><br><span class="line">names(MMPvalue) = paste(<span class="string">&quot;p.MM&quot;</span>, modNames, sep=<span class="string">&quot;&quot;</span>);</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<h4 id="计算性状与基因的相关性矩阵"><a href="#计算性状与基因的相关性矩阵" class="headerlink" title="计算性状与基因的相关性矩阵"></a>计算性状与基因的相关性矩阵</h4><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## 只有连续型性状才能只有计算</span></span><br><span class="line"><span class="comment">## 这里把是否属于 Luminal 表型这个变量用0,1进行数值化。</span></span><br><span class="line">Luminal = as.data.frame(design[,<span class="number">3</span>]);</span><br><span class="line">names(Luminal) = <span class="string">&quot;Luminal&quot;</span></span><br><span class="line">geneTraitSignificance = as.data.frame(cor(datExpr, Luminal, use = <span class="string">&quot;p&quot;</span>));</span><br><span class="line">GSPvalue = as.data.frame(corPvalueStudent(as.matrix(geneTraitSignificance), nSamples));</span><br><span class="line">names(geneTraitSignificance) = paste(<span class="string">&quot;GS.&quot;</span>, names(Luminal), sep=<span class="string">&quot;&quot;</span>);</span><br><span class="line">names(GSPvalue) = paste(<span class="string">&quot;p.GS.&quot;</span>, names(Luminal), sep=<span class="string">&quot;&quot;</span>);</span><br></pre></td></tr></table></figure>

<h4 id="指定感兴趣模块进行分析"><a href="#指定感兴趣模块进行分析" class="headerlink" title="指定感兴趣模块进行分析"></a>指定感兴趣模块进行分析</h4><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">module = <span class="string">&quot;brown&quot;</span></span><br><span class="line">column = match(module, modNames);</span><br><span class="line">moduleGenes = moduleColors==module;</span><br><span class="line">sizeGrWindow(<span class="number">7</span>, <span class="number">7</span>)</span><br><span class="line">par(mfrow = c(<span class="number">1</span>,<span class="number">1</span>))</span><br><span class="line">verboseScatterplot(abs(geneModuleMembership[moduleGenes, column]),</span><br><span class="line">                     abs(geneTraitSignificance[moduleGenes, <span class="number">1</span>]),</span><br><span class="line">                     xlab = paste(<span class="string">&quot;Module Membership in&quot;</span>, module, <span class="string">&quot;module&quot;</span>),</span><br><span class="line">                     ylab = <span class="string">&quot;Gene significance for Luminal&quot;</span>,</span><br><span class="line">                     main = paste(<span class="string">&quot;Module membership vs. gene significance\n&quot;</span>),</span><br><span class="line">                     cex.main = <span class="number">1.2</span>, cex.lab = <span class="number">1.2</span>, cex.axis = <span class="number">1.2</span>, col = module)</span><br></pre></td></tr></table></figure>

<p><img src="/images/wgcna/cor.png" alt="cor">  </p>
<p>可以看到这些基因不仅仅是跟其对应的模块高度相关，而且是跟其对应的性状高度相关，进一步说明了基因值得深度探究。  </p>
<h3 id="8-提取指定模块的基因名"><a href="#8-提取指定模块的基因名" class="headerlink" title="8. 提取指定模块的基因名"></a>8. 提取指定模块的基因名</h3><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Select module</span></span><br><span class="line">module = <span class="string">&quot;brown&quot;</span>;</span><br><span class="line"><span class="comment"># Select module probes</span></span><br><span class="line">probes = colnames(datExpr) <span class="comment">## 我们例子里面的probe就是基因</span></span><br><span class="line">inModule = (moduleColors==module);</span><br><span class="line">modProbes = probes[inModule]; </span><br><span class="line">head(modProbes)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 如果使用WGCNA包自带的热图就很丑。</span></span><br><span class="line">which.module=<span class="string">&quot;brown&quot;</span>;</span><br><span class="line">dat=datExpr[,moduleColors==which.module ] </span><br><span class="line">plotMat(t(scale(dat)),nrgcols=<span class="number">30</span>,rlabels=<span class="literal">T</span>,</span><br><span class="line">        clabels=<span class="literal">T</span>,rcols=which.module,</span><br><span class="line">        title=which.module )</span><br><span class="line">datExpr[<span class="number">1</span>:<span class="number">4</span>,<span class="number">1</span>:<span class="number">4</span>]</span><br><span class="line">dat=t(datExpr[,moduleColors==which.module ] )</span><br><span class="line"><span class="keyword">library</span>(pheatmap)</span><br><span class="line">pheatmap(dat ,show_colnames =<span class="literal">F</span>,show_rownames = <span class="literal">F</span>) <span class="comment">#对那些提取出来的1000个基因所在的每一行取出，组合起来为一个新的表达矩阵</span></span><br><span class="line">n=t(scale(t(log(dat+<span class="number">1</span>)))) <span class="comment"># &#x27;scale&#x27;可以对log-ratio数值进行归一化</span></span><br><span class="line">n[n&gt;<span class="number">2</span>]=<span class="number">2</span> </span><br><span class="line">n[n&lt; -<span class="number">2</span>]= -<span class="number">2</span></span><br><span class="line">n[<span class="number">1</span>:<span class="number">4</span>,<span class="number">1</span>:<span class="number">4</span>]</span><br><span class="line">pheatmap(n,show_colnames =<span class="literal">F</span>,show_rownames = <span class="literal">F</span>)</span><br><span class="line">group_list=datTraits$subtype</span><br><span class="line">ac=data.frame(g=group_list)</span><br><span class="line">rownames(ac)=colnames(n) </span><br><span class="line">pheatmap(n,show_colnames =<span class="literal">F</span>,show_rownames = <span class="literal">F</span>,</span><br><span class="line">         annotation_col=ac )</span><br><span class="line"><span class="comment"># 可以很清晰的看到，所有的形状相关的模块基因</span></span><br><span class="line"><span class="comment"># 其实未必就不是差异表达基因。</span></span><br></pre></td></tr></table></figure>

<p><img src="/images/wgcna/heatmap.png" alt="heatmap">  </p>
<p>有了基因信息，后边就可以进行GO/KEGG等功能数据库的注释  </p>
<h3 id="9-模块导出"><a href="#9-模块导出" class="headerlink" title="9. 模块导出"></a>9. 模块导出</h3><p>主要模块里面的基因直接的相互作用关系信息可以导出到<strong>cytoscape</strong>,<strong>VisANT</strong>等网络可视化软件。  </p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Select module</span></span><br><span class="line">module = <span class="string">&quot;brown&quot;</span>;</span><br><span class="line"><span class="comment"># Select module probes</span></span><br><span class="line">probes = colnames(datExpr) <span class="comment">## 我们例子里面的probe就是基因名</span></span><br><span class="line">inModule = (moduleColors==module);</span><br><span class="line">modProbes = probes[inModule]; </span><br><span class="line"><span class="comment">## 也是提取指定模块的基因名</span></span><br><span class="line"><span class="comment"># Select the corresponding Topological Overlap</span></span><br><span class="line">modTOM = TOM[inModule, inModule];</span><br><span class="line">dimnames(modTOM) = list(modProbes, modProbes)</span><br></pre></td></tr></table></figure>

<p>导出到<strong>cytoscape</strong>  </p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">cyt = exportNetworkToCytoscape(</span><br><span class="line">      modTOM,</span><br><span class="line">     edgeFile = paste(<span class="string">&quot;CytoscapeInput-edges-&quot;</span>, paste(module, collapse=<span class="string">&quot;-&quot;</span>), <span class="string">&quot;.txt&quot;</span>, sep=<span class="string">&quot;&quot;</span>),</span><br><span class="line">     nodeFile = paste(<span class="string">&quot;CytoscapeInput-nodes-&quot;</span>, paste(module, collapse=<span class="string">&quot;-&quot;</span>), <span class="string">&quot;.txt&quot;</span>, sep=<span class="string">&quot;&quot;</span>),</span><br><span class="line">     weighted = <span class="literal">TRUE</span>,</span><br><span class="line">     threshold = <span class="number">0.02</span>,</span><br><span class="line">     nodeNames = modProbes, </span><br><span class="line">     nodeAttr = moduleColors[inModule]);</span><br></pre></td></tr></table></figure>



<h3 id="Reference"><a href="#Reference" class="headerlink" title="Reference"></a>Reference</h3><ol>
<li><a target="_blank" rel="noopener" href="https://www.jianshu.com/p/e9cc3f43441d">WGCNA分析，简单全面的最新教程</a>  </li>
<li><a target="_blank" rel="noopener" href="https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/">https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/</a>  </li>
</ol>
      </div>
      
        <div class="prev-or-next">
          <div class="post-foot-next">
            
              <a href="/2020/08/12/single-cell-i/" target="_self">
                <i class="iconfont icon-chevronleft"></i>
                <span>上一页</span>
              </a>
            
          </div>
          <div class="post-attach">
            <span class="post-pubtime">
              <i class="iconfont icon-updatetime" title="更新时间"></i>
              2020-08-15
            </span>
            
                  <span class="post-tags">
                    <i class="iconfont icon-tags" title="标签"></i>
                    
                    <span class="span--tag">
                      <a href="/tags/WGCNA/" title="WGCNA">
                        <b>#</b> WGCNA
                      </a>
                    </span>
                    
                  </span>
              
          </div>
          <div class="post-foot-prev">
            
          </div>
        </div>
      
    </div>
    
  <div class="post-catalog" id="catalog">
    <div class="title">目录</div>
    <div class="catalog-content">
      <ol class="toc"><li class="toc-item toc-level-1"><a class="toc-link" href="#WGCNA-%E5%88%86%E6%9E%90"><span class="toc-text">WGCNA 分析</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#WGCNA%E5%9F%BA%E6%9C%AC%E6%A6%82%E5%BF%B5"><span class="toc-text">WGCNA基本概念</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#WGCNA%E5%B8%B8%E8%A7%81%E6%9C%AF%E8%AF%AD"><span class="toc-text">WGCNA常见术语</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%9F%BA%E6%9C%AC%E5%88%86%E6%9E%90%E6%B5%81%E7%A8%8B"><span class="toc-text">基本分析流程</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#WGCNA%E5%8C%85%E5%AE%9E%E6%88%98"><span class="toc-text">WGCNA包实战</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#%E5%87%86%E5%A4%87%E5%B7%A5%E4%BD%9C"><span class="toc-text">准备工作</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#1-%E7%A1%AE%E5%AE%9A%E6%9C%80%E4%BD%B3beta%E5%80%BC"><span class="toc-text">1. 确定最佳beta值</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#2-%E7%BD%91%E7%BB%9C%E6%9E%84%E5%BB%BA"><span class="toc-text">2. 网络构建</span></a><ol class="toc-child"><li class="toc-item toc-level-4"><a class="toc-link" href="#%E4%B8%80%E6%AD%A5%E6%B3%95%E6%9E%84%E5%BB%BA"><span class="toc-text">一步法构建</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#%E5%88%86%E6%AD%A5%E6%9E%84%E5%BB%BA"><span class="toc-text">分步构建</span></a></li></ol></li><li class="toc-item toc-level-3"><a class="toc-link" href="#3-%E6%A8%A1%E5%9D%97%E4%B9%8B%E9%97%B4%E7%9A%84%E7%9B%B8%E5%85%B3%E6%80%A7"><span class="toc-text">3. 模块之间的相关性</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#4-%E6%A0%B7%E6%9C%AC%E5%92%8C%E6%80%A7%E7%8A%B6%E7%9A%84%E8%81%9A%E7%B1%BB%E5%85%B3%E7%B3%BB"><span class="toc-text">4. 样本和性状的聚类关系</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#5-%E5%8F%AF%E8%A7%86%E5%8C%96%E5%9F%BA%E5%9B%A0%E7%BD%91%E7%BB%9C"><span class="toc-text">5. 可视化基因网络</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#6-%E6%A8%A1%E5%9D%97%E5%92%8C%E6%80%A7%E7%8A%B6%E7%9A%84%E5%85%B3%E7%B3%BB"><span class="toc-text">6. 模块和性状的关系</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#7-%E6%84%9F%E5%85%B4%E8%B6%A3%E6%A8%A1%E5%9D%97%E7%9A%84%E5%85%B7%E4%BD%93%E5%88%86%E6%9E%90"><span class="toc-text">7. 感兴趣模块的具体分析</span></a><ol class="toc-child"><li class="toc-item toc-level-4"><a class="toc-link" href="#%E8%AE%A1%E7%AE%97%E6%A8%A1%E5%9D%97%E4%B8%8E%E5%9F%BA%E5%9B%A0%E7%9A%84%E7%9B%B8%E5%85%B3%E6%80%A7%E7%9F%A9%E9%98%B5"><span class="toc-text">计算模块与基因的相关性矩阵</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#%E8%AE%A1%E7%AE%97%E6%80%A7%E7%8A%B6%E4%B8%8E%E5%9F%BA%E5%9B%A0%E7%9A%84%E7%9B%B8%E5%85%B3%E6%80%A7%E7%9F%A9%E9%98%B5"><span class="toc-text">计算性状与基因的相关性矩阵</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#%E6%8C%87%E5%AE%9A%E6%84%9F%E5%85%B4%E8%B6%A3%E6%A8%A1%E5%9D%97%E8%BF%9B%E8%A1%8C%E5%88%86%E6%9E%90"><span class="toc-text">指定感兴趣模块进行分析</span></a></li></ol></li><li class="toc-item toc-level-3"><a class="toc-link" href="#8-%E6%8F%90%E5%8F%96%E6%8C%87%E5%AE%9A%E6%A8%A1%E5%9D%97%E7%9A%84%E5%9F%BA%E5%9B%A0%E5%90%8D"><span class="toc-text">8. 提取指定模块的基因名</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#9-%E6%A8%A1%E5%9D%97%E5%AF%BC%E5%87%BA"><span class="toc-text">9. 模块导出</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#Reference"><span class="toc-text">Reference</span></a></li></ol></li></ol></li></ol>
    </div>
  </div>

  
<script src="/js/catalog.js"></script>




    
      <div class="comments-container">
        


  <script src='//unpkg.com/valine/dist/Valine.min.js'></script>

  <div id="vcomments"></div>

  <script>
    new Valine({
      el: '#vcomments',
      appId: 'DdMocNMQxnEQcChwW38b0JVQ-gzGzoHsz',
      appKey: 'ETFXPtdWGEvlSfsH7RwSMzAf',
      placeholder: 'Welcome!',
      avatar: 'retro',
      lang: 'zh-CN'
    })
  </script>

    <style>
      .comments-container .v .vempty {
        display: none!important;
      }
    </style>




      </div>
    
  </div>


        <div class="footer">
  <div class="social">
    <ul>
      
        <li>
          <a title="github" target="_blank" rel="noopener" href="https://github.com/Moonerss">
            <i class="iconfont icon-github"></i>
          </a>
        </li>
      
        <li>
          <a title="rss" href="/atom.xml">
            <i class="iconfont icon-rss"></i>
          </a>
        </li>
      
    </ul>
  </div>
  
    <div class="footer-more">
      <a target="_blank" rel="noopener" href="https://github.com/Moonerss">Copyright © mooner 2020</a>
    </div>
  
    <div class="footer-more">
      <a target="_blank" rel="noopener" href="https://github.com/zchengsite/hexo-theme-oranges">Theme by Oranges | Powered by Hexo</a>
    </div>
  
</div>

      </div>

      <div class="back-to-top hidden">
  <a href="javascript: void(0)">
    <i class="iconfont icon-chevronup"></i>
  </a>
</div>


<script src="/js/backtotop.js"></script>



      
  <div class="search-icon" id="search-icon">
    <a href="javascript: void(0)">
      <i class="iconfont icon-search"></i>
    </a>
  </div>

  <div class="search-overlay hidden">
    <div class="search-content" tabindex="0">
      <div class="search-title">
        <span class="search-icon-input">
          <a href="javascript: void(0)">
            <i class="iconfont icon-search"></i>
          </a>
        </span>
        
          <input type="text" class="search-input" id="search-input" placeholder="Search ...">
        
        <span class="search-close-icon" id="search-close-icon">
          <a href="javascript: void(0)">
            <i class="iconfont icon-close"></i>
          </a>
        </span>
      </div>
      <div class="search-result" id="search-result"></div>
    </div>
  </div>

  <script type="text/javascript">
    var inputArea = document.querySelector("#search-input")
    var searchOverlayArea = document.querySelector(".search-overlay")

    inputArea.onclick = function() {
      getSearchFile()
      this.onclick = null
    }

    inputArea.onkeydown = function() {
      if(event.keyCode == 13)
        return false
    }

    function openOrHideSearchContent() {
      let isHidden = searchOverlayArea.classList.contains('hidden')
      if (isHidden) {
        searchOverlayArea.classList.remove('hidden')
        document.body.classList.add('hidden')
        // inputArea.focus()
      } else {
        searchOverlayArea.classList.add('hidden')
        document.body.classList.remove('hidden')
      }
    }

    function blurSearchContent(e) {
      if (e.target === searchOverlayArea) {
        openOrHideSearchContent()
      }
    }

    document.querySelector("#search-icon").addEventListener("click", openOrHideSearchContent, false)
    document.querySelector("#search-close-icon").addEventListener("click", openOrHideSearchContent, false)
    searchOverlayArea.addEventListener("click", blurSearchContent, false)

    var searchFunc = function (path, search_id, content_id) {
      'use strict';
      var $input = document.getElementById(search_id);
      var $resultContent = document.getElementById(content_id);
      $resultContent.innerHTML = "<ul><span class='local-search-empty'>首次搜索，正在载入索引文件，请稍后……<span></ul>";
      $.ajax({
        // 0x01. load xml file
        url: path,
        dataType: "xml",
        success: function (xmlResponse) {
          // 0x02. parse xml file
          var datas = $("entry", xmlResponse).map(function () {
            return {
              title: $("title", this).text(),
              content: $("content", this).text(),
              url: $("url", this).text()
            };
          }).get();
          $resultContent.innerHTML = "";

          $input.addEventListener('input', function () {
            // 0x03. parse query to keywords list
            var str = '<ul class=\"search-result-list\">';
            var keywords = this.value.trim().toLowerCase().split(/[\s\-]+/);
            $resultContent.innerHTML = "";
            if (this.value.trim().length <= 0) {
              return;
            }
            // 0x04. perform local searching
            datas.forEach(function (data) {
              var isMatch = true;
              var content_index = [];
              if (!data.title || data.title.trim() === '') {
                data.title = "Untitled";
              }
              var orig_data_title = data.title.trim();
              var data_title = orig_data_title.toLowerCase();
              var orig_data_content = data.content.trim().replace(/<[^>]+>/g, "");
              var data_content = orig_data_content.toLowerCase();
              var data_url = data.url;
              var index_title = -1;
              var index_content = -1;
              var first_occur = -1;
              // only match artiles with not empty contents
              if (data_content !== '') {
                keywords.forEach(function (keyword, i) {
                  index_title = data_title.indexOf(keyword);
                  index_content = data_content.indexOf(keyword);

                  if (index_title < 0 && index_content < 0) {
                    isMatch = false;
                  } else {
                    if (index_content < 0) {
                      index_content = 0;
                    }
                    if (i == 0) {
                      first_occur = index_content;
                    }
                    // content_index.push({index_content:index_content, keyword_len:keyword_len});
                  }
                });
              } else {
                isMatch = false;
              }
              // 0x05. show search results
              if (isMatch) {
                str += "<li><a href='" + data_url + "' class='search-result-title'>" + orig_data_title + "</a>";
                var content = orig_data_content;
                if (first_occur >= 0) {
                  // cut out 100 characters
                  var start = first_occur - 20;
                  var end = first_occur + 80;

                  if (start < 0) {
                    start = 0;
                  }

                  if (start == 0) {
                    end = 100;
                  }

                  if (end > content.length) {
                    end = content.length;
                  }

                  var match_content = content.substr(start, end);

                  // highlight all keywords
                  keywords.forEach(function (keyword) {
                    var regS = new RegExp(keyword, "gi");
                    match_content = match_content.replace(regS, "<span class=\"search-keyword\">" + keyword + "</span>");
                  });

                  str += "<p class=\"search-result-abstract\">" + match_content + "...</p>"
                }
                str += "</li>";
              }
            });
            str += "</ul>";
            if (str.indexOf('<li>') === -1) {
              return $resultContent.innerHTML = "<ul><span class='local-search-empty'>没有找到内容，请尝试更换检索词。<span></ul>";
            }
            $resultContent.innerHTML = str;
          });
        },
        error: function(xhr, status, error) {
          $resultContent.innerHTML = ""
          if (xhr.status === 404) {
            $resultContent.innerHTML = "<ul><span class='local-search-empty'>未找到search.xml文件，具体请参考：<a href='https://github.com/zchengsite/hexo-theme-oranges#configuration' target='_black'>configuration</a><span></ul>";
          } else {
            $resultContent.innerHTML = "<ul><span class='local-search-empty'>请求失败，尝试重新刷新页面或稍后重试。<span></ul>";
          }
        }
      });
      $(document).on('click', '#search-close-icon', function() {
        $('#search-input').val('');
        $('#search-result').html('');
      });
    }

    var getSearchFile = function() {
        var path = "/search.xml";
        searchFunc(path, 'search-input', 'search-result');
    }
  </script>




    </div>
  </body>
</html>
